// precise-ui/dist/es6/components/StackPanel/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled from '../../utils/styled';
export var StackPanelDirection;
(function (StackPanelDirection) {
    StackPanelDirection["leftToRight"] = "left";
    StackPanelDirection["rightToLeft"] = "right";
    StackPanelDirection["topToBottom"] = "top";
    StackPanelDirection["bottomToTop"] = "bottom";
})(StackPanelDirection || (StackPanelDirection = {}));
const StackLayout = styled.div `
  display: flex;
  flex-direction: ${props => props.dir};
  flex-wrap: ${props => props.wrapping};
  justify-content: flex-start;
`;
function mapFlex(direction) {
    switch (direction) {
        case StackPanelDirection.rightToLeft:
            return 'row-reverse';
        case StackPanelDirection.topToBottom:
            return 'column';
        case StackPanelDirection.bottomToTop:
            return 'column-reverse';
    }
    return 'row';
}
/**
 * The stack panel displays a stack of children in a given direction.
 */
export const StackPanel = Object.assign(((_a) => {
    var { theme, direction, wrap, innerRef } = _a, props = __rest(_a, ["theme", "direction", "wrap", "innerRef"]);
    const wrapping = wrap ? 'wrap' : 'nowrap';
    const dir = mapFlex(direction);
    return React.createElement(StackLayout, Object.assign({ theme: theme, dir: dir, wrapping: wrapping, ref: innerRef }, props));
}), { inner: {
        get StackLayout() { return StackLayout; }
    } });
StackPanel.displayName = 'StackPanel';
