// precise-ui/dist/es5/components/Notifications/notifier.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var types_1 = require("./types");
var eventManager_1 = require("../../utils/eventManager");
var defaultOptions = {
    type: 'none',
    autoClose: undefined,
};
function mergeOptions(options, type) {
    return Object.assign({}, defaultOptions, options, { type: type });
}
function emitEvent(_a, type) {
    var content = _a.content, _b = _a.options, options = _b === void 0 ? {} : _b;
    if (type === void 0) { type = 'none'; }
    var eventManagerOption = options.events;
    var eventManager = eventManagerOption || eventManager_1.eventManagers[0];
    eventManager.emit(types_1.NotificationAction.showNotification, {
        content: content,
        options: mergeOptions(options, type),
    });
}
exports.notify = Object.assign(function (payload) { return emitEvent(payload, payload.options && payload.options.type); }, {
    success: function (payload) { return emitEvent(payload, 'success'); },
    info: function (payload) { return emitEvent(payload, 'info'); },
    warn: function (payload) { return emitEvent(payload, 'warning'); },
    warning: function (payload) { return emitEvent(payload, 'warning'); },
    error: function (payload) { return emitEvent(payload, 'error'); },
});
