// precise-ui/dist/es5/components/Notifications/ToastNotification.part.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var styled_1 = require("../../utils/styled");
var distance_1 = require("../../distance");
var remCalc_1 = require("../../utils/remCalc");
var Notification_1 = require("../Notification");
var animationDuration = 200; // ms
var offsetDistance = 32;
function sideMultiplier(position) {
    switch (position) {
        case 'bottom-left':
        case 'top-left':
            return -1;
        case 'top-right':
        case 'bottom-right':
            return 1;
        default:
            return 0;
    }
}
var InAnimation = function (startOffset) { return styled_1.keyframes(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  from {\n    opacity: 0;\n    transform: translate(", "px);\n  }\n  to {\n    opacity: 1;\n    transform: translate(0px);\n  }\n"], ["\n  from {\n    opacity: 0;\n    transform: translate(", "px);\n  }\n  to {\n    opacity: 1;\n    transform: translate(0px);\n  }\n"])), startOffset); };
var OutAnimation = function (endOffset) { return styled_1.keyframes(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  from {\n    opacity: 1;\n    transform: translate(0px);\n  }\n  to {\n    opacity: 0;\n    transform: translate(", "px);\n  }\n"], ["\n  from {\n    opacity: 1;\n    transform: translate(0px);\n  }\n  to {\n    opacity: 0;\n    transform: translate(", "px);\n  }\n"])), endOffset); };
var StyledNotification = styled_1.default(Notification_1.Notification)(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  opacity: 0;\n  animation: ", "\n    ", "s ease-in-out forwards;\n"], ["\n  opacity: 0;\n  animation: ",
    "\n    ", "s ease-in-out forwards;\n"])), function (props) { return props.closing
    ? OutAnimation(sideMultiplier(props.position) * offsetDistance)
    : InAnimation(sideMultiplier(props.position) * offsetDistance); }, animationDuration / 1000);
var ToastNotification = /** @class */ (function (_super) {
    __extends(ToastNotification, _super);
    function ToastNotification(props) {
        var _this = _super.call(this, props) || this;
        _this.onCloseHandler = function () {
            _this.close();
            return false;
        };
        _this.close = function () {
            var onClose = _this.props.onClose;
            _this.setState({ closing: true }, function () {
                _this.closeId = setTimeout(function () {
                    if (typeof onClose === 'function') {
                        onClose();
                    }
                }, animationDuration);
            });
        };
        _this.state = {
            closing: false,
        };
        return _this;
    }
    ToastNotification.prototype.UNSAFE_componentWillMount = function () {
        var autoClose = this.props.autoClose;
        if (autoClose) {
            this.autoCloseId = setTimeout(this.close, autoClose);
        }
    };
    ToastNotification.prototype.componentWillUnmount = function () {
        if (this.closeId) {
            clearTimeout(this.closeId);
        }
        if (this.autoCloseId) {
            clearTimeout(this.autoCloseId);
        }
    };
    ToastNotification.prototype.render = function () {
        var closing = this.state.closing;
        var _a = this.props, _0 = _a.autoClose, _1 = _a.events, _2 = _a.id, _4 = _a.onClose, position = _a.position, other = __rest(_a, ["autoClose", "events", "id", "onClose", "position"]);
        var toastTheme = {
            notificationPadding: distance_1.distance.small + " " + distance_1.distance.small + " " + distance_1.distance.medium,
            notificationBorderWidth: "0 0 0 5px",
            notificationTextFontSize: remCalc_1.remCalc('12px'),
            notificationTextLineHeight: "14px",
            notificationIconMarginRight: distance_1.distance.small,
        };
        return (React.createElement(StyledNotification, __assign({ position: position, closing: closing, theme: toastTheme, onClose: this.onCloseHandler }, other)));
    };
    ToastNotification.inner = {
        get StyledNotification() { return StyledNotification; }
    };
    return ToastNotification;
}(React.Component));
exports.ToastNotification = ToastNotification;
var templateObject_1, templateObject_2, templateObject_3;
