// precise-ui/dist/es6/hoc/withValidation.js
import * as React from 'react';
import { withInner } from 'typescript-plugin-inner-jsx/withInner';
/**
 * Provides automatic validation for a wrapped component.
 * @param validate The validation function.
 * @returns A constructor function taking a component to be wrapped with the validation.
 */
export function withValidation(validate) {
    return (Component) => {
        return withInner(class Validator extends React.PureComponent {
            constructor(props) {
                super(props);
                this.validate = (e) => {
                    const { onChange, onError, onSuccess } = this.props;
                    const error = validate(e);
                    if (error !== this.state.error) {
                        const notify = error ? onError : onSuccess;
                        this.setState({
                            error,
                        });
                        if (typeof notify === 'function') {
                            notify();
                        }
                    }
                    if (typeof onChange === 'function') {
                        onChange(e);
                    }
                };
                this.state = {
                    error: undefined,
                };
            }
            render() {
                const { error = this.props.error } = this.state;
                return React.createElement(Component, Object.assign({}, this.props, { error: error, onChange: this.validate }));
            }
        }, { Component });
    };
}
