// precise-ui/dist/es6/hoc/withResponsive.js
import * as React from 'react';
import { withInner } from 'typescript-plugin-inner-jsx/withInner';
/**
 * Wraps the component in a responsive construct that is sensitive to the size changes of its
 * container.
 * @param Component The component to be sensitive to its container dimensions.
 * @returns A component that can use the `dimensions` prop.
 */
export function withResponsive(Component) {
    return withInner(class Responsive extends React.Component {
        constructor(props) {
            super(props);
            this.sizeChanged = () => {
                const { width, height } = this.getDimension();
                if (width !== this.state.width || height !== this.state.height) {
                    this.setState({
                        width,
                        height,
                    });
                }
            };
            this.orientationChanged = () => this.setState({
                angle: (screen.orientation && screen.orientation.angle) || 0,
            });
            this.setNode = (node) => (this.node = node);
            const screenWithOrientaion = screen;
            this.state = {
                width: window.innerWidth,
                height: window.innerHeight,
                angle: (screenWithOrientaion.orientation && screenWithOrientaion.orientation.angle) || 0,
            };
        }
        componentDidMount() {
            window.addEventListener('resize', this.sizeChanged);
            window.addEventListener('orientationchange', this.orientationChanged);
        }
        componentWillUnmount() {
            window.removeEventListener('resize', this.sizeChanged);
            window.removeEventListener('orientationchange', this.orientationChanged);
        }
        getDimension() {
            if (this.node) {
                const boundingClientRect = this.node.getBoundingClientRect();
                return {
                    width: boundingClientRect.width,
                    height: boundingClientRect.height,
                };
            }
            else {
                const { innerWidth, innerHeight } = window;
                return {
                    width: innerWidth,
                    height: innerHeight,
                };
            }
        }
        render() {
            const additionalProps = {
                innerRef: this.setNode,
                dimensions: this.state,
            };
            const props = Object.assign({}, this.props, additionalProps);
            return React.createElement(Component, Object.assign({}, props));
        }
    }, { Component });
}
