// precise-ui/dist/es6/components/Breadcrumbs/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled, { themed } from '../../utils/styled';
import { Breadcrumb } from '../Breadcrumb';
import { dark } from '../../colors';
import { distance } from '../../distance';
import { OverflowMenu } from '../OverflowMenu';
import { getFontStyle } from '../../textStyles';
const BreadcrumbContainer = styled.div `
  ${getFontStyle({ size: 'small' })}
  
  color: ${themed(props => props.theme.ui8)};
`;
const BreadcrumbSeparator = styled.span `
  padding: 0 ${distance.medium};
  color: ${dark};
  font-size: inherit;
  &:after {
    content: '/';
  }
`;
function collapse(items, target, size) {
    items[target] = (React.createElement(OverflowMenu, { items: items.splice(target, 1 + items.length - size, ''), button: React.createElement(Breadcrumb, { title: "...", href: "#" }) }));
}
function insertSeparators(items) {
    for (let i = items.length; i-- > 1;) {
        items.splice(i, 0, React.createElement(BreadcrumbSeparator, null));
    }
}
/**
 * Display a list of Breadcrumb elements with optional grouping.
 */
export const Breadcrumbs = Object.assign(((_a) => {
    var { size = 5, children } = _a, props = __rest(_a, ["size", "children"]);
    const displayElements = [];
    const count = Math.max(size, 1);
    React.Children.forEach(children, child => {
        if (child) {
            displayElements.push(child);
        }
    });
    if (displayElements.length > count) {
        collapse(displayElements, +(count > 2), count);
    }
    insertSeparators(displayElements);
    return (React.createElement(BreadcrumbContainer, Object.assign({}, props), displayElements.map((element, i) => (React.createElement("span", { key: i }, element)))));
}), { inner: {
        get BreadcrumbContainer() { return BreadcrumbContainer; }
    } });
Breadcrumbs.displayName = 'Breadcrumbs';
