// precise-ui/dist/es5/hoc/withResponsive.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var withInner_1 = require("typescript-plugin-inner-jsx/withInner");
/**
 * Wraps the component in a responsive construct that is sensitive to the size changes of its
 * container.
 * @param Component The component to be sensitive to its container dimensions.
 * @returns A component that can use the `dimensions` prop.
 */
function withResponsive(Component) {
    return withInner_1.withInner(/** @class */ (function (_super) {
        __extends(Responsive, _super);
        function Responsive(props) {
            var _this = _super.call(this, props) || this;
            _this.sizeChanged = function () {
                var _a = _this.getDimension(), width = _a.width, height = _a.height;
                if (width !== _this.state.width || height !== _this.state.height) {
                    _this.setState({
                        width: width,
                        height: height,
                    });
                }
            };
            _this.orientationChanged = function () { return _this.setState({
                angle: (screen.orientation && screen.orientation.angle) || 0,
            }); };
            _this.setNode = function (node) { return (_this.node = node); };
            var screenWithOrientaion = screen;
            _this.state = {
                width: window.innerWidth,
                height: window.innerHeight,
                angle: (screenWithOrientaion.orientation && screenWithOrientaion.orientation.angle) || 0,
            };
            return _this;
        }
        Responsive.prototype.componentDidMount = function () {
            window.addEventListener('resize', this.sizeChanged);
            window.addEventListener('orientationchange', this.orientationChanged);
        };
        Responsive.prototype.componentWillUnmount = function () {
            window.removeEventListener('resize', this.sizeChanged);
            window.removeEventListener('orientationchange', this.orientationChanged);
        };
        Responsive.prototype.getDimension = function () {
            if (this.node) {
                var boundingClientRect = this.node.getBoundingClientRect();
                return {
                    width: boundingClientRect.width,
                    height: boundingClientRect.height,
                };
            }
            else {
                var innerWidth_1 = window.innerWidth, innerHeight_1 = window.innerHeight;
                return {
                    width: innerWidth_1,
                    height: innerHeight_1,
                };
            }
        };
        Responsive.prototype.render = function () {
            var additionalProps = {
                innerRef: this.setNode,
                dimensions: this.state,
            };
            var props = __assign({}, this.props, additionalProps);
            return React.createElement(Component, __assign({}, props));
        };
        return Responsive;
    }(React.Component)), { Component: Component });
}
exports.withResponsive = withResponsive;
