// precise-ui/dist/es5/hoc/withResponsiveMode.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var withResponsive_1 = require("./withResponsive");
var withInner_1 = require("typescript-plugin-inner-jsx/withInner");
/**
 * Provides `mode` prop (if it's not passed manually) to the component depending on
 * the container width. When mode switches, HOC calls `onModeChange` prop if it is
 * defined.
 * @param getMode The function to determine the mode from the given width.
 * @returns A constructor function to wrap a component with the `mode` prop determined
 * by the getMode function.
 */
function withResponsiveMode(getMode) {
    return function (Component) {
        return withInner_1.withInner(withResponsive_1.withResponsive(/** @class */ (function (_super) {
            __extends(ModeProvider, _super);
            function ModeProvider(props) {
                var _this = _super.call(this, props) || this;
                _this.state = {
                    controlled: props.mode !== undefined,
                    mode: props.mode || getMode(props.dimensions && props.dimensions.width),
                };
                return _this;
            }
            ModeProvider.getDerivedStateFromProps = function (props, state) {
                var change = props.onModeChange;
                if (state.controlled) {
                    var mode = props.mode || getMode();
                    if (state.mode !== mode) {
                        if (typeof change === 'function') {
                            change({ mode: mode });
                        }
                        return {
                            mode: mode,
                        };
                    }
                }
                else {
                    var dimensions = props.dimensions;
                    var mode = state.mode;
                    var nextMode = getMode(dimensions && dimensions.width);
                    if (mode !== nextMode) {
                        if (typeof change === 'function') {
                            change({ mode: nextMode });
                        }
                        return {
                            mode: nextMode,
                        };
                    }
                }
                return {};
            };
            ModeProvider.prototype.componentDidMount = function () {
                var mode = this.state.mode;
                var onModeChange = this.props.onModeChange;
                if (mode && typeof onModeChange === 'function') {
                    onModeChange({ mode: mode });
                }
            };
            ModeProvider.prototype.render = function () {
                var mode = this.state.mode;
                var props = __assign({}, this.props, { mode: mode });
                return React.createElement(Component, __assign({}, props));
            };
            return ModeProvider;
        }(React.Component))), { Component: Component });
    };
}
exports.withResponsiveMode = withResponsiveMode;
