// precise-ui/dist/es6/components/Form/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled from '../../utils/styled';
import { Prompt } from '../Prompt';
import { FormContext } from '../../contexts';
const StyledForm = styled.form `
  margin: 0;
`;
function isDifferent(a, b) {
    if (a !== b) {
        if (Array.isArray(a) && Array.isArray(b)) {
            if (a.length === b.length) {
                for (let i = 0; i < a.length; i++) {
                    if (isDifferent(a[i], b[i])) {
                        return true;
                    }
                }
                return false;
            }
        }
        else if (typeof a === 'object' && typeof b === 'object') {
            const keysA = Object.keys(a);
            const keysB = Object.keys(b);
            if (keysA.length === keysB.length) {
                for (const key of keysA) {
                    if (isDifferent(a[key], b[key])) {
                        return true;
                    }
                }
                return false;
            }
        }
        return true;
    }
    return false;
}
function isChanged(initial, current) {
    const keys = Object.keys(current);
    for (const key of keys) {
        if (isDifferent(current[key], initial[key])) {
            return true;
        }
    }
    return false;
}
/**
 * Represents a field aggregator that enables easily creating forms.
 */
export class Form extends React.Component {
    constructor(props) {
        super(props);
        this.fields = [];
        this.ctx = this.createContext();
        this.submit = (e) => {
            const { onSubmit, disabled } = this.props;
            const { current, changed, errors } = this.state;
            this.setErrors(current);
            if (!disabled && typeof onSubmit === 'function') {
                const arrayErrors = Object.keys(errors).reduce((arrayErrors, field) => {
                    const error = errors[field];
                    if (error) {
                        arrayErrors.push({ field, error });
                    }
                    return arrayErrors;
                }, []);
                this.setState({
                    changed: false,
                    initial: current,
                }, () => onSubmit({
                    data: current,
                    errors: arrayErrors,
                    changed,
                }));
            }
            e.preventDefault();
            return false;
        };
        const data = props.value || props.defaultValue || {};
        this.state = {
            changed: false,
            controlled: props.value !== undefined,
            initial: data,
            current: data,
            errors: {},
        };
    }
    UNSAFE_componentWillReceiveProps(nextProps) {
        const { controlled, initial } = this.state;
        if (controlled) {
            const { value = {} } = nextProps;
            const changed = isChanged(initial, value);
            this.setValues(value, changed);
        }
    }
    setValues(current, changed) {
        const keys = Object.keys(current);
        this.setState({
            current,
            changed,
        });
        for (const key of keys) {
            const value = current[key];
            for (const field of this.fields) {
                if (field.props.name === key && field.state.value !== value) {
                    field.setState({
                        value,
                    });
                }
            }
        }
    }
    getError(name, value) {
        const validator = this.props.validationRules && this.props.validationRules[name];
        const validationResult = validator ? validator(value) : true;
        const error = validationResult === true ? undefined : validationResult;
        return error;
    }
    setFieldError(name, error) {
        for (const fieldEntity of this.fields) {
            if (fieldEntity.props.name === name) {
                fieldEntity.setState({ error });
                return;
            }
        }
    }
    setError({ name, value }) {
        const error = this.getError(name, value);
        this.setFieldError(name, error);
        this.setState({ errors: Object.assign({}, this.state.errors, { [name]: error }) });
    }
    setErrors(current) {
        const keys = Object.keys(current);
        const errors = Object.assign({}, this.state.errors);
        for (const key of keys) {
            const value = current[key];
            const error = this.getError(key, value);
            errors[key] = error;
            this.setFieldError(key, error);
        }
        this.setState({ errors });
    }
    createContext() {
        return {
            change: (field) => {
                const { onChange } = this.props;
                const { controlled, current, initial } = this.state;
                const proposed = Object.assign({}, current, { [field.name]: field.value });
                const changed = isChanged(initial, proposed);
                if (!controlled) {
                    this.setValues(proposed, changed);
                }
                this.setError(field);
                if (typeof onChange === 'function') {
                    onChange({
                        changed,
                        value: proposed,
                    });
                }
            },
            subscribe: (field) => {
                const { current, errors = {} } = this.state;
                const { name } = field.props;
                if (name) {
                    this.fields.push(field);
                    let error;
                    if (name in current) {
                        const value = current[name];
                        error = this.getError(name, value);
                        field.setState({
                            value,
                        });
                    }
                    else {
                        const value = field.state.value;
                        current[name] = value;
                        error = this.getError(name, value);
                    }
                    if (error) {
                        this.setState({ errors: Object.assign({}, errors, { [name]: error }) });
                    }
                }
            },
            unsubscribe: (field) => {
                const index = this.fields.indexOf(field);
                index >= 0 && this.fields.splice(index, 1);
            },
        };
    }
    render() {
        const _a = this.props, { value: _0, defaultValue: _1, onChange: _2, onSubmit: _3, disabled: _4, children, prompt } = _a, rest = __rest(_a, ["value", "defaultValue", "onChange", "onSubmit", "disabled", "children", "prompt"]);
        const { changed } = this.state;
        return (React.createElement(StyledForm, Object.assign({}, rest, { onSubmit: this.submit }),
            prompt && (typeof prompt === 'function' ? prompt(changed) : React.createElement(Prompt, { when: changed, message: prompt })),
            React.createElement(FormContext.Provider, { value: this.ctx }, children)));
    }
}
Form.inner = {
    get StyledForm() { return StyledForm; },
    get Prompt() { return Prompt; }
};
