// precise-ui/dist/es6/components/MetroRadioTile/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import * as colors from '../../colors';
import styled from '../../utils/styled';
import { RadioButton } from '../RadioButton';
import { StyledTileIcon } from '../../quarks/StyledTileIcon';
import { getFontStyle } from '../../textStyles';
const Tile = styled(RadioButton) `
  cursor: pointer;
  position: relative;
  display: inline-block;
  width: 96px;
  height: 96px;
  padding: 8px;
  overflow: hidden;
  margin: 10px !important;
  user-select: none;
  color: ${colors.black};
  background: ${colors.grey7};
  ${props => (props.checked ? `outline: ${colors.cyan} solid 4px;` : '')};

  &:hover {
    ${props => (!props.checked ? `outline: rgba(29, 29, 29, 0.1) solid 4px;` : '')};
  }
`;
const TileTitle = styled.span `
  ${getFontStyle({ size: 'small', weight: 'medium' })}

  position: absolute;
  bottom: 8px;
  right: 8px;

  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  z-index: 2;
`;
export class MetroRadioTile extends React.Component {
    constructor(props) {
        super(props);
        this.change = (e) => {
            const { onChange } = this.props;
            if (!this.state.controlled) {
                this.setState({
                    checked: e.value,
                });
            }
            if (typeof onChange === 'function') {
                onChange(e);
            }
        };
        this.state = {
            controlled: props.value !== undefined,
            checked: props.value || props.defaultValue || false,
        };
    }
    UNSAFE_componentWillReceiveProps(props) {
        if (this.state.controlled) {
            this.setState({
                checked: props.value || false,
            });
        }
    }
    render() {
        const _a = this.props, { image, title } = _a, props = __rest(_a, ["image", "title"]);
        const { checked } = this.state;
        return (React.createElement(Tile, Object.assign({ onChange: this.change, checked: checked }, props),
            image && React.createElement(StyledTileIcon, { src: image, width: "50px", height: "50px" }),
            title && React.createElement(TileTitle, null, title)));
    }
}
MetroRadioTile.inner = {
    get Tile() { return Tile; },
    get StyledTileIcon() { return StyledTileIcon; },
    get TileTitle() { return TileTitle; }
};
