// precise-ui/dist/es6/components/GridArea/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled from '../../utils/styled';
const GridAreaContainer = styled.div `
  height: 100%;
`;
/**
 * The GridArea component to declare areas within a Grid instance.
 */
export class GridArea extends React.PureComponent {
    constructor() {
        super(...arguments);
        this.clicked = (e) => {
            this.tapped(e.currentTarget, e.clientX, e.clientY);
        };
        this.touched = (e) => {
            const touch = e.touches[0];
            this.tapped(e.currentTarget, touch.clientX, touch.clientY);
        };
    }
    tapped(node, clientX, clientY) {
        const { onTap } = this.props;
        if (typeof onTap === 'function') {
            const r = node.getBoundingClientRect();
            onTap({
                node,
                x: clientX - r.left,
                y: clientY - r.top,
                width: r.width,
                height: r.height,
            });
        }
    }
    render() {
        const _a = this.props, { colSpan, rowSpan, column, row, onTap, hidden } = _a, props = __rest(_a, ["colSpan", "rowSpan", "column", "row", "onTap", "hidden"]);
        return React.createElement(GridAreaContainer, Object.assign({}, props, { onMouseDown: this.clicked, onTouchStartCapture: this.touched }));
    }
}
GridArea.inner = {
    get GridAreaContainer() { return GridAreaContainer; }
};
