// precise-ui/dist/es6/components/Expander/index.js
import * as React from 'react';
import styled from '../../utils/styled';
function getExpandingContainerHeight(status, expandedHeight, collapsedHeight) {
    if (!expandedHeight) {
        return '0';
    }
    switch (status) {
        case 'entering':
            return `${expandedHeight}px`;
        case 'entered':
            return 'auto';
        case 'exitingStart':
            return `${expandedHeight}px`;
        case 'exiting':
        default:
            return `${collapsedHeight || '0'}`;
    }
}
const ExpandingContainer = styled('div')(({ status, expandedHeight, collapsedHeight, timeout }) => `
    overflow: hidden;
    transition-duration: ${timeout}ms;
    height: ${getExpandingContainerHeight(status, expandedHeight, collapsedHeight)}
  `);
/**
 * A functional expander component for expanding / collapsing content.
 */
export class Expander extends React.Component {
    constructor(props) {
        super(props);
        this.wrapperRef = React.createRef();
        this.setContainer = (ref) => {
            this.rootContainer = ref;
        };
        this.state = {
            status: 'unmounted',
        };
    }
    componentDidMount() {
        this.handleChange();
    }
    componentDidUpdate(prevProps, prevState) {
        const { expand } = this.props;
        const { status } = this.state;
        if (prevProps.expand !== expand || prevState.status !== status) {
            this.handleChange();
        }
    }
    componentWillUnmount() {
        this.clearTimer();
    }
    handleChange() {
        const { expand } = this.props;
        const { status } = this.state;
        if (expand) {
            if (status === 'unmounted') {
                this.setState({
                    status: 'mounted',
                });
            }
            else if (status !== 'entering' && status !== 'entered') {
                this.performEnter();
            }
        }
        else {
            if (status === 'entering' || status === 'entered') {
                this.performExit();
            }
        }
    }
    performEnter() {
        this.setState({
            status: 'entering',
        }, () => this.setDelayedStatus('entered'));
    }
    performExit() {
        const { unmountOnExit } = this.props;
        this.setState({ status: 'exitingStart' }, () => this.setState({ status: 'exiting' }, () => this.setDelayedStatus(unmountOnExit ? 'unmounted' : 'exited')));
    }
    clearTimer() {
        if (this.timer) {
            clearTimeout(this.timer);
        }
    }
    setDelayedStatus(status) {
        const { timeout = 0 } = this.props;
        this.clearTimer();
        this.timer = setTimeout(() => {
            this.setState({
                status,
            });
        }, timeout);
    }
    render() {
        const { status } = this.state;
        const { children, collapsedHeight, timeout = 0 } = this.props;
        const contentRef = this.wrapperRef.current;
        if (status !== 'unmounted') {
            return (React.createElement(ExpandingContainer, { ref: this.setContainer, expandedHeight: contentRef ? contentRef.clientHeight : undefined, collapsedHeight: collapsedHeight, status: status, timeout: timeout, style: { minHeight: collapsedHeight } },
                React.createElement("div", { ref: this.wrapperRef }, children)));
        }
        return false;
    }
}
Expander.inner = {
    get ExpandingContainer() { return ExpandingContainer; }
};
