// precise-ui/dist/es5/components/Modal/index.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var styled_1 = require("../../utils/styled");
var Headline_1 = require("../Headline");
var colors_1 = require("../../colors");
var Blocker_1 = require("../Blocker");
var CloseButton_1 = require("../CloseButton");
var distance_1 = require("../../distance");
var textStyles_1 = require("../../textStyles");
var openAnimationDuration = 300;
var closeAnimationDuration = 200;
var blockerAnimationDuration = 200;
var InAnimation = function (startOffset) { return styled_1.keyframes(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  from {\n    opacity: 0;\n    transform: translate(0, ", "px);\n  }\n  to {\n    opacity: 1;\n    transform: translate(0px);\n  }\n"], ["\n  from {\n    opacity: 0;\n    transform: translate(0, ", "px);\n  }\n  to {\n    opacity: 1;\n    transform: translate(0px);\n  }\n"])), startOffset); };
var OutAnimation = function () { return styled_1.keyframes(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  from {\n    opacity: 1;\n  }\n  to {\n    opacity: 0;\n  }\n"], ["\n  from {\n    opacity: 1;\n  }\n  to {\n    opacity: 0;\n  }\n"]))); };
var BlockerInAnimation = function () { return styled_1.keyframes(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  from {\n    opacity: 0;\n  }\n  to {\n    opacity: 1;\n  }\n"], ["\n  from {\n    opacity: 0;\n  }\n  to {\n    opacity: 1;\n  }\n"]))); };
var StyledModal = styled_1.default('div')(function (_a) {
    var width = _a.width, closing = _a.closing;
    return styled_1.css(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n    outline: none;\n    color: ", ";\n    ", ";\n    margin: ", " auto;\n    display: flex;\n    align-items: center;\n    min-height: calc(100% - (", " * 2));\n    animation: ", "\n      ", "ms cubic-bezier(0, 0, 0.25, 1);\n    animation-fill-mode: forwards;\n\n    @media screen and (max-width: ", ") {\n      width: 100%;\n      min-height: 100%;\n      margin: 0;\n      align-items: stretch;\n    }\n  "], ["\n    outline: none;\n    color: ", ";\n    ", ";\n    margin: ", " auto;\n    display: flex;\n    align-items: center;\n    min-height: calc(100% - (", " * 2));\n    animation: ", "\n      ", "ms cubic-bezier(0, 0, 0.25, 1);\n    animation-fill-mode: forwards;\n\n    @media screen and (max-width: ", ") {\n      width: 100%;\n      min-height: 100%;\n      margin: 0;\n      align-items: stretch;\n    }\n  "])), colors_1.dark, width ? "width: " + width : 'max-width: 500px', distance_1.distance.xlarge, distance_1.distance.xlarge, closing ? OutAnimation() : InAnimation(-72), closing ? closeAnimationDuration : openAnimationDuration, width || '500px');
});
var StyledBlocker = styled_1.default(Blocker_1.Blocker)(function (_a) {
    var closing = _a.closing;
    return styled_1.css(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n    animation: ", " ", "ms\n      cubic-bezier(0, 0, 0.25, 1);\n    animation-fill-mode: forwards;\n  "], ["\n    animation: ", " ", "ms\n      cubic-bezier(0, 0, 0.25, 1);\n    animation-fill-mode: forwards;\n  "])), closing ? OutAnimation() : BlockerInAnimation(), blockerAnimationDuration);
});
var ModalContent = styled_1.default.div(function (_a) {
    var minHeight = _a.minHeight;
    return "\n    position: relative;\n    display: flex;\n    flex-direction: column;\n    background: " + colors_1.white + ";\n    box-shadow: 0 12px 24px 0 rgba(0, 0, 0, 0.1);\n    border-top: 4px solid " + colors_1.cyan + ";\n    width: 100%;\n    " + (minHeight ? "min-height: " + minHeight : '') + ";\n  ";
});
var ModalLabel = styled_1.default.span(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  ", "\n\n  padding-bottom: ", ";\n"], ["\n  ", "\n\n  padding-bottom: ", ";\n"])), textStyles_1.getFontStyle({ size: 'small' }), distance_1.distance.small);
var StyledModalBody = styled_1.default.div(templateObject_7 || (templateObject_7 = __makeTemplateObject(["\n  ", "\n\n  text-align: left;\n  padding: 0 ", " ", ";\n  margin-bottom: auto;\n"], ["\n  ", "\n\n  text-align: left;\n  padding: 0 ", " ", ";\n  margin-bottom: auto;\n"])), textStyles_1.getFontStyle({ size: 'medium', weight: 'regular' }), distance_1.distance.xxlarge, distance_1.distance.xxlarge);
var StyledModalHeader = styled_1.default.div(templateObject_8 || (templateObject_8 = __makeTemplateObject(["\n  margin-bottom: ", ";\n  padding: ", " ", " 0;\n"], ["\n  margin-bottom: ", ";\n  padding: ", " ", " 0;\n"])), distance_1.distance.medium, distance_1.distance.xxlarge, distance_1.distance.xxlarge);
var StyledModalFooter = styled_1.default.div(templateObject_9 || (templateObject_9 = __makeTemplateObject(["\n  background: ", ";\n  padding: ", " ", ";\n  text-align: right;\n"], ["\n  background: ", ";\n  padding: ", " ", ";\n  text-align: right;\n"])), colors_1.grey6, distance_1.distance.large, distance_1.distance.xxlarge);
/**
 * A simple modal dialog for requiring user interaction.
 */
var Modal = /** @class */ (function (_super) {
    __extends(Modal, _super);
    function Modal(props) {
        var _this = _super.call(this, props) || this;
        _this.closeBackground = function () {
            _this.closeFrom('background');
        };
        _this.closeButton = function () {
            _this.closeFrom('button');
        };
        _this.state = {
            closing: false,
        };
        return _this;
    }
    Modal.prototype.closeFrom = function (origin) {
        var _this = this;
        var _a = this.props, onClose = _a.onClose, onBeforeClose = _a.onBeforeClose;
        var closeEvent = { origin: origin };
        this.callOnBeforeCloseHandler(onBeforeClose, closeEvent).then(function (closePermitted) {
            if (!closePermitted) {
                return;
            }
            _this.setState({ closing: true }, function () { return setTimeout(function () {
                if (typeof onClose === 'function') {
                    onClose(closeEvent);
                }
                _this.setState({ closing: false });
            }, closeAnimationDuration); });
        });
    };
    Modal.prototype.callOnBeforeCloseHandler = function (onBeforeClose, closeEvent) {
        if (typeof onBeforeClose !== 'function') {
            return Promise.resolve(true);
        }
        var canClose = onBeforeClose(closeEvent);
        return Promise.resolve(canClose);
    };
    Modal.prototype.render = function () {
        var _a = this.props, _0 = _a.onBeforeClose, children = _a.children, onClose = _a.onClose, _b = _a.open, open = _b === void 0 ? false : _b, minHeight = _a.minHeight, rest = __rest(_a, ["onBeforeClose", "children", "onClose", "open", "minHeight"]);
        var canClose = typeof onClose === 'function';
        var closing = this.state.closing;
        return (open && (React.createElement(StyledBlocker, { closing: closing, onClose: this.closeBackground },
            React.createElement(StyledModal, __assign({ tabIndex: 0, closing: closing }, rest),
                React.createElement(ModalContent, { minHeight: minHeight },
                    children,
                    canClose && React.createElement(CloseButton_1.CloseButton, { onClick: this.closeButton }))))));
    };
    Modal.inner = {
        get StyledBlocker() { return StyledBlocker; },
        get StyledModal() { return StyledModal; },
        get ModalContent() { return ModalContent; },
        get CloseButton() { return CloseButton_1.CloseButton; }
    };
    return Modal;
}(React.PureComponent));
exports.Modal = Modal;
/**
 * Styles the body of a modal dialog.
 */
exports.ModalBody = Object.assign((function (props) { return (React.createElement(StyledModalBody, __assign({}, props))); }), { inner: {
        get StyledModalBody() { return StyledModalBody; }
    } });
exports.ModalBody.displayName = 'ModalBody';
/**
 * Styles the header of a modal dialog
 */
exports.ModalHeader = Object.assign((function (_a) {
    var title = _a.title, label = _a.label, rest = __rest(_a, ["title", "label"]);
    return (React.createElement(StyledModalHeader, __assign({}, rest),
        label && React.createElement(ModalLabel, null, label),
        title && React.createElement(Headline_1.Headline, { level: 2 }, title)));
}), { inner: {
        get StyledModalHeader() { return StyledModalHeader; },
        get ModalLabel() { return ModalLabel; },
        get Headline() { return Headline_1.Headline; }
    } });
exports.ModalHeader.displayName = 'ModalHeader';
/**
 * Styles the footer of a modal dialog.
 */
exports.ModalFooter = Object.assign((function (props) { return React.createElement(StyledModalFooter, __assign({}, props)); }), { inner: {
        get StyledModalFooter() { return StyledModalFooter; }
    } });
exports.ModalFooter.displayName = 'ModalFooter';
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8, templateObject_9;
