// precise-ui/dist/es5/colors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cyan = 'rgba(0,139,208,1)';
exports.white = 'rgba(255,255,255,1)';
exports.whiteSmoke = 'rgba(239,239,239,1)';
exports.whiterSmoke = 'rgba(248,248,248,1)';
exports.lighterGray = 'rgba(211,211,211,1)';
exports.tuna = 'rgba(75,78,82,1)';
exports.eclipse = 'rgba(62,62,62,1)';
exports.charcoal = 'rgba(114,114,114,1)';
exports.pinkSwan = 'rgba(180,180,180,1)';
exports.transparent = 'rgba(255, 255, 255, 0)';
exports.black = 'rgba(0,0,0,1)';
exports.steel = 'rgba(116,118,120,1)';
exports.teal = 'rgba(77,144,161,1)';
exports.blue = 'rgba(0,94,210,1)';
exports.darkGray = 'rgba(132,132,132,1)';
exports.dark = 'rgba(64,64,64,1)';
exports.pacificBlue = exports.cyan;
exports.grey1 = 'rgba(77,79,83,1)';
exports.grey2 = 'rgba(116,118,120,1)';
exports.grey3 = 'rgba(139,141,142,1)';
exports.grey4 = 'rgba(154,155,156,1)';
exports.grey5 = 'rgba(188,189,188,1)';
exports.grey6 = 'rgba(224,225,221,1)';
exports.grey6m = 'rgba(223,227,230,1)';
exports.grey7 = 'rgba(242,242,242,1)';
exports.brightBlue = 'rgba(235,240,247,1)';
exports.lightGray = 'rgba(178,178,178,1)';
exports.disabledBlue = 'rgba(203,228,245,1)';
exports.indigo = 'rgba(20,30,140,1)';
exports.azur = 'rgba(5,94,210,1)';
exports.azur1 = '#9bbeed';
exports.azur2 = '#699ee4';
exports.azur3 = '#377edb';
exports.azur4 = '#044ba8';
exports.azur5 = '#03387e';
exports.azur6 = '#022554';
exports.skyBlue = 'rgba(106,176,226,1)';
exports.skyBlue1 = '#c3dff3';
exports.skyBlue2 = '#a5cfed';
exports.skyBlue3 = '#87bfe7';
exports.skyBlue4 = '#548cb4';
exports.skyBlue5 = '#3f6987';
exports.skyBlue6 = '#2a465a';
exports.green = 'rgba(60,138,46,1)';
exports.green1 = '#b1d0ab';
exports.green2 = '#8ab881';
exports.green3 = '#63a157';
exports.green4 = '#306e24';
exports.green5 = '#24521b';
exports.green6 = '#183712';
exports.greenNeon = 'rgba(225,255,50,1)';
exports.lightGreen = 'rgba(217,233,6,1)';
exports.brightLemon = 'rgba(254,203,0,1)';
exports.orange = 'rgba(236,101,0,1)';
exports.orange1 = '#f7c199';
exports.orange2 = '#f3a266';
exports.orange3 = '#ef8333';
exports.orange4 = '#bc5000';
exports.orange5 = '#8d3c00';
exports.orange6 = '#5e5800';
exports.orangeNeon = 'rgba(255,26,0,1)';
exports.red1 = '#ffa399';
exports.red2 = '#ff7566';
exports.red3 = '#ff4733';
exports.red4 = '#cc1400';
exports.red5 = '#990f00';
exports.red6 = '#660a00';
exports.purpleRed = 'rgba(167,2,64,1)';
exports.purpleRed1 = '#edccd9';
exports.purpleRed2 = '#d381a0';
exports.purpleRed3 = '#b93566';
exports.purpleRed4 = '#96023a';
exports.purpleRed5 = '#75012d';
exports.purpleRed6 = '#43011a';
exports.midnight = 'rgba(0,80,123,1)';
exports.ocean = 'rgba(0,102,156,1)';
exports.all = {
    cyan: exports.cyan,
    white: exports.white,
    whiteSmoke: exports.whiteSmoke,
    whiterSmoke: exports.whiterSmoke,
    lighterGray: exports.lighterGray,
    tuna: exports.tuna,
    eclipse: exports.eclipse,
    charcoal: exports.charcoal,
    pinkSwan: exports.pinkSwan,
    transparent: exports.transparent,
    black: exports.black,
    steel: exports.steel,
    teal: exports.teal,
    blue: exports.blue,
    darkGray: exports.darkGray,
    dark: exports.dark,
    pacificBlue: exports.pacificBlue,
    grey1: exports.grey1,
    grey2: exports.grey2,
    grey3: exports.grey3,
    grey4: exports.grey4,
    grey5: exports.grey5,
    grey6: exports.grey6,
    grey6m: exports.grey6m,
    grey7: exports.grey7,
    brightBlue: exports.brightBlue,
    lightGray: exports.lightGray,
    disabledBlue: exports.disabledBlue,
    indigo: exports.indigo,
    azur: exports.azur,
    azur1: exports.azur1,
    azur2: exports.azur2,
    azur3: exports.azur3,
    azur4: exports.azur4,
    azur5: exports.azur5,
    azur6: exports.azur6,
    skyBlue: exports.skyBlue,
    skyBlue1: exports.skyBlue1,
    skyBlue2: exports.skyBlue2,
    skyBlue3: exports.skyBlue3,
    skyBlue4: exports.skyBlue4,
    skyBlue5: exports.skyBlue5,
    skyBlue6: exports.skyBlue6,
    green: exports.green,
    green1: exports.green1,
    green2: exports.green2,
    green3: exports.green3,
    green4: exports.green4,
    green5: exports.green5,
    green6: exports.green6,
    greenNeon: exports.greenNeon,
    lightGreen: exports.lightGreen,
    brightLemon: exports.brightLemon,
    orange: exports.orange,
    orange1: exports.orange1,
    orange2: exports.orange2,
    orange3: exports.orange3,
    orange4: exports.orange4,
    orange5: exports.orange5,
    orange6: exports.orange6,
    orangeNeon: exports.orangeNeon,
    red1: exports.red1,
    red2: exports.red2,
    red3: exports.red3,
    red4: exports.red4,
    red5: exports.red5,
    red6: exports.red6,
    purpleRed: exports.purpleRed,
    purpleRed1: exports.purpleRed1,
    purpleRed2: exports.purpleRed2,
    purpleRed3: exports.purpleRed3,
    purpleRed4: exports.purpleRed4,
    purpleRed5: exports.purpleRed5,
    purpleRed6: exports.purpleRed6,
    midnight: exports.midnight,
    ocean: exports.ocean,
};
