// precise-ui/dist/es6/components/PasswordField/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled from '../../utils/styled';
import { TextField } from '../TextField';
import { IconLink } from '../IconLink';
const StyledTextField = styled(TextField) `
  font-family: ${props => (props.reveal ? 'inherit' : 'sans-serif')};
`;
/**
 * A password field for custom user input.
 */
export class PasswordField extends React.Component {
    constructor(props) {
        super(props);
        this.showPassword = (e) => {
            this.setState({
                reveal: true,
            });
            e.preventDefault();
        };
        this.hidePassword = (e) => {
            this.setState({
                reveal: false,
            });
            e.preventDefault();
        };
        this.state = {
            reveal: false,
        };
    }
    renderDefaultSuffix(reveal, theme) {
        const iconName = reveal ? 'VisibilityOff' : 'Visibility';
        return React.createElement(IconLink, { icon: iconName, onClick: reveal ? this.hidePassword : this.showPassword, theme: theme });
    }
    render() {
        const _a = this.props, { theme, suffix } = _a, rest = __rest(_a, ["theme", "suffix"]);
        const { reveal } = this.state;
        const props = Object.assign({ suffix: suffix === undefined ? this.renderDefaultSuffix(reveal, theme) : suffix, type: reveal ? 'text' : 'password', theme,
            reveal }, rest);
        return React.createElement(StyledTextField, Object.assign({}, props));
    }
}
PasswordField.inner = {
    get IconLink() { return IconLink; },
    get StyledTextField() { return StyledTextField; }
};
