// precise-ui/dist/es6/components/MetroInfoTile/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled, { themed } from '../../utils/styled';
import { Anchor } from '../Anchor';
import { StyledTileIcon } from '../../quarks';
import { getFontStyle } from '../../textStyles';
const Tile = styled(Anchor) `
  cursor: pointer;
  display: block;
  position: relative;
  width: ${themed(({ theme }) => theme.metroInfoTile.size)};
  height: ${themed(({ theme }) => theme.metroInfoTile.size)};
  margin: 0 12px 12px 0;
  overflow: hidden;
  user-select: none;
  background: ${themed(({ theme }) => theme.metroInfoTile.background)};
  color: ${themed(({ theme }) => theme.metroInfoTile.textColor)};

  &:hover {
    outline: rgba(29, 29, 29, 0.1) solid 4px;
  }
`;
const TileTitle = styled.span `
  ${getFontStyle({ size: 'small', weight: 'medium' })}

  position: absolute;
  left: 0;
  bottom: 0;
  right: 0;
  padding: 0 10px 10px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  z-index: 2;
`;
/**
 * Provides a simple tile component that gives a flat info appareance.
 */
export const MetroInfoTile = Object.assign(((_a) => {
    var { children, image, title } = _a, props = __rest(_a, ["children", "image", "title"]);
    return (React.createElement(Tile, Object.assign({}, props),
        image && React.createElement(StyledTileIcon, { src: image, width: "33%", height: "33%" }),
        title && React.createElement(TileTitle, null, title),
        children));
}), { inner: {
        get Tile() { return Tile; },
        get StyledTileIcon() { return StyledTileIcon; },
        get TileTitle() { return TileTitle; }
    } });
MetroInfoTile.displayName = 'MetroInfoTile';
