// precise-ui/dist/es6/components/MenuListItem/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled, { themed, css } from '../../utils/styled';
import { distance } from '../../distance';
const ActiveStyle = css `
  &:before {
    position: absolute;
    content: '';
    width: 4px;
    height: 100%;
    background: ${themed(({ theme }) => theme.ui0)};
    left: 0;
  }
`;
const StyledMenuListItem = styled('li') `
  display: flex;
  position: relative;
  align-items: center;
  color: ${themed(props => props.theme.text6)};
  text-decoration: none;
  height: ${({ condensed }) => (condensed ? distance.xlarge : distance.xxlarge)};
  padding: 0 0 0 ${distance.large};
  ${({ active }) => (active ? ActiveStyle : '')};
`;
export const MenuListItem = Object.assign(((_a) => {
    var { active = false, condensed = false } = _a, props = __rest(_a, ["active", "condensed"]);
    return (React.createElement(StyledMenuListItem, Object.assign({}, props, { condensed: condensed, active: active })));
}), { inner: {
        get StyledMenuListItem() { return StyledMenuListItem; }
    } });
MenuListItem.displayName = 'MenuListItem';
