// precise-ui/dist/es5/components/DropdownPanel/DropdownPanel.part.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var styled_1 = require("../../utils/styled");
var InteractiveList_1 = require("../InteractiveList");
var colors_1 = require("../../colors");
var Icon_1 = require("../Icon");
var textStyles_1 = require("../../textStyles");
var react_onclickoutside_1 = require("react-onclickoutside");
var StyledWrapper = styled_1.default.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  box-sizing: border-box;\n  box-shadow: none;\n  margin: 0 0 ", " 0;\n  padding: 0;\n  border: 1px solid ", ";\n"], ["\n  box-sizing: border-box;\n  box-shadow: none;\n  margin: 0 0 ", " 0;\n  padding: 0;\n  border: 1px solid ", ";\n"])), function (props) { return (props.open ? '10px' : '0'); }, styled_1.themed(function (_a) {
    var theme = _a.theme, open = _a.open;
    return (open ? theme.ui0 : colors_1.transparent);
}));
var MenuCaret = styled_1.default.div(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  position: absolute;\n  top: 8px;\n  right: 9px;\n  fill: rgba(0, 139, 208, 1);\n"], ["\n  position: absolute;\n  top: 8px;\n  right: 9px;\n  fill: rgba(0, 139, 208, 1);\n"])));
var MenuTitle = styled_1.default.span(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  ", "\n\n  border-bottom: 2px solid black;\n"], ["\n  ", "\n\n  border-bottom: 2px solid black;\n"])), textStyles_1.getFontStyle({ size: 'large', weight: 'regular' }));
var SelectedMenu = styled_1.default.div(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  position: relative;\n  padding: 10px;\n  cursor: pointer;\n"], ["\n  position: relative;\n  padding: 10px;\n  cursor: pointer;\n"])));
var CustomWrapper = function (_a) {
    var _0 = _a.border, _1 = _a.direction, _2 = _a.open, props = __rest(_a, ["border", "direction", "open"]);
    return (React.createElement("div", __assign({}, props)));
};
var SlideDownTabsInt = /** @class */ (function (_super) {
    __extends(SlideDownTabsInt, _super);
    function SlideDownTabsInt(props) {
        var _this = _super.call(this, props) || this;
        _this.hide = function () {
            _this.setState({ isOpen: false });
        };
        _this.toggle = function (e) {
            _this.setState({ isOpen: !_this.state.isOpen });
            e.preventDefault();
        };
        _this.onChange = function (_a) {
            var value = _a.value;
            _this.props.onChange(value[0]);
            _this.setState({
                isOpen: false,
            });
        };
        _this.handleClickOutside = _this.hide;
        _this.state = {
            isOpen: false,
        };
        return _this;
    }
    SlideDownTabsInt.prototype.render = function () {
        var isOpen = this.state.isOpen;
        var _a = this.props, data = _a.data, selectedIndex = _a.selectedIndex;
        var interactiveList = data.map(function (item, index) { return (index === selectedIndex ? undefined : item); });
        return (React.createElement(StyledWrapper, { open: isOpen },
            React.createElement(SelectedMenu, { onMouseDown: this.toggle },
                React.createElement(MenuTitle, null, data[selectedIndex].content),
                React.createElement(MenuCaret, null,
                    React.createElement(Icon_1.Icon, { name: isOpen ? 'KeyboardArrowUp' : 'KeyboardArrowDown', size: "24px", color: colors_1.cyan }))),
            React.createElement(InteractiveList_1.InteractiveList, { data: interactiveList, open: isOpen, onBlur: this.hide, autoFocus: true, onChange: this.onChange, customWrapper: CustomWrapper, onClickOutside: function () { } })));
    };
    SlideDownTabsInt.inner = {
        get StyledWrapper() { return StyledWrapper; },
        get SelectedMenu() { return SelectedMenu; },
        get MenuTitle() { return MenuTitle; },
        get MenuCaret() { return MenuCaret; },
        get Icon() { return Icon_1.Icon; },
        get InteractiveList() { return InteractiveList_1.InteractiveList; }
    };
    return SlideDownTabsInt;
}(React.Component));
exports.SlideDownTabs = react_onclickoutside_1.default(SlideDownTabsInt);
exports.DropdownContainer = styled_1.default.div(templateObject_5 || (templateObject_5 = __makeTemplateObject([""], [""])));
exports.TabContent = styled_1.default.div(templateObject_6 || (templateObject_6 = __makeTemplateObject([""], [""])));
exports.TabItem = styled_1.default.div(templateObject_7 || (templateObject_7 = __makeTemplateObject(["\n  ", ";\n"], ["\n  ", ";\n"])), function (props) { return (props.active ? '' : 'display: none;'); });
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7;
