// precise-ui/dist/es5/components/AccordionCard/AccordionCardInt.part.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var styled_1 = require("../../utils/styled");
var distance_1 = require("../../distance");
var labels_1 = require("../../utils/labels");
var textStyles_1 = require("../../textStyles");
labels_1.setLabels({
    openLabel: 'Open Details',
    closeLabel: 'Close Details',
});
var animationDuration = '0.3s';
var animationFunction = 'cubic-bezier(0, 0, 0.25, 1)';
var StyledContainer = styled_1.default('div')(styled_1.themed(function (_a) {
    var open = _a.open, _b = _a.theme, _c = _b.accordionCard, borderColor = _c.borderColor, openedBorderColor = _c.openedBorderColor, text1 = _b.text1;
    return "\n    " + textStyles_1.getFontStyle({ size: 'medium' }) + "\n    \n    position: relative;\n    color: " + text1 + ";\n    border: 1px solid " + (open ? openedBorderColor : borderColor) + ";\n  ";
}));
var StyledHeaderContainer = styled_1.default('div')(styled_1.themed(function (_a) {
    var open = _a.open, _b = _a.theme.accordionCard, headerPadding = _b.headerPadding, openedHeaderBorderColor = _b.openedHeaderBorderColor, openedHeaderBackground = _b.openedHeaderBackground, headerBackground = _b.headerBackground;
    return "\n      padding: " + headerPadding + ";\n      background: " + (open ? openedHeaderBackground : headerBackground) + ";\n      border-bottom: " + (open ? "1px solid " + openedHeaderBorderColor : 'none') + ";\n    ";
}));
exports.StyledActionContainer = styled_1.default.div(styled_1.themed(function (_a) {
    var open = _a.open, _b = _a.theme, ui0 = _b.ui0, ui1 = _b.ui1;
    return styled_1.css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n      text-align: center;\n      padding: ", " ", ";\n      cursor: pointer;\n      background: ", ";\n      border: ", ";\n      margin: -1px -1px -1px -1px;\n      color: ", ";\n    "], ["\n      text-align: center;\n      padding: ", " ", ";\n      cursor: pointer;\n      background: ", ";\n      border: ", ";\n      margin: -1px -1px -1px -1px;\n      color: ", ";\n    "])), distance_1.distance.small, distance_1.distance.medium, open ? ui1 : ui0, open ? "1px solid " + ui0 : 'none', open ? ui0 : ui1);
}));
var StyledDetailsContainerAnimator = styled_1.default('div')(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  transition: all ", " ", ";\n  max-height: ", ";\n  opacity: ", ";\n  overflow: hidden;\n"], ["\n  transition: all ", " ", ";\n  max-height: ", ";\n  opacity: ", ";\n  overflow: hidden;\n"])), animationDuration, animationFunction, function (_a) {
    var open = _a.open;
    return (open ? '10000px' : '0');
}, function (_a) {
    var open = _a.open;
    return (open ? '1' : '0');
});
var AccordionCard = /** @class */ (function (_super) {
    __extends(AccordionCard, _super);
    function AccordionCard(props) {
        var _this = _super.call(this, props) || this;
        _this.handleClick = function () {
            var onActionClick = _this.props.onActionClick;
            var _a = _this.state, opened = _a.opened, controlled = _a.controlled;
            !controlled && _this.setState({ opened: !opened });
            if (typeof onActionClick === 'function') {
                onActionClick({ opened: controlled ? opened : !opened });
            }
        };
        _this.renderDefaultAction = function (_) {
            var theme = _this.props.theme;
            var opened = _this.state.opened;
            return (React.createElement(exports.StyledActionContainer, { theme: theme, onClick: _this.handleClick, open: opened }, labels_1.getPropLabel(_this.props, opened ? 'closeLabel' : 'openLabel')));
        };
        _this.state = {
            opened: props.opened !== undefined ? props.opened : false,
            controlled: props.opened !== undefined,
        };
        return _this;
    }
    AccordionCard.getDerivedStateFromProps = function (props, state) {
        return state.controlled ? { opened: props.opened } : {};
    };
    AccordionCard.prototype.renderActions = function () {
        var _a = this.props.renderAction, renderAction = _a === void 0 ? this.renderDefaultAction : _a;
        var opened = this.state.opened;
        return renderAction({ opened: opened });
    };
    AccordionCard.prototype.render = function () {
        var _a = this.props, _0 = _a.onActionClick, theme = _a.theme, children = _a.children, header = _a.header, props = __rest(_a, ["onActionClick", "theme", "children", "header"]);
        var opened = this.state.opened;
        return (React.createElement(StyledContainer, __assign({}, props, { theme: theme, open: opened }),
            React.createElement(StyledHeaderContainer, { theme: theme, open: opened }, header),
            React.createElement(StyledDetailsContainerAnimator, { theme: theme, open: opened }, children),
            this.renderActions()));
    };
    AccordionCard.inner = {
        get StyledActionContainer() { return exports.StyledActionContainer; },
        get StyledContainer() { return StyledContainer; },
        get StyledHeaderContainer() { return StyledHeaderContainer; },
        get StyledDetailsContainerAnimator() { return StyledDetailsContainerAnimator; }
    };
    return AccordionCard;
}(React.Component));
exports.AccordionCard = AccordionCard;
var templateObject_1, templateObject_2;
