// precise-ui/dist/es5/components/Flyout/index.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var react_onclickoutside_1 = require("react-onclickoutside");
var styled_1 = require("../../utils/styled");
var FlyoutWindow_part_1 = require("./FlyoutWindow.part");
var FlyoutContainer = styled_1.default.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  position: relative;\n  display: inline-block;\n  width: fit-content;\n"], ["\n  position: relative;\n  display: inline-block;\n  width: fit-content;\n"])));
var StyledTargetWrapper = styled_1.default.div(templateObject_2 || (templateObject_2 = __makeTemplateObject([""], [""])));
var FlyoutInt = /** @class */ (function (_super) {
    __extends(FlyoutInt, _super);
    function FlyoutInt(props) {
        var _this = _super.call(this, props) || this;
        _this.handleClickOutside = function () {
            _this.state.open && !_this.state.controlled && _this.setOpen(false);
        };
        _this.setTargetRef = function (el) {
            _this.targetContainer = el;
        };
        _this.onClick = function () {
            if (!_this.state.controlled) {
                _this.setOpen(!_this.state.open);
            }
        };
        _this.state = {
            controlled: props.open !== undefined,
            open: props.open || false,
            targetRect: undefined,
            dirtyFlag: false,
        };
        return _this;
    }
    FlyoutInt.prototype.componentDidMount = function () {
        this.updateMeasurements();
    };
    FlyoutInt.prototype.UNSAFE_componentWillReceiveProps = function (nextProps) {
        if (this.state.controlled && nextProps.open !== undefined) {
            this.setOpen(nextProps.open);
        }
    };
    FlyoutInt.prototype.componentDidUpdate = function () {
        var dirtyFlag = this.state.dirtyFlag;
        if (dirtyFlag) {
            this.updateMeasurements();
            this.setState({
                dirtyFlag: false,
            });
        }
    };
    FlyoutInt.prototype.updateMeasurements = function () {
        if (this.targetContainer) {
            var targetRect = this.targetContainer.getBoundingClientRect();
            this.setState({
                targetRect: targetRect,
            });
        }
    };
    FlyoutInt.prototype.setOpen = function (open) {
        var onChange = this.props.onChange;
        if (typeof onChange === 'function') {
            onChange({ open: open });
        }
        this.setState({
            open: open,
            dirtyFlag: open === true,
        });
    };
    FlyoutInt.prototype.render = function () {
        var _a = this.props, children = _a.children, content = _a.content, _0 = _a.open, _1 = _a.onChange, props = __rest(_a, ["children", "content", "open", "onChange"]);
        var targetRect = this.state.targetRect;
        var open = this.state.open;
        return (React.createElement(FlyoutContainer, { ref: this.setTargetRef },
            React.createElement(StyledTargetWrapper, { onClick: this.onClick }, children),
            !!content && targetRect && open && (React.createElement(FlyoutWindow_part_1.FlyoutWindow, __assign({}, props, { targetRect: targetRect }), content))));
    };
    FlyoutInt.inner = {
        get FlyoutContainer() { return FlyoutContainer; },
        get StyledTargetWrapper() { return StyledTargetWrapper; },
        get FlyoutWindow() { return FlyoutWindow_part_1.FlyoutWindow; }
    };
    return FlyoutInt;
}(React.Component));
exports.Flyout = react_onclickoutside_1.default(FlyoutInt);
exports.Flyout.displayName = 'Flyout';
var templateObject_1, templateObject_2;
