// precise-ui/dist/es5/components/InteractiveSurface/index.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var styled_1 = require("../../utils/styled");
var interactivity_1 = require("../../utils/interactivity");
var Container = styled_1.default.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  position: relative;\n  touch-action: ", ";\n  user-select: ", ";\n"], ["\n  position: relative;\n  touch-action: ", ";\n  user-select: ", ";\n"])), function (props) { return (props.disabled ? 'auto' : 'none'); }, function (props) { return (props.disabled ? 'auto' : 'none'); });
/**
 * The interactive surface provides a component that reacts on touch (or click) interactivity.
 */
var InteractiveSurface = /** @class */ (function (_super) {
    __extends(InteractiveSurface, _super);
    function InteractiveSurface() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.mounted = false;
        _this.moved = false;
        _this.updateValue = function (x, y, active, rect, moved) {
            var canceled = true;
            if (_this.mounted) {
                var onChange = _this.props.onChange;
                _this.moved = _this.moved || moved;
                if (typeof onChange === 'function') {
                    onChange({
                        x: x,
                        y: y,
                        moved: _this.moved,
                        active: active,
                        rect: rect,
                        release: function () {
                            canceled = false;
                        },
                    });
                }
            }
            return canceled;
        };
        _this.changeValue = function (e) {
            var disabled = _this.props.disabled;
            _this.moved = false;
            if (!disabled) {
                interactivity_1.trackMouse(e, _this.updateValue);
            }
        };
        _this.preventValue = function (e) {
            var _a = _this.props, disabled = _a.disabled, opaque = _a.opaque;
            if (!disabled && !opaque && _this.moved) {
                e.preventDefault();
                e.stopPropagation();
            }
        };
        _this.changeTouchValue = function (e) {
            var disabled = _this.props.disabled;
            _this.moved = false;
            if (!disabled) {
                interactivity_1.trackTouch(e, _this.updateValue);
            }
        };
        return _this;
    }
    InteractiveSurface.prototype.UNSAFE_componentWillMount = function () {
        this.mounted = true;
    };
    InteractiveSurface.prototype.componentWillUnmount = function () {
        this.mounted = false;
    };
    InteractiveSurface.prototype.render = function () {
        var _a = this.props, children = _a.children, disabled = _a.disabled, _0 = _a.onChange, _1 = _a.opaque, props = __rest(_a, ["children", "disabled", "onChange", "opaque"]);
        return (React.createElement(Container, __assign({}, props, { onMouseDownCapture: this.changeValue, onClickCapture: this.preventValue, onTouchStartCapture: this.changeTouchValue }), children));
    };
    InteractiveSurface.inner = {
        get Container() { return Container; }
    };
    return InteractiveSurface;
}(React.Component));
exports.InteractiveSurface = InteractiveSurface;
var templateObject_1;
