// precise-ui/dist/es6/components/Sidebar/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import { SidebarPopup } from './SidebarPopup.part';
import { SidebarActivatorContainer } from './SidebarActivatorContainer.part';
export class Sidebar extends React.Component {
    constructor(props) {
        super(props);
        this.show = () => {
            if (!this.controlled) {
                this.setState({ open: true });
            }
            typeof this.props.onShow === 'function' && this.props.onShow();
        };
        this.hide = () => {
            if (!this.controlled) {
                this.setState({ open: false });
            }
            typeof this.props.onClose === 'function' && this.props.onClose();
        };
        this.onKeyPress = (e) => {
            if (e.keyCode === 27 /* escape */) {
                this.hide();
            }
        };
        const { open } = this.props;
        this.controlled = open !== undefined;
        this.state = {
            open: open !== undefined ? open : false,
        };
    }
    UNSAFE_componentWillReceiveProps(props) {
        if (this.controlled) {
            this.setState({ open: Boolean(props.open) });
        }
    }
    componentDidMount() {
        document.addEventListener('keydown', this.onKeyPress);
    }
    componentWillUnmount() {
        document.removeEventListener('keydown', this.onKeyPress);
    }
    render() {
        const _a = this.props, { onShow: _0, activator, size = '275px' } = _a, props = __rest(_a, ["onShow", "activator", "size"]);
        const { open } = this.state;
        return (React.createElement(React.Fragment, null,
            React.createElement(SidebarActivatorContainer, { onClick: this.show }, activator),
            React.createElement(SidebarPopup, Object.assign({}, props, { open: open, onClose: this.hide, size: size }))));
    }
}
Sidebar.inner = {
    get SidebarActivatorContainer() { return SidebarActivatorContainer; },
    get SidebarPopup() { return SidebarPopup; }
};
