// precise-ui/dist/es5/components/Tooltip/index.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var styled_1 = require("../../utils/styled");
var Flyout_1 = require("../Flyout");
var colors_1 = require("../../colors");
var remCalc_1 = require("../../utils/remCalc");
var TooltipContainer = styled_1.default.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  position: relative;\n  width: fit-content;\n  display: inline-block;\n"], ["\n  position: relative;\n  width: fit-content;\n  display: inline-block;\n"])));
var Tooltip = /** @class */ (function (_super) {
    __extends(Tooltip, _super);
    function Tooltip(props) {
        var _this = _super.call(this, props) || this;
        _this.setTargetRef = function (el) {
            _this.targetContainer = el;
        };
        _this.onMouseOver = function () {
            if (!_this.state.controlled) {
                _this.setOpen(true);
            }
        };
        _this.onMouseOut = function () {
            if (!_this.state.controlled) {
                _this.setOpen(false);
            }
        };
        _this.state = {
            controlled: props.open !== undefined,
            open: props.open || false,
            targetRect: { top: 0, right: 0, bottom: 0, left: 0, width: 0, height: 0 },
            dirtyFlag: false,
        };
        return _this;
    }
    Tooltip.prototype.componentDidMount = function () {
        this.updateMeasurements();
    };
    Tooltip.prototype.UNSAFE_componentWillReceiveProps = function (nextProps) {
        if (this.state.controlled && nextProps.open !== undefined) {
            this.setOpen(nextProps.open);
        }
    };
    Tooltip.prototype.componentDidUpdate = function () {
        var dirtyFlag = this.state.dirtyFlag;
        if (dirtyFlag) {
            this.updateMeasurements();
            this.setState({
                dirtyFlag: false,
            });
        }
    };
    Tooltip.prototype.updateMeasurements = function () {
        if (this.targetContainer) {
            var targetRect = this.targetContainer.getBoundingClientRect();
            this.setState({
                targetRect: targetRect,
            });
        }
    };
    Tooltip.prototype.setOpen = function (open) {
        var onChange = this.props.onChange;
        if (typeof onChange === 'function') {
            onChange({ open: open });
        }
        this.setState({
            open: open,
            dirtyFlag: open === true,
        });
    };
    Tooltip.prototype.render = function () {
        var _a = this.props, _0 = _a.open, theme = _a.theme, props = __rest(_a, ["open", "theme"]);
        var open = this.state.open;
        var tooltipFlyoutTheme = __assign({ flyout: { background: colors_1.tuna, textColor: colors_1.white, fontSize: remCalc_1.remCalc('14px'), maxWidth: '250px' } }, theme);
        return (React.createElement(TooltipContainer, { ref: this.setTargetRef, onFocus: this.onMouseOver, onBlur: this.onMouseOut, onMouseOver: this.onMouseOver, onMouseOut: this.onMouseOut },
            React.createElement(Flyout_1.Flyout, __assign({}, props, { open: open, theme: tooltipFlyoutTheme }))));
    };
    Tooltip.inner = {
        get TooltipContainer() { return TooltipContainer; },
        get Flyout() { return Flyout_1.Flyout; }
    };
    return Tooltip;
}(React.Component));
exports.Tooltip = Tooltip;
var templateObject_1;
