// precise-ui/dist/es6/components/ProgressBar/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled, { themed, keyframes } from '../../utils/styled';
import { distance } from '../../distance';
import { getFontStyle } from '../../textStyles';
const spinningWidth = 50;
const ProgressContainer = styled.div `
  color: ${themed(props => props.theme.text6)};
`;
const ProgressRail = styled('div') `
  position: relative;
  border: 0;
  background: ${themed(props => props.theme.ui4)};
  height: ${({ type }) => (type === 'secondary' ? distance.xsmall : distance.xxsmall)};
`;
const ProgressTitle = styled.div `
  ${getFontStyle({ size: 'medium' })}

  margin: 0 0 ${distance.medium};
  padding: 0;
  display: block;
`;
const ProgressDescription = styled.div `
  ${getFontStyle({ size: 'small' })}

  margin: ${distance.small} 0 0;
`;
const ProgressIndicator = styled.div `
  border-radius: inherit;
  border: 0;
  background: ${themed(props => props.theme.ui0)};
  height: 100%;
  width: 0;
`;
const ProgressStop = styled('div') `
  border-radius: ${distance.small};
  position: absolute;
  width: ${distance.small};
  height: ${distance.small};
  background: ${themed(props => (props.active ? props.theme.ui7 : props.theme.text2))};
  top: 50%;
  transform: translateY(-50%);
`;
const SpinningAnimation = keyframes `
  from {
    left: -${spinningWidth}%;
  }
  to {
    left: 100%;
  }
`;
const MaskContainer = styled.div `
  position: relative;
  overflow: hidden;
  height: 100%;
`;
const ProgressAnimation = styled.div `
  position: absolute;
  height: 100%;
  width: 100%;
  left: -${spinningWidth}%;
  animation: ${SpinningAnimation} 2s linear infinite;
`;
function getPercent(current, minimum, maximum) {
    const value = ~~((100 * (current - minimum)) / (maximum - minimum));
    const percent = Math.min(Math.max(0, value), 100);
    return `${percent}%`;
}
function spinning(child) {
    return (React.createElement(MaskContainer, null,
        React.createElement(ProgressAnimation, null, child)));
}
/**
 * The progress bar component displays a progress indicator ranging from min to max.
 */
export const ProgressBar = Object.assign(((_a) => {
    var { minimum = 0, maximum = 100, value: propValue, title, description, animate, stops = [], type = 'primary', theme } = _a, rest = __rest(_a, ["minimum", "maximum", "value", "title", "description", "animate", "stops", "type", "theme"]);
    const value = (animate === 'spinning' ? spinningWidth : propValue) || 0;
    const width = getPercent(value, minimum, maximum);
    const transition = animate ? 'width 200ms' : 'none';
    const indicator = React.createElement(ProgressIndicator, { style: { width, transition }, theme: theme });
    return (React.createElement(ProgressContainer, { theme: theme },
        !!title && React.createElement(ProgressTitle, { theme: theme }, title),
        React.createElement(ProgressRail, Object.assign({ type: type, theme: theme }, rest),
            animate === 'spinning' ? spinning(indicator) : indicator,
            stops.map((stop, index) => (React.createElement(ProgressStop, { key: `${index}-at-${stop}`, style: { left: getPercent(stop, minimum, maximum) }, active: stop <= value, theme: theme })))),
        !!description && React.createElement(ProgressDescription, { theme: theme }, description)));
}), { inner: {
        get ProgressIndicator() { return ProgressIndicator; },
        get ProgressContainer() { return ProgressContainer; },
        get ProgressTitle() { return ProgressTitle; },
        get ProgressRail() { return ProgressRail; },
        get ProgressStop() { return ProgressStop; },
        get ProgressDescription() { return ProgressDescription; }
    } });
ProgressBar.displayName = 'ProgressBar';
