// precise-ui/dist/es6/components/FileUploaderDetails/helpers.js
import { setLabels } from '../../utils/labels';
export function mergeData(newData, currentData) {
    const all = {};
    for (const file of currentData) {
        all[file.fileId] = file;
    }
    for (const file of newData) {
        if (file.deleted) {
            delete all[file.fileId];
        }
        else if (!file.canceled || all[file.fileId]) {
            all[file.fileId] = file;
        }
    }
    return Object.keys(all).map(key => all[key]);
}
export const iconNames = {
    error: 'Error',
    canceled: 'Error',
    progress: 'Sync',
    scanning: 'Sync',
    success: 'CheckCircle',
};
setLabels({
    tableHeaderFileLabel: 'File',
    tableHeaderStatusLabel: 'Status',
    errorTableUploadLabel: 'Upload failed',
    canceledTableUploadLabel: 'Upload canceled',
    progressTableUploadLabel: 'Uploading...',
    successTableUploadLabel: 'Upload successful',
    scanningTableUploadLabel: 'Scanning for viruses...',
    uploadModalTitleLabel: 'Upload Details',
    uploadProgressLabel: 'Uploading...',
    uploadSuccessLabel: 'Upload successful',
    uploadErrorLabel: 'Upload not successful',
    uploadScanningLabel: 'Scanning for viruses',
    viewDetailsLabel: 'View Details',
    itemPluralLabel: 'Files',
    itemSingularLabel: 'File',
    cancelAllLabel: 'Cancel all uploads',
});
export function getStatus(file) {
    if (file.error) {
        return 'error';
    }
    else if (file.canceled || file.deleted) {
        return 'canceled';
    }
    else if (file.progress < 100) {
        return 'progress';
    }
    else if (file.scanning) {
        return 'scanning';
    }
    return 'success';
}
