// precise-ui/dist/es5/components/Sidebar/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var SidebarPopup_part_1 = require("./SidebarPopup.part");
var SidebarActivatorContainer_part_1 = require("./SidebarActivatorContainer.part");
var Sidebar = /** @class */ (function (_super) {
    __extends(Sidebar, _super);
    function Sidebar(props) {
        var _this = _super.call(this, props) || this;
        _this.show = function () {
            if (!_this.controlled) {
                _this.setState({ open: true });
            }
            typeof _this.props.onShow === 'function' && _this.props.onShow();
        };
        _this.hide = function () {
            if (!_this.controlled) {
                _this.setState({ open: false });
            }
            typeof _this.props.onClose === 'function' && _this.props.onClose();
        };
        _this.onKeyPress = function (e) {
            if (e.keyCode === 27 /* escape */) {
                _this.hide();
            }
        };
        var open = _this.props.open;
        _this.controlled = open !== undefined;
        _this.state = {
            open: open !== undefined ? open : false,
        };
        return _this;
    }
    Sidebar.prototype.UNSAFE_componentWillReceiveProps = function (props) {
        if (this.controlled) {
            this.setState({ open: Boolean(props.open) });
        }
    };
    Sidebar.prototype.componentDidMount = function () {
        document.addEventListener('keydown', this.onKeyPress);
    };
    Sidebar.prototype.componentWillUnmount = function () {
        document.removeEventListener('keydown', this.onKeyPress);
    };
    Sidebar.prototype.render = function () {
        var _a = this.props, _0 = _a.onShow, activator = _a.activator, _b = _a.size, size = _b === void 0 ? '275px' : _b, props = __rest(_a, ["onShow", "activator", "size"]);
        var open = this.state.open;
        return (React.createElement(React.Fragment, null,
            React.createElement(SidebarActivatorContainer_part_1.SidebarActivatorContainer, { onClick: this.show }, activator),
            React.createElement(SidebarPopup_part_1.SidebarPopup, __assign({}, props, { open: open, onClose: this.hide, size: size }))));
    };
    Sidebar.inner = {
        get SidebarActivatorContainer() { return SidebarActivatorContainer_part_1.SidebarActivatorContainer; },
        get SidebarPopup() { return SidebarPopup_part_1.SidebarPopup; }
    };
    return Sidebar;
}(React.Component));
exports.Sidebar = Sidebar;
