// precise-ui/dist/es5/components/Sidebar/SidebarContainer.part.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var styled_1 = require("../../utils/styled");
exports.SidebarContainer = styled_1.default.div(function (_a) {
    var open = _a.open, _b = _a.position, position = _b === void 0 ? 'right' : _b, _c = _a.size, size = _c === void 0 ? '275px' : _c;
    return "\n    transition: all 0.3s cubic-bezier(0, 0, 0.25, 1);\n    user-select: none;\n    position: fixed;\n    background-color: white;\n    z-index: 100001;\n    " + (!open
        ? "\n        visibility: hidden;\n        opacity: 0;\n        overflow: initial;\n        "
        : "\n        visibility: visible;\n        opacity: 1;\n        overflow: auto;\n        ") + "\n    " + getSizeStyles(position, size) + "\n    " + getPositionStyles(position, open, size) + "\n  ";
});
function getSizeStyles(position, size) {
    if (position === 'top' || position === 'bottom') {
        return "\n      height: " + size + "; \n      width: 100%;\n    ";
    }
    else {
        return "\n      width: " + size + "; \n      height: 100%;\n    ";
    }
}
function getPositionStyles(position, open, size) {
    var opposite = {
        top: 'bottom',
        bottom: 'top',
        left: 'right',
        right: 'left',
    };
    var positionValue = {
        top: '0',
        bottom: '0',
        left: '0',
        right: '0',
    };
    positionValue[opposite[position]] = 'auto';
    if (!open) {
        var match = size.match(/^(\d+)(.+?)$/);
        if (match) {
            positionValue[position] = "-" + parseInt(match[1], undefined) / 2 + match[2];
        }
        else {
            positionValue[position] = "-50%";
        }
    }
    return (Object.keys(positionValue)
        .map(function (positionName) { return positionName + ": " + positionValue[positionName]; })
        .join(';') + ';');
}
