// precise-ui/dist/es5/components/GridArea/index.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var styled_1 = require("../../utils/styled");
var GridAreaContainer = styled_1.default.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  height: 100%;\n"], ["\n  height: 100%;\n"])));
/**
 * The GridArea component to declare areas within a Grid instance.
 */
var GridArea = /** @class */ (function (_super) {
    __extends(GridArea, _super);
    function GridArea() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.clicked = function (e) {
            _this.tapped(e.currentTarget, e.clientX, e.clientY);
        };
        _this.touched = function (e) {
            var touch = e.touches[0];
            _this.tapped(e.currentTarget, touch.clientX, touch.clientY);
        };
        return _this;
    }
    GridArea.prototype.tapped = function (node, clientX, clientY) {
        var onTap = this.props.onTap;
        if (typeof onTap === 'function') {
            var r = node.getBoundingClientRect();
            onTap({
                node: node,
                x: clientX - r.left,
                y: clientY - r.top,
                width: r.width,
                height: r.height,
            });
        }
    };
    GridArea.prototype.render = function () {
        var _a = this.props, colSpan = _a.colSpan, rowSpan = _a.rowSpan, column = _a.column, row = _a.row, onTap = _a.onTap, hidden = _a.hidden, props = __rest(_a, ["colSpan", "rowSpan", "column", "row", "onTap", "hidden"]);
        return React.createElement(GridAreaContainer, __assign({}, props, { onMouseDown: this.clicked, onTouchStartCapture: this.touched }));
    };
    GridArea.inner = {
        get GridAreaContainer() { return GridAreaContainer; }
    };
    return GridArea;
}(React.PureComponent));
exports.GridArea = GridArea;
var templateObject_1;
