// precise-ui/dist/es6/components/Image/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import { Spinner } from '../Spinner';
import styled, { themed } from '../../utils/styled';
// Wrapper used only in case when preloader enabled
const StyledImageWrapper = styled.div `
  position: relative;
  display: inline-block;
  width: 100%;
  height: 100%;
  flex: 1 1 auto;
  box-sizing: border-box;
`;
const StyledImage = styled.img `
  margin: 0;
  padding: 0;
  border: 0;
  display: inline-block;
  vertical-align: middle;
  max-width: 100%;
  height: auto;
  border-style: none;
`;
const StyledError = styled.span `
  background: ${themed(props => props.theme.inputError)};
  color: ${themed(props => props.theme.text7)};
  font-family: inherit;
`;
/**
 * Image component is meant to be a simple wrapper for standard DOM `img` element.
 * Component is fluid by default, meaning that it has max-width: 100% and height: auto,
 * so that it scales with it's parent.
 */
export class Image extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            status: 'idle',
        };
    }
    componentDidMount() {
        const { preload, src } = this.props;
        if (preload) {
            const image = document.createElement('img');
            image.onload = () => this.setState({ status: 'loaded' });
            image.onerror = () => this.setState({ status: 'error' });
            image.src = src;
        }
    }
    render() {
        const _a = this.props, { children, preload, src, alt, preloader, error } = _a, rest = __rest(_a, ["children", "preload", "src", "alt", "preloader", "error"]);
        const { status } = this.state;
        return preload ? (React.createElement(StyledImageWrapper, null, (() => {
            switch (status) {
                case 'loaded':
                    return React.createElement(StyledImage, Object.assign({ src: src, alt: alt }, rest));
                case 'error':
                    return error || React.createElement(StyledError, null,
                        "Error loading image from ",
                        src,
                        ".");
                case 'loading':
                case 'idle':
                default:
                    return preloader || React.createElement(Spinner, null);
            }
        })())) : (React.createElement(StyledImage, Object.assign({ src: src, alt: alt }, rest)));
    }
}
Image.inner = {
    get StyledImageWrapper() { return StyledImageWrapper; },
    get StyledImage() { return StyledImage; },
    get StyledError() { return StyledError; },
    get Spinner() { return Spinner; }
};
