// precise-ui/dist/es5/components/AccordionTable/AccordionTableCard.part.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var styled_1 = require("../../utils/styled");
var indexHelper_1 = require("./helpers/indexHelper");
var TableShared_part_1 = require("../Table/TableShared.part");
var TableCard_part_1 = require("../Table/TableCard.part");
var AccordionCard_1 = require("../AccordionCard");
var ListItem_1 = require("../ListItem");
var List_1 = require("../List");
var distance_1 = require("../../distance");
var textStyles_1 = require("../../textStyles");
var PropContainer = styled_1.default.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  margin-bottom: ", ";\n\n  &:last-child {\n    margin-bottom: 0;\n  }\n"], ["\n  margin-bottom: ", ";\n\n  &:last-child {\n    margin-bottom: 0;\n  }\n"])), distance_1.distance.medium);
var PropName = styled_1.default.div(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  ", "\n"], ["\n  ", "\n"])), textStyles_1.getFontStyle({ size: 'xSmall' }));
var PropValue = styled_1.default.div(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  ", "\n"], ["\n  ", "\n"])), textStyles_1.getFontStyle({ weight: 'medium' }));
var PlaceholderContainer = styled_1.default.div(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  text-align: center;\n"], ["\n  text-align: center;\n"])));
var StyledList = styled_1.default(List_1.List)(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  overflow: visible;\n"], ["\n  overflow: visible;\n"])));
var StyledListItem = styled_1.default(ListItem_1.ListItem)(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  padding: 0;\n  margin-bottom: ", ";\n  border: none;\n\n  &:last-child {\n    margin-bottom: 0;\n  }\n"], ["\n  padding: 0;\n  margin-bottom: ", ";\n  border: none;\n\n  &:last-child {\n    margin-bottom: 0;\n  }\n"])), distance_1.distance.xlarge);
var AccordionTableCardHost = Object.assign((function (props) { return React.createElement(StyledList, __assign({}, props)); }), { inner: {
        get StyledList() { return StyledList; }
    } });
var AccordionTableCard = /** @class */ (function (_super) {
    __extends(AccordionTableCard, _super);
    function AccordionTableCard(props) {
        var _this = _super.call(this, props) || this;
        _this.renderItem = function (_a) {
            var item = _a.item, index = _a.index, open = _a.open, keys = _a.keys, closeLabel = _a.closeLabel, openLabel = _a.openLabel;
            var _b = _this.props, detailsRenderer = _b.detailsRenderer, theme = _b.theme;
            return (React.createElement(StyledListItem, { theme: theme, key: index },
                React.createElement(AccordionCard_1.AccordionCard, { openLabel: openLabel, closeLabel: closeLabel, theme: theme, header: _this.renderItemProps(item, keys, index), opened: open, onActionClick: function () { return _this.handleClick(index, item); } }, open && detailsRenderer && detailsRenderer({ data: item, index: index }))));
        };
        _this.state = {
            selectedIndexes: indexHelper_1.normalizeIndex(props.selectedIndex),
            controlled: props.selectedIndex !== undefined,
        };
        return _this;
    }
    AccordionTableCard.getDerivedStateFromProps = function (props, state) {
        var selectedIndex = props.selectedIndex;
        if (state.controlled && selectedIndex !== undefined) {
            return {
                selectedIndexes: indexHelper_1.normalizeIndex(props.selectedIndex),
            };
        }
        return state;
    };
    AccordionTableCard.prototype.handleClick = function (target, data) {
        var _a = this.props, onChange = _a.onChange, multiple = _a.multiple;
        var _b = this.state, controlled = _b.controlled, selectedIndexes = _b.selectedIndexes;
        var nextIndexes = indexHelper_1.toggleIndex(selectedIndexes, target, multiple);
        if (typeof onChange === 'function') {
            onChange({
                selectedIndex: multiple ? nextIndexes : nextIndexes[0] !== undefined ? nextIndexes[0] : -1,
                previousIndex: multiple ? selectedIndexes : selectedIndexes[0] !== undefined ? selectedIndexes[0] : -1,
                data: data,
            });
        }
        if (!controlled) {
            this.setState({
                selectedIndexes: nextIndexes,
            });
        }
    };
    AccordionTableCard.prototype.getHeader = function (key) {
        var columns = this.props.columns;
        if (columns) {
            var column = columns[key];
            return typeof column === 'string' ? column : column.header;
        }
        return key;
    };
    AccordionTableCard.prototype.renderItemProps = function (item, keys, row) {
        var _this = this;
        return keys.map(function (key, index) { return _this.renderItemProp(key, item[key], index, row); });
    };
    AccordionTableCard.prototype.renderItemProp = function (key, propValue, index, row) {
        var _a = this.props.cellRenderer, cellRenderer = _a === void 0 ? TableShared_part_1.defaultCellRenderer : _a;
        var propKey = this.getHeader(key);
        return (React.createElement(PropContainer, { key: index },
            React.createElement(PropName, null, propKey),
            React.createElement(PropValue, null, cellRenderer({
                row: row,
                column: index,
                key: key,
                value: propValue,
                render: TableShared_part_1.defaultCellRenderer,
            }))));
    };
    AccordionTableCard.prototype.render = function () {
        var _this = this;
        var _a = this.props, _b = _a.data, data = _b === void 0 ? [] : _b, columns = _a.columns, placeholder = _a.placeholder, theme = _a.theme, _c = _a.cardRenderer, cardRenderer = _c === void 0 ? this.renderItem : _c, _d = _a.bodyRenderer, bodyRenderer = _d === void 0 ? TableCard_part_1.defaultBodyRenderer : _d, openLabel = _a.openLabel, closeLabel = _a.closeLabel, props = __rest(_a, ["data", "columns", "placeholder", "theme", "cardRenderer", "bodyRenderer", "openLabel", "closeLabel"]);
        var keys = Object.keys(columns || data[0] || {});
        var rows = data.length === 0
            ? placeholder
                ? [
                    React.createElement(StyledListItem, { theme: theme, key: 0 },
                        React.createElement(PlaceholderContainer, { theme: theme }, placeholder)),
                ]
                : []
            : data.map(function (item, index) { return cardRenderer({
                item: item,
                index: index,
                open: indexHelper_1.hasIndex(_this.state.selectedIndexes, index),
                keys: keys,
                openLabel: openLabel,
                closeLabel: closeLabel,
            }); });
        return bodyRenderer({
            rows: rows,
            mode: 'card',
            props: __assign({ theme: theme, borderless: true }, props),
            table: AccordionTableCardHost,
        });
    };
    AccordionTableCard.inner = {
        get StyledListItem() { return StyledListItem; },
        get AccordionCard() { return AccordionCard_1.AccordionCard; },
        get PropContainer() { return PropContainer; },
        get PropName() { return PropName; },
        get PropValue() { return PropValue; },
        get PlaceholderContainer() { return PlaceholderContainer; }
    };
    return AccordionTableCard;
}(React.Component));
exports.AccordionTableCard = AccordionTableCard;
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6;
