// precise-ui/dist/es5/components/FileUploaderDetails/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var labels_1 = require("../../utils/labels");
function mergeData(newData, currentData) {
    var all = {};
    for (var _i = 0, currentData_1 = currentData; _i < currentData_1.length; _i++) {
        var file = currentData_1[_i];
        all[file.fileId] = file;
    }
    for (var _a = 0, newData_1 = newData; _a < newData_1.length; _a++) {
        var file = newData_1[_a];
        if (file.deleted) {
            delete all[file.fileId];
        }
        else if (!file.canceled || all[file.fileId]) {
            all[file.fileId] = file;
        }
    }
    return Object.keys(all).map(function (key) { return all[key]; });
}
exports.mergeData = mergeData;
exports.iconNames = {
    error: 'Error',
    canceled: 'Error',
    progress: 'Sync',
    scanning: 'Sync',
    success: 'CheckCircle',
};
labels_1.setLabels({
    tableHeaderFileLabel: 'File',
    tableHeaderStatusLabel: 'Status',
    errorTableUploadLabel: 'Upload failed',
    canceledTableUploadLabel: 'Upload canceled',
    progressTableUploadLabel: 'Uploading...',
    successTableUploadLabel: 'Upload successful',
    scanningTableUploadLabel: 'Scanning for viruses...',
    uploadModalTitleLabel: 'Upload Details',
    uploadProgressLabel: 'Uploading...',
    uploadSuccessLabel: 'Upload successful',
    uploadErrorLabel: 'Upload not successful',
    uploadScanningLabel: 'Scanning for viruses',
    viewDetailsLabel: 'View Details',
    itemPluralLabel: 'Files',
    itemSingularLabel: 'File',
    cancelAllLabel: 'Cancel all uploads',
});
function getStatus(file) {
    if (file.error) {
        return 'error';
    }
    else if (file.canceled || file.deleted) {
        return 'canceled';
    }
    else if (file.progress < 100) {
        return 'progress';
    }
    else if (file.scanning) {
        return 'scanning';
    }
    return 'success';
}
exports.getStatus = getStatus;
