// precise-ui/dist/es6/components/PaginationBar/ItemControlsView.part.js
import * as React from 'react';
import styled from '../../utils/styled';
import { SelectButton } from '../SelectButton';
import { dark } from '../../colors';
import { distance } from '../../distance';
const StyledItemControls = styled.div `
  margin-right: auto;
  display: flex;
  align-items: center;
`;
const StyledItemsInfo = styled.div `
  position: relative;
  padding-left: ${distance.small};

  &:before {
    content: '';
    background: ${dark};
    position: absolute;
    left: 0;
    top: 50%;
    transform: translateY(-50%);
    width: 1px;
    height: 100%;
  }
`;
export const ItemControls = Object.assign((({ label, data, value, children, onChange }) => (React.createElement(StyledItemControls, null,
    label,
    React.createElement(SelectButton, { data: data, value: value, onChange: onChange }),
    React.createElement(StyledItemsInfo, null, children)))), { inner: {
        get StyledItemControls() { return StyledItemControls; },
        get SelectButton() { return SelectButton; },
        get StyledItemsInfo() { return StyledItemsInfo; }
    } });
