// precise-ui/dist/es6/components/PaginationBar/PageControls.part.js
import * as React from 'react';
import styled from '../../utils/styled';
import { Container } from '../Container';
import { Icon } from '../Icon';
import { SelectButton } from '../SelectButton';
import { cyan, grey1 } from '../../colors';
import { distance } from '../../distance';
import { getFontStyle } from '../../textStyles';
const StyledContainer = styled(Container) `
  width: auto;
`;
const PageControlsStyled = styled.div `
  display: flex;
  align-items: center;
  margin-left: auto;
`;
const ChangePage = styled.button `
  ${getFontStyle({ size: 'xLarge' })}
  position: relative;
  cursor: ${props => (props.disabled ? 'not-allowed' : 'pointer')};
  outline: none;
  display: flex;
  background: transparent;
  border: none;
  align-items: center;
  padding: ${distance.small};
  margin: 0 ${distance.small};
  color: ${props => (!props.disabled ? cyan : grey1)};
  &:before,
  &:after {
    content: '';
    position: absolute;
    width: 1px;
    background: #dfe3e6;
    left: 0;
    top: 50%;
    transform: translateY(-50%);
    height: 100%;
  }
  &:after {
    right: 0;
    left: auto;
  }
  &:last-child {
    margin-right: 0;
    &:after {
      display: none;
    }
  }
`;
export const PageControls = Object.assign((({ children, current, total, changeToPrevious, changeToNext, changeToSelect, pages, }) => (React.createElement(PageControlsStyled, null,
    React.createElement(StyledContainer, null, children),
    React.createElement(ChangePage, { disabled: current <= 0, onClick: changeToPrevious, type: "button" },
        React.createElement(Icon, { name: "KeyboardArrowLeft" })),
    React.createElement(SelectButton, { data: pages, value: `${current + 1}`, onChange: changeToSelect }),
    React.createElement(ChangePage, { disabled: current >= total - 1, onClick: changeToNext, type: "button" },
        React.createElement(Icon, { name: "KeyboardArrowRight" }))))), { inner: {
        get PageControlsStyled() { return PageControlsStyled; },
        get StyledContainer() { return StyledContainer; },
        get ChangePage() { return ChangePage; },
        get Icon() { return Icon; },
        get SelectButton() { return SelectButton; }
    } });
