// precise-ui/dist/es6/components/Headline/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled, { themed, css } from '../../utils/styled';
import { distance } from '../../distance';
import { getFontStyle } from '../../textStyles';
const Headlines = {};
function getComponentName(level) {
    return `h${level >= 1 && level <= 5 ? level : 3}`;
}
function getHeadlineStyle(level) {
    switch (level) {
        case 1:
            return getFontStyle({ size: 'xxxLarge', weight: 'light' });
        case 2:
            return getFontStyle({ size: 'xxLarge', weight: 'light' });
        case 3:
            return getFontStyle({ size: 'xLarge', weight: 'medium' });
        case 4:
            return getFontStyle({ size: 'large', weight: 'regular' });
        case 5:
            return getFontStyle({ size: 'medium', weight: 'medium' });
        default:
            return '';
    }
}
function getStyledHeadline(level) {
    const component = getComponentName(level);
    const Headline = Headlines[component];
    if (!Headline) {
        const NewHeadline = styled(component)(themed(props => css `
          ${getHeadlineStyle(props.level)}

          margin: 0;
          padding: ${props.theme.headingsPadding || `0 ${distance.small} 0 0`};
          font-family: ${props.theme.fontFamily || 'inherit'};
          color: ${props.subheader ? props.theme.text5 : 'inherit'};
        `));
        Headlines[component] = NewHeadline;
        return NewHeadline;
    }
    return Headline;
}
/**
 * Headline component with styles for all headline levels.
 */
export const Headline = (_a) => {
    var { level = 3, children } = _a, rest = __rest(_a, ["level", "children"]);
    const StyledHeadline = getStyledHeadline(level);
    return (React.createElement(StyledHeadline, Object.assign({ level: level }, rest), children));
};
Headline.displayName = 'Headline';
