// precise-ui/dist/es6/utils/displayTo.js
import { css } from './styled';
import { ScreenSizeList } from '../common';
import { breakpoints } from '../themes';
export function getMediaQueries(breakpoints) {
    return ScreenSizeList.map(x => ({ screen: x, breakpoints: getScreenSizeBreakpoints(x, breakpoints) })).reduce((acc, next) => (next.breakpoints ? Object.assign({}, acc, { [next.screen]: getWidthBreakpointsQuery(next.breakpoints) }) : acc), {});
}
// create max-width media query including screen size
export function displayUpTo(screen) {
    const screenBreakpoints = getScreenSizeBreakpoints(screen, breakpoints);
    if (!screenBreakpoints) {
        throw new Error('Invalid screen size');
    }
    const query = getWidthBreakpointsQuery({ max: screenBreakpoints.max });
    return getMediaQuery(query);
}
// create min-width media query including screen size
export function displayFrom(screen) {
    const screenBreakpoints = getScreenSizeBreakpoints(screen, breakpoints);
    if (!screenBreakpoints) {
        throw new Error('Invalid screen size');
    }
    const query = getWidthBreakpointsQuery({ min: screenBreakpoints.min });
    return getMediaQuery(query);
}
export function displayTo(screen) {
    const screenBreakpoints = getScreenSizeBreakpoints(screen, breakpoints);
    const query = screenBreakpoints ? getWidthBreakpointsQuery(screenBreakpoints) : screen;
    return getMediaQuery(query);
}
// create media query based on previous and next breakpoints
export function getWidthBreakpointsQuery({ max, min }) {
    if (min === undefined && max === undefined) {
        throw new Error('Both breakpoints cannot be `undefined`');
    }
    if (min !== undefined && max != undefined && max < min) {
        throw new Error(`The min(${min}) breakpoint must be less than max(${max})`);
    }
    const queries = [];
    if (min !== undefined) {
        queries.push(`(min-width: ${min}px)`);
    }
    if (max !== undefined) {
        queries.push(`(max-width: ${max - 1}px)`);
    }
    return `${queries.join(' and ')}`;
}
export function getScreenSizeBreakpoints(screen, breakpoints) {
    switch (screen) {
        case 'small':
            return { max: breakpoints.medium };
        case 'medium':
            return { min: breakpoints.medium, max: breakpoints.large };
        case 'large':
            return { min: breakpoints.large, max: breakpoints.xLarge };
        case 'xLarge':
            return { min: breakpoints.xLarge, max: breakpoints.max };
        case 'max':
            return { min: breakpoints.max };
        default:
            return undefined;
    }
}
function getMediaQuery(query) {
    return (strings, ...interpolations) => css `
      @media ${query} {
        ${css(strings, ...interpolations)};
      }
    `;
}
