// precise-ui/dist/es6/utils/styled.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as styledComponents from 'styled-components';
import { light } from './../themes';
const { default: styled, css, keyframes, createGlobalStyle, withTheme, ThemeProvider, ThemeConsumer, ThemeContext, isStyledComponent, ServerStyleSheet, StyleSheetManager, } = styledComponents;
/*
  Helper function which insures that theme is always available in the interpolation callback.
*/
export const themed = (interpolation) => (props) => {
    const _a = props, { theme } = _a, rest = __rest(_a, ["theme"]);
    const hasNestedProps = (prop) => typeof prop === 'object' && !Array.isArray(prop) && Object.keys(prop).length > 0;
    const mergedTheme = Object.keys(theme).reduce((acc, prop) => (hasNestedProps(theme[prop]) ? (acc[prop] = Object.assign({}, acc[prop], theme[prop])) : (acc[prop] = theme[prop]), acc), Object.assign({}, light));
    return interpolation(Object.assign({ theme: mergedTheme }, rest));
};
export { css, keyframes, createGlobalStyle, withTheme, ThemeProvider, ThemeConsumer, ThemeContext, isStyledComponent, ServerStyleSheet, StyleSheetManager, };
export default styled;
