// precise-ui/dist/es6/components/Badge/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled, { themed } from '../../utils/styled';
import { distance } from '../../distance';
import { getFontStyle } from '../../textStyles';
const BasicBadge = styled.div `
  position: relative;
  border-radius: 0.25em;
  text-align: center;
`;
const FilledBadge = styled(BasicBadge) `
  color: ${themed(({ theme }) => theme.badgeBackground)};
  background: ${themed(({ theme }) => theme.badgeColor)};
`;
const NormalBadge = styled(BasicBadge) `
  ${getFontStyle({ size: 'xSmall' })}
  display: inline-block;
  color: ${themed(({ theme }) => theme.badgeColor)};
  background: ${themed(({ theme }) => theme.badgeBackground)};
  min-width: 2em;
  padding: 0 ${distance.xsmall};
  border: solid 1px ${themed(({ theme }) => theme.badgeColor)};
`;
/**
 * The badge component is a simple informative display.
 */
export const Badge = Object.assign(((_a) => {
    var { fill = false } = _a, props = __rest(_a, ["fill"]);
    return fill ? React.createElement(FilledBadge, Object.assign({}, props)) : React.createElement(NormalBadge, Object.assign({}, props));
}), { inner: {
        get FilledBadge() { return FilledBadge; },
        get NormalBadge() { return NormalBadge; }
    } });
Badge.displayName = 'Badge';
