// precise-ui/dist/es5/components/OverflowMenu/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var react_onclickoutside_1 = require("react-onclickoutside");
var InteractiveList_1 = require("../InteractiveList");
var Flyout_1 = require("../Flyout");
var CustomWrapper = function (_a) {
    var _0 = _a.open, _1 = _a.border, _2 = _a.direction, props = __rest(_a, ["open", "border", "direction"]);
    return (React.createElement("div", __assign({}, props)));
};
CustomWrapper.displayName = 'CustomWrapper';
function getItems(group) {
    return group.map(function (item, index) { return ({
        content: item,
        key: index.toString(),
    }); });
}
var OverflowMenuInt = /** @class */ (function (_super) {
    __extends(OverflowMenuInt, _super);
    function OverflowMenuInt(props) {
        var _this = _super.call(this, props) || this;
        _this.handleClickOutside = function () {
            _this.state.open &&
                _this.setState({
                    open: false,
                });
        };
        _this.toggleGroup = function (e) {
            var open = _this.state.open;
            _this.setState({
                open: !open,
            });
            e.preventDefault();
        };
        _this.closeList = function () {
            _this.setState({
                open: false,
            });
        };
        _this.state = {
            open: false,
            items: getItems(props.items),
        };
        return _this;
    }
    OverflowMenuInt.prototype.UNSAFE_componentWillReceiveProps = function (nextProps) {
        this.setState({
            items: getItems(nextProps.items),
        });
    };
    OverflowMenuInt.prototype.render = function () {
        var _a = this.state, open = _a.open, items = _a.items;
        var _b = this.props, toggleButton = _b.button, flyoutProps = _b.flyoutProps, interactiveListProps = _b.listProps;
        return (React.createElement(Flyout_1.Flyout, __assign({ noGutter: true, content: React.createElement(InteractiveList_1.InteractiveList, __assign({ open: open, autoFocus: true, data: items, onBlur: this.closeList, onChange: this.closeList, customWrapper: CustomWrapper }, interactiveListProps)) }, flyoutProps, { open: open }),
            React.createElement("div", { onClick: this.toggleGroup }, toggleButton || '...')));
    };
    OverflowMenuInt.inner = {
        get Flyout() { return Flyout_1.Flyout; }
    };
    return OverflowMenuInt;
}(React.Component));
exports.OverflowMenu = react_onclickoutside_1.default(OverflowMenuInt);
