// precise-ui/dist/es6/components/Flyout/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import onClickOutside from 'react-onclickoutside';
import styled from '../../utils/styled';
import { FlyoutWindow } from './FlyoutWindow.part';
const FlyoutContainer = styled.div `
  position: relative;
  display: inline-block;
  width: fit-content;
`;
const StyledTargetWrapper = styled.div ``;
class FlyoutInt extends React.Component {
    constructor(props) {
        super(props);
        this.handleClickOutside = () => {
            this.state.open && !this.state.controlled && this.setOpen(false);
        };
        this.setTargetRef = (el) => {
            this.targetContainer = el;
        };
        this.onClick = () => {
            if (!this.state.controlled) {
                this.setOpen(!this.state.open);
            }
        };
        this.state = {
            controlled: props.open !== undefined,
            open: props.open || false,
            targetRect: undefined,
            dirtyFlag: false,
        };
    }
    componentDidMount() {
        this.updateMeasurements();
    }
    UNSAFE_componentWillReceiveProps(nextProps) {
        if (this.state.controlled && nextProps.open !== undefined) {
            this.setOpen(nextProps.open);
        }
    }
    componentDidUpdate() {
        const { dirtyFlag } = this.state;
        if (dirtyFlag) {
            this.updateMeasurements();
            this.setState({
                dirtyFlag: false,
            });
        }
    }
    updateMeasurements() {
        if (this.targetContainer) {
            const targetRect = this.targetContainer.getBoundingClientRect();
            this.setState({
                targetRect,
            });
        }
    }
    setOpen(open) {
        const { onChange } = this.props;
        if (typeof onChange === 'function') {
            onChange({ open });
        }
        this.setState({
            open,
            dirtyFlag: open === true,
        });
    }
    render() {
        const _a = this.props, { children, content, open: _0, onChange: _1 } = _a, props = __rest(_a, ["children", "content", "open", "onChange"]);
        const { targetRect } = this.state;
        const { open } = this.state;
        return (React.createElement(FlyoutContainer, { ref: this.setTargetRef },
            React.createElement(StyledTargetWrapper, { onClick: this.onClick }, children),
            !!content && targetRect && open && (React.createElement(FlyoutWindow, Object.assign({}, props, { targetRect: targetRect }), content))));
    }
}
FlyoutInt.inner = {
    get FlyoutContainer() { return FlyoutContainer; },
    get StyledTargetWrapper() { return StyledTargetWrapper; },
    get FlyoutWindow() { return FlyoutWindow; }
};
export const Flyout = onClickOutside(FlyoutInt);
Flyout.displayName = 'Flyout';
