// precise-ui/dist/es6/components/Dashboard/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled from '../../utils/styled';
import { InteractiveSurface } from '../InteractiveSurface';
import { Grid } from '../Grid';
import { GridArea } from '../GridArea';
function clamp(lower, value, upper) {
    return Math.min(Math.max(lower, value), upper);
}
function calc(rel, off, total, dim) {
    const value = clamp(0, rel * total - off, total - dim);
    return `${value}px`;
}
function repeat(count, dim) {
    const value = dim !== undefined ? `${dim}px` : '1fr';
    if (count !== undefined) {
        const values = [];
        for (let i = 0; i < count; i++) {
            values.push(value);
        }
        return values;
    }
    return value;
}
function collides(a, b) {
    if (a.hidden || b.hidden) {
        return false;
    }
    else if (a.column !== undefined && a.row !== undefined && b.column !== undefined && b.row !== undefined) {
        const { colSpan: acp = 1, rowSpan: arp = 1 } = a;
        const { colSpan: bcp = 1, rowSpan: brp = 1 } = b;
        const acs = a.column;
        const ace = a.column + acp - 1;
        const ars = a.row;
        const are = a.row + arp - 1;
        const bcs = b.column;
        const bce = b.column + bcp - 1;
        const brs = b.row;
        const bre = b.row + brp - 1;
        return acs <= bce && ace >= bcs && ars <= bre && are >= brs;
    }
    return false;
}
function notEqual(a, b) {
    if (a !== b) {
        if (a.length !== b.length) {
            return true;
        }
        for (let i = a.length; i--;) {
            const at = a[i];
            const bt = b[i];
            if (at !== bt ||
                at.colSpan !== bt.colSpan ||
                at.column !== bt.column ||
                at.id !== bt.id ||
                at.row !== bt.row ||
                at.rowSpan !== bt.rowSpan) {
                return true;
            }
        }
    }
    return false;
}
function resetStyle(node) {
    // tslint:disable-next-line
    const defaultValue = null;
    const style = node.style;
    style.position = 'static';
    style.cursor = defaultValue;
    style.width = defaultValue;
    style.height = defaultValue;
    style.left = defaultValue;
    style.top = defaultValue;
    style.zIndex = defaultValue;
}
const Preview = styled.div `
  background: #eee;
  border: 1px dashed #ccc;
  width: 100%;
  height: 100%;
`;
function getPreview(preview, tile) {
    const content = preview === true ? React.createElement(Preview, null) : preview;
    return (React.createElement(GridArea, { colSpan: tile.colSpan, rowSpan: tile.rowSpan, column: tile.column, row: tile.row }, content));
}
const defaultActiveTile = {
    x: 0,
    y: 0,
    width: 0,
    height: 0,
};
function changeTile(oldTiles, newTile) {
    let changed = false;
    const newTiles = oldTiles.map(tile => {
        if (tile.id === newTile.id) {
            changed =
                changed ||
                    tile.colSpan !== newTile.colSpan ||
                    tile.column !== newTile.column ||
                    tile.hidden !== newTile.hidden ||
                    tile.row !== newTile.row ||
                    tile.rowSpan !== newTile.rowSpan;
            return newTile;
        }
        else if (collides(tile, newTile)) {
            changed = true;
            return {
                id: tile.id,
                colSpan: tile.colSpan,
                rowSpan: tile.rowSpan,
            };
        }
        return tile;
    });
    return changed ? newTiles : oldTiles;
}
/**
 * Dashboard component.
 */
export class Dashboard extends React.Component {
    constructor(props) {
        super(props);
        this.setters = [];
        this.dragTile = (e) => {
            const { activeTile } = this;
            if (activeTile) {
                const node = activeTile.node;
                const current = activeTile.current;
                if (!e.moved) {
                }
                else if (e.active) {
                    const style = node.style;
                    style.left = calc(e.x, activeTile.x, e.rect.width, activeTile.width);
                    style.top = calc(e.y, activeTile.y, e.rect.height, activeTile.height);
                    this.previewDrag(e.x, e.y, current);
                }
                else {
                    const pos = this.getTargetPosition(e.x, e.y, current);
                    resetStyle(node);
                    this.finishDrag(Object.assign({}, current, { column: pos.column, row: pos.row }));
                }
            }
            else {
                e.release();
            }
        };
        this.setLayout = ({ layout }) => {
            this.layout = layout;
        };
        const { defaultTiles = [] } = props;
        this.state = {
            tiles: [...defaultTiles],
            live: undefined,
        };
    }
    UNSAFE_componentWillReceiveProps(nextProps) {
        const { defaultTiles: currTiles = [] } = this.props;
        const { defaultTiles: nextTiles = [] } = nextProps;
        if (notEqual(currTiles, nextTiles)) {
            this.setState({
                tiles: [...nextTiles],
                live: undefined,
            });
        }
    }
    finishDrag(newTile) {
        const { onChange } = this.props;
        const { tiles: oldTiles } = this.state;
        const newTiles = changeTile(oldTiles, newTile);
        this.activeTile = undefined;
        if (newTiles !== oldTiles) {
            this.setState({
                tiles: newTiles,
                live: undefined,
            });
            if (typeof onChange === 'function') {
                onChange({
                    tile: newTile,
                    tiles: newTiles,
                });
            }
        }
        else {
            this.setState({
                live: undefined,
            });
        }
    }
    previewDrag(h, v, tile) {
        const active = this.activeTile;
        const { preview } = this.props;
        const { live, tiles } = this.state;
        if (preview && active && live) {
            const current = active.current;
            const pos = this.getTargetPosition(h, v, tile);
            const last = live[active.index];
            if (pos.column !== last.column || pos.row !== last.row) {
                this.setState({
                    live: changeTile(tiles, Object.assign({}, current, { column: pos.column, row: pos.row })),
                });
            }
        }
    }
    getTargetPosition(h, v, tile) {
        const { columns, rows } = this.layout;
        const { colSpan = 1, rowSpan = 1 } = tile;
        const { x, y, width, height } = this.activeTile || defaultActiveTile;
        const totalRows = rows.length;
        const totalColumns = columns.length;
        const columnOffset = ~~((x * colSpan) / width);
        const rowOffset = ~~((y * rowSpan) / height);
        return {
            column: clamp(0, ~~(h * totalColumns) - columnOffset, totalColumns - colSpan),
            row: clamp(0, ~~(v * totalRows) - rowOffset, totalRows - rowSpan),
        };
    }
    setActiveTile(e, index) {
        const { disabled } = this.props;
        const { tiles } = this.state;
        const current = tiles[index];
        if (!disabled) {
            this.activeTile = Object.assign({}, e, { index,
                current });
            const style = e.node.style;
            style.cursor = 'move';
            style.position = 'absolute';
            style.width = `${e.width}px`;
            style.height = `${e.height}px`;
            style.zIndex = '100000000';
            this.setState({
                live: tiles,
            });
        }
    }
    render() {
        const _a = this.props, { columnWidth, columnCount = 5, rowHeight, rowCount, spacing = 10, theme, children, disabled, preview, defaultTiles: _0, onChange: _1, emptyTiles } = _a, props = __rest(_a, ["columnWidth", "columnCount", "rowHeight", "rowCount", "spacing", "theme", "children", "disabled", "preview", "defaultTiles", "onChange", "emptyTiles"]);
        const { tiles, live } = this.state;
        const currentTiles = live || tiles;
        const columns = repeat(columnCount, columnWidth);
        const rows = repeat(rowCount, rowHeight);
        const active = this.activeTile;
        return (React.createElement(InteractiveSurface, { theme: theme, onChange: this.dragTile, disabled: disabled },
            React.createElement(Grid, Object.assign({ theme: theme, rows: rows, columns: columns, spacing: `${spacing}px`, showEmptyCells: emptyTiles, onLayout: this.setLayout }, props),
                React.Children.map(children, (child, index) => {
                    const tile = currentTiles[index];
                    const setters = this.setters;
                    if (setters[index] === undefined) {
                        setters[index] = e => this.setActiveTile(e, index);
                    }
                    return (tile && (React.createElement(GridArea, { key: tile.id, theme: theme, onTap: setters[index], column: tile.column, colSpan: tile.colSpan, rowSpan: tile.rowSpan, hidden: tile.hidden, row: tile.row }, child)));
                }),
                preview && active && getPreview(preview, currentTiles[active.index]))));
    }
}
Dashboard.inner = {
    get InteractiveSurface() { return InteractiveSurface; },
    get Grid() { return Grid; },
    get GridArea() { return GridArea; }
};
