// precise-ui/dist/es5/components/ColorPicker/index.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var styled_1 = require("../../utils/styled");
var Slider_1 = require("../Slider");
var IndicatorKnob_1 = require("../IndicatorKnob");
var InteractiveSurface_1 = require("../InteractiveSurface");
var colors_1 = require("../../utils/colors");
var withFormContext_1 = require("../../hoc/withFormContext");
var distance_1 = require("../../distance");
var PaddedContainer_1 = require("../PaddedContainer");
var InputNotification_1 = require("../InputNotification");
function rgb(c) {
    return "rgb(" + c.r + ", " + c.g + ", " + c.b + ")";
}
function computeColor(color) {
    var fc = color;
    if (typeof color === 'string') {
        color = colors_1.parseColor(color);
    }
    else if (typeof fc.h === 'number') {
        return {
            hsv: fc,
            color: color,
            base: colors_1.hsvToRgb(fc.h, 1, 1),
        };
    }
    else {
        color = {
            r: color.r,
            g: color.g,
            b: color.b,
            a: color.a,
        };
    }
    var hsv = colors_1.rgbToHsv(color.r, color.g, color.b, color.a);
    return {
        hsv: hsv,
        color: color,
        base: colors_1.hsvToRgb(hsv.h, 1, 1),
    };
}
var PickerContainer = styled_1.default.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  margin: ", ";\n"], ["\n  margin: ", ";\n"])), distance_1.distance.small);
var PickerSurface = styled_1.default(InteractiveSurface_1.InteractiveSurface)(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  width: ", ";\n  height: ", ";\n  position: relative;\n  background-color: ", ";\n"], ["\n  width: ", ";\n  height: ", ";\n  position: relative;\n  background-color: ", ";\n"])), function (props) { return props.width; }, function (props) { return props.height; }, function (props) { return props.color; });
var PickerSaturation = styled_1.default.div(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  position: absolute;\n  left: 0;\n  right: 0;\n  top: 0;\n  bottom: 0;\n  background: linear-gradient(to bottom, transparent 0%, black 100%);\n"], ["\n  position: absolute;\n  left: 0;\n  right: 0;\n  top: 0;\n  bottom: 0;\n  background: linear-gradient(to bottom, transparent 0%, black 100%);\n"])));
var PickerValue = styled_1.default.div(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  position: absolute;\n  left: 0;\n  right: 0;\n  top: 0;\n  bottom: 0;\n  background: linear-gradient(to right, white 0%, transparent 100%);\n"], ["\n  position: absolute;\n  left: 0;\n  right: 0;\n  top: 0;\n  bottom: 0;\n  background: linear-gradient(to right, white 0%, transparent 100%);\n"])));
var ColorSlider = styled_1.default(Slider_1.Slider)(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  margin: ", " 0 -", " 0;\n  background: linear-gradient(\n    to left,\n    red 0,\n    #f09 10%,\n    #cd00ff 20%,\n    #3200ff 30%,\n    #06f 40%,\n    #00fffd 50%,\n    #0f6 60%,\n    #35ff00 70%,\n    #cdff00 80%,\n    #f90 90%,\n    red 100%\n  );\n"], ["\n  margin: ", " 0 -", " 0;\n  background: linear-gradient(\n    to left,\n    red 0,\n    #f09 10%,\n    #cd00ff 20%,\n    #3200ff 30%,\n    #06f 40%,\n    #00fffd 50%,\n    #0f6 60%,\n    #35ff00 70%,\n    #cdff00 80%,\n    #f90 90%,\n    red 100%\n  );\n"])), distance_1.distance.large, distance_1.distance.xsmall);
var OpacitySlider = styled_1.default(Slider_1.Slider)(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  margin: ", " 0 -", " 0;\n  background: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAYAAACNMs+9AAAAJUlEQVQYV2N89erVfwY0ICYmxoguxjgUFKI7GsTH5m4M3w1ChQC1/Ca8i2n1WgAAAABJRU5ErkJggg==');\n  &:before {\n    background: linear-gradient(to right, transparent 0px, rgba(0, 117, 120, 0.357) 100%);\n    border-radius: 10px;\n    position: absolute;\n    left: 0;\n    right: 0;\n    top: 0;\n    bottom: 0;\n    content: '';\n  }\n"], ["\n  margin: ", " 0 -", " 0;\n  background: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAYAAACNMs+9AAAAJUlEQVQYV2N89erVfwY0ICYmxoguxjgUFKI7GsTH5m4M3w1ChQC1/Ca8i2n1WgAAAABJRU5ErkJggg==');\n  &:before {\n    background: linear-gradient(to right, transparent 0px, rgba(0, 117, 120, 0.357) 100%);\n    border-radius: 10px;\n    position: absolute;\n    left: 0;\n    right: 0;\n    top: 0;\n    bottom: 0;\n    content: '';\n  }\n"])), distance_1.distance.large, distance_1.distance.xsmall);
var ColorIndicator = styled_1.default(IndicatorKnob_1.IndicatorKnob)(templateObject_7 || (templateObject_7 = __makeTemplateObject(["\n  box-shadow: 0 0 20px -5px ", ";\n  border: 2px solid ", ";\n"], ["\n  box-shadow: 0 0 20px -5px ", ";\n  border: 2px solid ", ";\n"])), function (props) { return (props.disabled ? 'transparent' : '#000'); }, function (props) { return (props.disabled ? '#ccc' : props.active ? '#eee' : '#fff'); });
var ColorPickerInt = /** @class */ (function (_super) {
    __extends(ColorPickerInt, _super);
    function ColorPickerInt(props) {
        var _this = _super.call(this, props) || this;
        _this.changeColor = function (e) {
            var s = e.x;
            var v = 1 - e.y;
            var notify = _this.props.onChange;
            var current = _this.state.value;
            var rgb = colors_1.hsvToRgb(current.h, s, v, current.a);
            var selected = __assign({}, current, { s: s, v: v }, rgb);
            if (!_this.state.controlled) {
                _this.changeTo({
                    value: selected,
                });
            }
            _this.setState({
                active: e.active,
            });
            if (typeof notify === 'function') {
                notify({
                    value: selected,
                });
            }
        };
        _this.changeBackground = function (e) {
            var h = e.value;
            var notify = _this.props.onChange;
            var current = _this.state.value;
            var rgb = colors_1.hsvToRgb(h, current.s, current.v, current.a);
            var selected = __assign({}, current, { h: h, s: current.s, v: current.v }, rgb);
            if (!_this.state.controlled) {
                var base = colors_1.hsvToRgb(h, 1, 1);
                _this.changeTo({
                    base: base,
                    value: selected,
                });
            }
            if (typeof notify === 'function') {
                notify({
                    value: selected,
                });
            }
        };
        _this.changeOpacity = function (e) {
            var alpha = e.value;
            var notify = _this.props.onChange;
            var current = _this.state.value;
            var selected = __assign({}, current, { a: alpha });
            if (!_this.state.controlled) {
                _this.changeTo({
                    value: selected,
                });
            }
            if (typeof notify === 'function') {
                notify({
                    value: selected,
                });
            }
        };
        var _a = computeColor(props.value || props.defaultValue || { r: 0, g: 0, b: 0 }), hsv = _a.hsv, color = _a.color, base = _a.base;
        _this.state = {
            controlled: props.value !== undefined,
            active: false,
            value: __assign({ a: 1 }, color, { h: hsv.h, s: hsv.s, v: hsv.v }),
            base: base,
            error: props.error,
        };
        return _this;
    }
    ColorPickerInt.prototype.UNSAFE_componentWillReceiveProps = function (_a) {
        var value = _a.value, error = _a.error;
        if (value && value !== this.state.value) {
            var _b = computeColor(value || this.state.value), hsv = _b.hsv, color = _b.color, base = _b.base;
            this.setState({
                value: __assign({ a: 1 }, color, { h: hsv.h, s: hsv.s, v: hsv.v }),
                base: base,
            });
        }
        this.setState({ error: error });
    };
    ColorPickerInt.prototype.componentDidMount = function () {
        var form = this.props.form;
        var controlled = this.state.controlled;
        if (!controlled && form) {
            form.subscribe(this);
        }
    };
    ColorPickerInt.prototype.componentWillUnmount = function () {
        var form = this.props.form;
        var controlled = this.state.controlled;
        if (!controlled && form) {
            form.unsubscribe(this);
        }
    };
    ColorPickerInt.prototype.changeTo = function (state) {
        var _a = this.props, form = _a.form, _b = _a.name, name = _b === void 0 ? '' : _b;
        if (form) {
            form.change({
                name: name,
                value: state.value,
            });
        }
        else {
            this.setState(state);
        }
    };
    ColorPickerInt.prototype.render = function () {
        var _a = this.state, value = _a.value, active = _a.active, base = _a.base, error = _a.error;
        var _b = this.props, _0 = _b.defaultValue, _1 = _b.value, _2 = _b.onChange, _3 = _b.onInput, allowOpacity = _b.allowOpacity, hideBar = _b.hideBar, _c = _b.width, width = _c === void 0 ? '100%' : _c, _d = _b.height, height = _d === void 0 ? '200px' : _d, info = _b.info, props = __rest(_b, ["defaultValue", "value", "onChange", "onInput", "allowOpacity", "hideBar", "width", "height", "info"]);
        var baseRgb = rgb(base);
        var currentRgb = rgb(value);
        var surface = {
            color: baseRgb,
            width: width,
            height: height,
        };
        return (React.createElement(PickerContainer, __assign({}, props),
            React.createElement(PickerSurface, __assign({}, surface, { onChange: this.changeColor }),
                React.createElement(PickerValue, null),
                React.createElement(PickerSaturation, null),
                React.createElement(ColorIndicator, { x: value.s, y: 1 - value.v, color: currentRgb, active: active })),
            !hideBar && React.createElement(ColorSlider, { onChange: this.changeBackground, value: value.h, color: baseRgb }),
            allowOpacity && React.createElement(OpacitySlider, { onChange: this.changeOpacity, value: value.a }),
            (error || info) && (React.createElement(PaddedContainer_1.PaddedContainer, { top: "small", bottom: "xsmall" },
                React.createElement(InputNotification_1.InputNotification, { error: error, info: info })))));
    };
    ColorPickerInt.inner = {
        get PickerContainer() { return PickerContainer; },
        get PickerSurface() { return PickerSurface; },
        get PickerValue() { return PickerValue; },
        get PickerSaturation() { return PickerSaturation; },
        get ColorIndicator() { return ColorIndicator; },
        get ColorSlider() { return ColorSlider; },
        get OpacitySlider() { return OpacitySlider; },
        get PaddedContainer() { return PaddedContainer_1.PaddedContainer; },
        get InputNotification() { return InputNotification_1.InputNotification; }
    };
    return ColorPickerInt;
}(React.PureComponent));
/**
 * A color picker for selecting a color in the UI.
 */
exports.ColorPicker = withFormContext_1.withFormContext(ColorPickerInt);
exports.ColorPicker.displayName = 'ColorPicker';
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7;
