// precise-ui/dist/es6/components/OverflowMenu/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import onClickOutside from 'react-onclickoutside';
import { InteractiveList, } from '../InteractiveList';
import { Flyout } from '../Flyout';
const CustomWrapper = (_a) => {
    var { open: _0, border: _1, direction: _2 } = _a, props = __rest(_a, ["open", "border", "direction"]);
    return (React.createElement("div", Object.assign({}, props)));
};
CustomWrapper.displayName = 'CustomWrapper';
function getItems(group) {
    return group.map((item, index) => ({
        content: item,
        key: index.toString(),
    }));
}
class OverflowMenuInt extends React.Component {
    constructor(props) {
        super(props);
        this.handleClickOutside = () => {
            this.state.open &&
                this.setState({
                    open: false,
                });
        };
        this.toggleGroup = (e) => {
            const { open } = this.state;
            this.setState({
                open: !open,
            });
            e.preventDefault();
        };
        this.closeList = () => {
            this.setState({
                open: false,
            });
        };
        this.state = {
            open: false,
            items: getItems(props.items),
        };
    }
    UNSAFE_componentWillReceiveProps(nextProps) {
        this.setState({
            items: getItems(nextProps.items),
        });
    }
    render() {
        const { open, items } = this.state;
        const { button: toggleButton, flyoutProps, listProps: interactiveListProps } = this.props;
        return (React.createElement(Flyout, Object.assign({ noGutter: true, content: React.createElement(InteractiveList, Object.assign({ open: open, autoFocus: true, data: items, onBlur: this.closeList, onChange: this.closeList, customWrapper: CustomWrapper }, interactiveListProps)) }, flyoutProps, { open: open }),
            React.createElement("div", { onClick: this.toggleGroup }, toggleButton || '...')));
    }
}
OverflowMenuInt.inner = {
    get Flyout() { return Flyout; }
};
export const OverflowMenu = onClickOutside(OverflowMenuInt);
