// precise-ui/dist/es6/components/ColorPicker/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled from '../../utils/styled';
import { Slider } from '../Slider';
import { IndicatorKnob } from '../IndicatorKnob';
import { InteractiveSurface } from '../InteractiveSurface';
import { rgbToHsv, hsvToRgb, parseColor } from '../../utils/colors';
import { withFormContext } from '../../hoc/withFormContext';
import { distance } from '../../distance';
import { PaddedContainer } from '../PaddedContainer';
import { InputNotification } from '../InputNotification';
function rgb(c) {
    return `rgb(${c.r}, ${c.g}, ${c.b})`;
}
function computeColor(color) {
    const fc = color;
    if (typeof color === 'string') {
        color = parseColor(color);
    }
    else if (typeof fc.h === 'number') {
        return {
            hsv: fc,
            color,
            base: hsvToRgb(fc.h, 1, 1),
        };
    }
    else {
        color = {
            r: color.r,
            g: color.g,
            b: color.b,
            a: color.a,
        };
    }
    const hsv = rgbToHsv(color.r, color.g, color.b, color.a);
    return {
        hsv,
        color,
        base: hsvToRgb(hsv.h, 1, 1),
    };
}
const PickerContainer = styled.div `
  margin: ${distance.small};
`;
const PickerSurface = styled(InteractiveSurface) `
  width: ${props => props.width};
  height: ${props => props.height};
  position: relative;
  background-color: ${props => props.color};
`;
const PickerSaturation = styled.div `
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  background: linear-gradient(to bottom, transparent 0%, black 100%);
`;
const PickerValue = styled.div `
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  background: linear-gradient(to right, white 0%, transparent 100%);
`;
const ColorSlider = styled(Slider) `
  margin: ${distance.large} 0 -${distance.xsmall} 0;
  background: linear-gradient(
    to left,
    red 0,
    #f09 10%,
    #cd00ff 20%,
    #3200ff 30%,
    #06f 40%,
    #00fffd 50%,
    #0f6 60%,
    #35ff00 70%,
    #cdff00 80%,
    #f90 90%,
    red 100%
  );
`;
const OpacitySlider = styled(Slider) `
  margin: ${distance.large} 0 -${distance.xsmall} 0;
  background: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAYAAACNMs+9AAAAJUlEQVQYV2N89erVfwY0ICYmxoguxjgUFKI7GsTH5m4M3w1ChQC1/Ca8i2n1WgAAAABJRU5ErkJggg==');
  &:before {
    background: linear-gradient(to right, transparent 0px, rgba(0, 117, 120, 0.357) 100%);
    border-radius: 10px;
    position: absolute;
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    content: '';
  }
`;
const ColorIndicator = styled(IndicatorKnob) `
  box-shadow: 0 0 20px -5px ${props => (props.disabled ? 'transparent' : '#000')};
  border: 2px solid ${props => (props.disabled ? '#ccc' : props.active ? '#eee' : '#fff')};
`;
class ColorPickerInt extends React.PureComponent {
    constructor(props) {
        super(props);
        this.changeColor = (e) => {
            const s = e.x;
            const v = 1 - e.y;
            const notify = this.props.onChange;
            const current = this.state.value;
            const rgb = hsvToRgb(current.h, s, v, current.a);
            const selected = Object.assign({}, current, { s: s, v: v }, rgb);
            if (!this.state.controlled) {
                this.changeTo({
                    value: selected,
                });
            }
            this.setState({
                active: e.active,
            });
            if (typeof notify === 'function') {
                notify({
                    value: selected,
                });
            }
        };
        this.changeBackground = (e) => {
            const h = e.value;
            const notify = this.props.onChange;
            const current = this.state.value;
            const rgb = hsvToRgb(h, current.s, current.v, current.a);
            const selected = Object.assign({}, current, { h, s: current.s, v: current.v }, rgb);
            if (!this.state.controlled) {
                const base = hsvToRgb(h, 1, 1);
                this.changeTo({
                    base,
                    value: selected,
                });
            }
            if (typeof notify === 'function') {
                notify({
                    value: selected,
                });
            }
        };
        this.changeOpacity = (e) => {
            const alpha = e.value;
            const notify = this.props.onChange;
            const current = this.state.value;
            const selected = Object.assign({}, current, { a: alpha });
            if (!this.state.controlled) {
                this.changeTo({
                    value: selected,
                });
            }
            if (typeof notify === 'function') {
                notify({
                    value: selected,
                });
            }
        };
        const { hsv, color, base } = computeColor(props.value || props.defaultValue || { r: 0, g: 0, b: 0 });
        this.state = {
            controlled: props.value !== undefined,
            active: false,
            value: Object.assign({ a: 1 }, color, { h: hsv.h, s: hsv.s, v: hsv.v }),
            base,
            error: props.error,
        };
    }
    UNSAFE_componentWillReceiveProps({ value, error }) {
        if (value && value !== this.state.value) {
            const { hsv, color, base } = computeColor(value || this.state.value);
            this.setState({
                value: Object.assign({ a: 1 }, color, { h: hsv.h, s: hsv.s, v: hsv.v }),
                base,
            });
        }
        this.setState({ error });
    }
    componentDidMount() {
        const { form } = this.props;
        const { controlled } = this.state;
        if (!controlled && form) {
            form.subscribe(this);
        }
    }
    componentWillUnmount() {
        const { form } = this.props;
        const { controlled } = this.state;
        if (!controlled && form) {
            form.unsubscribe(this);
        }
    }
    changeTo(state) {
        const { form, name = '' } = this.props;
        if (form) {
            form.change({
                name,
                value: state.value,
            });
        }
        else {
            this.setState(state);
        }
    }
    render() {
        const { value, active, base, error } = this.state;
        const _a = this.props, { defaultValue: _0, value: _1, onChange: _2, onInput: _3, allowOpacity, hideBar, width = '100%', height = '200px', info } = _a, props = __rest(_a, ["defaultValue", "value", "onChange", "onInput", "allowOpacity", "hideBar", "width", "height", "info"]);
        const baseRgb = rgb(base);
        const currentRgb = rgb(value);
        const surface = {
            color: baseRgb,
            width,
            height,
        };
        return (React.createElement(PickerContainer, Object.assign({}, props),
            React.createElement(PickerSurface, Object.assign({}, surface, { onChange: this.changeColor }),
                React.createElement(PickerValue, null),
                React.createElement(PickerSaturation, null),
                React.createElement(ColorIndicator, { x: value.s, y: 1 - value.v, color: currentRgb, active: active })),
            !hideBar && React.createElement(ColorSlider, { onChange: this.changeBackground, value: value.h, color: baseRgb }),
            allowOpacity && React.createElement(OpacitySlider, { onChange: this.changeOpacity, value: value.a }),
            (error || info) && (React.createElement(PaddedContainer, { top: "small", bottom: "xsmall" },
                React.createElement(InputNotification, { error: error, info: info })))));
    }
}
ColorPickerInt.inner = {
    get PickerContainer() { return PickerContainer; },
    get PickerSurface() { return PickerSurface; },
    get PickerValue() { return PickerValue; },
    get PickerSaturation() { return PickerSaturation; },
    get ColorIndicator() { return ColorIndicator; },
    get ColorSlider() { return ColorSlider; },
    get OpacitySlider() { return OpacitySlider; },
    get PaddedContainer() { return PaddedContainer; },
    get InputNotification() { return InputNotification; }
};
/**
 * A color picker for selecting a color in the UI.
 */
export const ColorPicker = withFormContext(ColorPickerInt);
ColorPicker.displayName = 'ColorPicker';
