// precise-ui/dist/es6/components/StepIndicator/StepIndicatorVertical.part.js
import * as React from 'react';
import styled, { themed } from '../../utils/styled';
import { ListItemNumber } from './StepIndicatorShared.part';
import { distance } from '../../distance';
import { getFontStyle } from '../../textStyles';
const StyledProgressContent = styled.div `
  min-height: ${distance.xxxlarge};
`;
const StyledProgressStep = styled.li `
  display: block;
  position: relative;
  padding-left: 40px;
  color: ${themed(({ theme }) => theme.text1)};

  &:before {
    content: '';
    display: block;
    border-radius: 50%;
    width: ${distance.small};
    height: ${distance.small};
    position: absolute;
    top: ${distance.small};
    left: ${distance.small};
    z-index: 1;
    background-color: ${themed(({ theme }) => theme.text1)};
  }

  &:not(:last-child):after {
    content: '';
    position: absolute;
    top: ${distance.small};
    left: 11px;
    width: 2px;
    height: 100%;
    background-color: ${themed(({ theme }) => theme.text1)};
  }

  &.step-active {
    ${getFontStyle({ weight: 'medium' })}
    
    color: ${themed(({ theme }) => theme.text1)};

    ~ li {
      color: ${themed(({ theme }) => theme.text2)};

      &:before {
        background-color: ${themed(({ theme }) => theme.text2)};
      }

      &:after {
        background-color: ${themed(({ theme }) => theme.text2)};
      }
    }

    &:after {
      background-color: ${themed(({ theme }) => theme.text2)};
    }

    &:before {
      background-color: ${themed(({ theme }) => theme.ui1)};
      border: 0.5em solid ${themed(({ theme }) => theme.text1)};
      left: 0;
      top: 0;
    }
  }
`;
const StyledProgressText = styled.div `
  display: block;
`;
const StyledProgressHost = styled.ol `
  width: 100%;
  display: block;
  padding: 0;
  margin: 0;

  & ${StyledProgressStep}:last-of-type ${StyledProgressContent} {
    min-height: auto;
  }
`;
function createStep(header, content, i, active, numbered, theme) {
    return (React.createElement(StyledProgressStep, { theme: theme, key: i, className: active ? 'step-active' : '' },
        numbered && React.createElement(ListItemNumber, { theme: theme }, i + 1),
        React.createElement(StyledProgressText, { theme: theme }, header),
        React.createElement(StyledProgressContent, null, content)));
}
function getHeaders(steps, current, numbered, theme) {
    return steps.map((step, i) => createStep(step, undefined, i, i === current, numbered, theme));
}
function getContent(children, current, numbered, theme) {
    const items = [];
    React.Children.forEach(children, (element, i) => {
        if (element && React.isValidElement(element)) {
            const { header, children } = element.props;
            items.push(createStep(header, children, i, i === current, numbered, theme));
        }
    });
    return items;
}
export const StepIndicatorVertical = Object.assign((({ theme, steps, current = 0, numbered, children, }) => {
    const items = steps !== undefined ? getHeaders(steps, current, numbered, theme) : getContent(children, current, numbered, theme);
    return React.createElement(StyledProgressHost, null, items);
}), { inner: {
        get StyledProgressHost() { return StyledProgressHost; }
    } });
