// precise-ui/dist/es5/components/StepIndicator/StepIndicatorHorizontal.part.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var styled_1 = require("../../utils/styled");
var StepIndicatorShared_part_1 = require("./StepIndicatorShared.part");
var distance_1 = require("../../distance");
var textStyles_1 = require("../../textStyles");
var StyledProgressHost = styled_1.default.ol(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  width: 100%;\n  display: table;\n  table-layout: fixed;\n  padding: 0;\n  margin: 0;\n  height: 4em;\n  box-sizing: border-box;\n  padding-left: ", ";\n"], ["\n  width: 100%;\n  display: table;\n  table-layout: fixed;\n  padding: 0;\n  margin: 0;\n  height: 4em;\n  box-sizing: border-box;\n  padding-left: ", ";\n"])), distance_1.distance.medium);
var HorizontalNumber = styled_1.default(StepIndicatorShared_part_1.ListItemNumber)(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  top: ", ";\n  left: -12px;\n"], ["\n  top: ", ";\n  left: -12px;\n"])), distance_1.distance.small);
var StyledProgressStep = styled_1.default.li(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  display: table-cell;\n  text-align: center;\n  vertical-align: top;\n  position: relative;\n  color: ", ";\n\n  &:before {\n    content: '';\n    display: block;\n    border-radius: 50%;\n    margin: 1em auto;\n    width: ", ";\n    height: ", ";\n    position: relative;\n    left: -50%;\n    z-index: 1;\n    background-color: ", ";\n  }\n\n  &:not(:first-child):after {\n    content: '';\n    position: absolute;\n    top: 19px;\n    left: -100%;\n    width: 100%;\n    height: 2px;\n    background-color: ", ";\n  }\n\n  &.step-active {\n    ", "\n    \n    color: ", ";\n\n    ~ li {\n      color: ", ";\n\n      &:before {\n        background-color: ", ";\n      }\n\n      &:after {\n        background-color: ", ";\n      }\n    }\n\n    &:before {\n      background-color: ", ";\n      border: 0.5em solid ", ";\n      margin: 0.5em auto;\n    }\n  }\n"], ["\n  display: table-cell;\n  text-align: center;\n  vertical-align: top;\n  position: relative;\n  color: ", ";\n\n  &:before {\n    content: '';\n    display: block;\n    border-radius: 50%;\n    margin: 1em auto;\n    width: ", ";\n    height: ", ";\n    position: relative;\n    left: -50%;\n    z-index: 1;\n    background-color: ", ";\n  }\n\n  &:not(:first-child):after {\n    content: '';\n    position: absolute;\n    top: 19px;\n    left: -100%;\n    width: 100%;\n    height: 2px;\n    background-color: ", ";\n  }\n\n  &.step-active {\n    ", "\n    \n    color: ", ";\n\n    ~ li {\n      color: ", ";\n\n      &:before {\n        background-color: ", ";\n      }\n\n      &:after {\n        background-color: ", ";\n      }\n    }\n\n    &:before {\n      background-color: ", ";\n      border: 0.5em solid ", ";\n      margin: 0.5em auto;\n    }\n  }\n"])), styled_1.themed(function (_a) {
    var theme = _a.theme;
    return theme.text1;
}), distance_1.distance.small, distance_1.distance.small, styled_1.themed(function (_a) {
    var theme = _a.theme;
    return theme.text1;
}), styled_1.themed(function (_a) {
    var theme = _a.theme;
    return theme.text1;
}), textStyles_1.getFontStyle({ weight: 'medium' }), styled_1.themed(function (_a) {
    var theme = _a.theme;
    return theme.text1;
}), styled_1.themed(function (_a) {
    var theme = _a.theme;
    return theme.text2;
}), styled_1.themed(function (_a) {
    var theme = _a.theme;
    return theme.text2;
}), styled_1.themed(function (_a) {
    var theme = _a.theme;
    return theme.text2;
}), styled_1.themed(function (_a) {
    var theme = _a.theme;
    return theme.ui1;
}), styled_1.themed(function (_a) {
    var theme = _a.theme;
    return theme.text1;
}));
var StyledProgressText = styled_1.default.div(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  display: inline-block;\n  overflow: hidden;\n  white-space: nowrap;\n  text-overflow: ellipsis;\n  position: absolute;\n  left: 0;\n  top: 2.5em;\n"], ["\n  display: inline-block;\n  overflow: hidden;\n  white-space: nowrap;\n  text-overflow: ellipsis;\n  position: absolute;\n  left: 0;\n  top: 2.5em;\n"])));
function createStep(header, i, active, numbered, theme) {
    return (React.createElement(StyledProgressStep, { theme: theme, key: i, className: active ? 'step-active' : '' },
        numbered && React.createElement(HorizontalNumber, { theme: theme }, i + 1),
        React.createElement(StyledProgressText, { theme: theme }, header)));
}
function getHeaders(steps, current, numbered, theme) {
    var items = steps.map(function (step, i) { return createStep(step, i, i === current, numbered, theme); });
    return {
        items: items,
        content: undefined,
    };
}
function getContent(children, current, numbered, theme) {
    var items = [];
    var content = undefined;
    React.Children.forEach(children, function (element, i) {
        if (element && React.isValidElement(element)) {
            var _a = element.props, header = _a.header, children_1 = _a.children;
            var active = i === current;
            items.push(createStep(header, i, active, numbered, theme));
            if (active) {
                content = children_1;
            }
        }
    });
    return {
        items: items,
        content: content,
    };
}
exports.StepIndicatorHorizontal = Object.assign((function (_a) {
    var theme = _a.theme, steps = _a.steps, _b = _a.current, current = _b === void 0 ? 0 : _b, numbered = _a.numbered, children = _a.children, props = __rest(_a, ["theme", "steps", "current", "numbered", "children"]);
    var _c = steps !== undefined ? getHeaders(steps, current, numbered, theme) : getContent(children, current, numbered, theme), items = _c.items, content = _c.content;
    return (React.createElement("div", __assign({}, props),
        React.createElement(StyledProgressHost, { theme: theme }, items),
        content));
}), { inner: {
        get StyledProgressHost() { return StyledProgressHost; }
    } });
var templateObject_1, templateObject_2, templateObject_3, templateObject_4;
