// precise-ui/dist/es6/components/IncreaseDecrease.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled, { themed, css } from '../utils/styled';
const IncreaseDecreaseContainer = styled.span `
  width: 4px;
  position: relative;
  &:before {
    content: ' ';
    white-space: pre;
  }
`;
const IncreaseDecreaseArrow = styled.div `
  width: 8px;
  height: 4px;
  position: absolute;
  cursor: pointer;
  left: 0;

  &:before {
    content: '';
    position: absolute;
    left: 0;
    right: 0;
    width: 0;
    height: 0;
    border-left: 4px solid transparent;
    border-right: 4px solid transparent;
  }
`;
const getBorderStyle = (position, active = false) => css `
  border-${position}: 4px solid ${themed(({ theme }) => (active ? theme.ui5 : theme.ui4))};
`;
const IncreaseDecreaseArrowUp = styled(IncreaseDecreaseArrow) `
  top: 4px;

  &:before {
    ${({ active }) => getBorderStyle('bottom', active)}
  }

  &:hover:before {
    ${getBorderStyle('bottom', true)}
  }
`;
const IncreaseDecreaseArrowDown = styled(IncreaseDecreaseArrow) `
  bottom: 4px;

  &:before {
    ${({ active }) => getBorderStyle('top', active)}
  }

  &:hover:before {
    ${getBorderStyle('top', true)}
  }
`;
export const IncreaseDecrease = Object.assign(((_a) => {
    var { onIncrease, onDecrease, value } = _a, props = __rest(_a, ["onIncrease", "onDecrease", "value"]);
    return (React.createElement(IncreaseDecreaseContainer, Object.assign({}, props),
        React.createElement(IncreaseDecreaseArrowUp, { onClick: onIncrease, active: value === 'increase' }),
        React.createElement(IncreaseDecreaseArrowDown, { onClick: onDecrease, active: value === 'decrease' })));
}), { inner: {
        get IncreaseDecreaseContainer() { return IncreaseDecreaseContainer; },
        get IncreaseDecreaseArrowUp() { return IncreaseDecreaseArrowUp; },
        get IncreaseDecreaseArrowDown() { return IncreaseDecreaseArrowDown; }
    } });
