// precise-ui/dist/es6/components/Spinner/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled, { keyframes, themed } from '../../utils/styled';
import { distance } from '../../distance';
import { remCalc } from '../../utils/remCalc';
function strToSize(size) {
    switch (size) {
        case 'x-small':
            return 5;
        case 'small':
            return 7;
        case 'large':
            return 12;
        case 'x-large':
            return 15;
    }
    return 10;
}
function strToFontSize(size) {
    return remCalc(size === 'x-small' ? '12px' : '14px');
}
const SpinnerAnimation = keyframes `
  25% {
    transform: translateX(2.625em) rotate(-90deg) scale(0.5);
    -webkit-transform: translateX(2.625em) rotate(-90deg) scale(0.5);
  } 50% {
    transform: translateX(2.625em) translateY(2.625em) rotate(-179deg);
    -webkit-transform: translateX(2.625em) translateY(2.625em) rotate(-179deg);
  } 50.1% {
    transform: translateX(2.625em) translateY(2.625em) rotate(-180deg);
    -webkit-transform: translateX(2.625em) translateY(2.625em) rotate(-180deg);
  } 75% {
    transform: translateX(0px) translateY(2.625em) rotate(-270deg) scale(0.5);
    -webkit-transform: translateX(0px) translateY(2.625em) rotate(-270deg) scale(0.5);
  } 100% {
    transform: rotate(-360deg);
    -webkit-transform: rotate(-360deg);
  }
`;
const SpinningContainer = styled.div `
  text-align: center;
  visibility: ${props => (props.hidden ? 'hidden' : 'visible')};
`;
const SpinningLabel = styled('label') `
  color: ${themed(props => props.theme.text6)};
  display: ${props => (strToSize(props.size) >= 20 ? 'block' : 'inline-block')};
  margin: ${distance.xsmall};
  color: ${themed(props => props.theme.textDisabled)};
  font-size: ${props => strToFontSize(props.size)};
`;
const Cubes = styled.div `
  width: ${props => strToSize(props.size) * 4.6}px;
  height: ${props => strToSize(props.size) * 4.6}px;
  position: relative;
  font-size: ${props => strToSize(props.size)}px;
  margin: auto;
`;
const Cube0 = styled.div `
  background-color: #141e8c;
  width: 2em;
  height: 2em;
  position: absolute;
  top: 0;
  left: 0;
  animation: ${SpinnerAnimation} 1.8s infinite ease-in-out;
`;
const Cube1 = styled(Cube0) `
  animation-delay: -0.9s;
`;
/**
 * The spinner component displays a loading spinner optionally equipped with a label text.
 */
export const Spinner = Object.assign(((_a) => {
    var { theme, size = 'medium', children } = _a, props = __rest(_a, ["theme", "size", "children"]);
    const spinningProps = { theme, size, hidden: props.hidden };
    return (React.createElement(SpinningContainer, Object.assign({}, props),
        React.createElement(Cubes, Object.assign({}, spinningProps),
            React.createElement(Cube0, null),
            React.createElement(Cube1, null)),
        children && React.createElement(SpinningLabel, Object.assign({}, spinningProps), children)));
}), { inner: {
        get SpinningContainer() { return SpinningContainer; },
        get Cubes() { return Cubes; },
        get Cube0() { return Cube0; },
        get Cube1() { return Cube1; },
        get SpinningLabel() { return SpinningLabel; }
    } });
Spinner.displayName = 'Spinner';
