// precise-ui/dist/es5/components/StackPanel/index.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var styled_1 = require("../../utils/styled");
var StackPanelDirection;
(function (StackPanelDirection) {
    StackPanelDirection["leftToRight"] = "left";
    StackPanelDirection["rightToLeft"] = "right";
    StackPanelDirection["topToBottom"] = "top";
    StackPanelDirection["bottomToTop"] = "bottom";
})(StackPanelDirection = exports.StackPanelDirection || (exports.StackPanelDirection = {}));
var StackLayout = styled_1.default.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  display: flex;\n  flex-direction: ", ";\n  flex-wrap: ", ";\n  justify-content: flex-start;\n"], ["\n  display: flex;\n  flex-direction: ", ";\n  flex-wrap: ", ";\n  justify-content: flex-start;\n"])), function (props) { return props.dir; }, function (props) { return props.wrapping; });
function mapFlex(direction) {
    switch (direction) {
        case StackPanelDirection.rightToLeft:
            return 'row-reverse';
        case StackPanelDirection.topToBottom:
            return 'column';
        case StackPanelDirection.bottomToTop:
            return 'column-reverse';
    }
    return 'row';
}
/**
 * The stack panel displays a stack of children in a given direction.
 */
exports.StackPanel = Object.assign((function (_a) {
    var theme = _a.theme, direction = _a.direction, wrap = _a.wrap, innerRef = _a.innerRef, props = __rest(_a, ["theme", "direction", "wrap", "innerRef"]);
    var wrapping = wrap ? 'wrap' : 'nowrap';
    var dir = mapFlex(direction);
    return React.createElement(StackLayout, __assign({ theme: theme, dir: dir, wrapping: wrapping, ref: innerRef }, props));
}), { inner: {
        get StackLayout() { return StackLayout; }
    } });
exports.StackPanel.displayName = 'StackPanel';
var templateObject_1;
