// precise-ui/dist/es6/components/NavBar/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled from '../../utils/styled';
import { StackPanel } from '../StackPanel';
import { Anchor } from '../Anchor';
import { Icon } from '../Icon';
import { cyan, black, grey2 } from '../../colors';
import { distance } from '../../distance';
import { multiply } from '../../utils/multiply';
import { getFontStyle } from '../../textStyles';
const HeaderStackPanel = styled(StackPanel) `
  align-items: center;
  justify-content: space-between;
  flex-shrink: 0;
`;
const StyledNavBar = styled(HeaderStackPanel) `
  padding-top: ${distance.xlarge};
  margin-bottom: ${distance.xxxlarge};
  flex-wrap: wrap;
`;
const StyledMenuLink = styled.a `
  ${getFontStyle({ size: 'xxLarge' })}

  position: relative;

  cursor: pointer;
  color: ${black};
  text-decoration: none;
  display: flex;
  margin-left: ${distance.medium};
  align-items: center;
  i {
    vertical-align: text-bottom;
  }
  span {
    margin-left: ${distance.small};
  }
  &:hover {
    color: ${cyan};
  }
`;
const StyledMenu = styled.div `
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: right;
  max-height: 0;
  overflow: hidden;
  &.open {
    max-height: 500px;
  }
`;
const StyledMenuItems = styled.ul `
  padding: 0;
  margin: 0;
  list-style-type: none;
  width: 100%;
`;
const StyledMenuItem = styled.li `
  ${getFontStyle({ size: 'medium' })}

  margin: ${distance.small} 0;

  text-align: right;
  & a {
    text-decoration: none;
    color: ${grey2};
  }
  & a:hover {
    color: ${cyan};
  }
`;
const StyledLogoLink = styled(Anchor) `
  margin-left: ${distance.xlarge};
  margin-right: ${distance.medium};
  height: 57px;
`;
const StyledLinks = styled.div `
  margin-right: auto;
  margin-left: ${multiply(distance.xxxlarge, 2)};
`;
const StyledLead = styled.div `
  ${getFontStyle({ size: 'xxLarge' })}

  margin-right: ${distance.medium};
`;
const StyledPin = styled.div `
  ${getFontStyle({ size: 'xxLarge' })}
  
  margin-left: ${distance.medium};
`;
const breakPoint = 780;
function isFullMode(width) {
    return width > breakPoint;
}
const DesktopHeader = Object.assign((({ leads, pins }) => (React.createElement(React.Fragment, null,
    React.createElement(StyledLinks, null, leads.map(item => (React.createElement(StyledLead, { key: item.id }, item.content)))),
    React.createElement(HeaderStackPanel, null, pins.map(item => (React.createElement(StyledPin, { key: item.id }, item.content))))))), { inner: {
        get StyledLinks() { return StyledLinks; },
        get StyledLead() { return StyledLead; },
        get HeaderStackPanel() { return HeaderStackPanel; },
        get StyledPin() { return StyledPin; }
    } });
const MobileHeader = Object.assign((({ entries, onToggle, open }) => (React.createElement(React.Fragment, null,
    React.createElement(HeaderStackPanel, null,
        React.createElement(StyledMenuLink, { onClick: onToggle },
            React.createElement(Icon, { name: open ? 'Close' : 'Menu' }))),
    React.createElement(StyledMenu, { onClick: onToggle, className: open ? 'open' : '' },
        React.createElement(StyledMenuItems, null, entries.map(item => (React.createElement(StyledMenuItem, { key: item.id }, item.content)))))))), { inner: {
        get HeaderStackPanel() { return HeaderStackPanel; },
        get StyledMenuLink() { return StyledMenuLink; },
        get Icon() { return Icon; },
        get StyledMenu() { return StyledMenu; },
        get StyledMenuItems() { return StyledMenuItems; },
        get StyledMenuItem() { return StyledMenuItem; }
    } });
/**
 * A simple navigation bar that is responsive by nature.
 */
export class NavBar extends React.Component {
    constructor(props) {
        super(props);
        this.toggle = () => {
            const { onToggle } = this.props;
            const { open, controlledOpen } = this.state;
            if (!controlledOpen) {
                this.setState({
                    open: !open,
                });
            }
            if (typeof onToggle === 'function') {
                onToggle({
                    shouldOpen: !open,
                });
            }
        };
        this.state = {
            controlledMode: props.mode !== undefined,
            controlledOpen: props.open !== undefined,
            open: !!props.open,
            mode: props.mode,
        };
    }
    UNSAFE_componentWillReceiveProps(nextProps) {
        if (this.state.controlledMode) {
            this.setState({
                mode: nextProps.mode,
            });
        }
        if (this.state.controlledOpen) {
            this.setState({
                open: !!nextProps.open,
            });
        }
    }
    render() {
        const _a = this.props, { logo, hidden, items, mode: _0, children } = _a, rest = __rest(_a, ["logo", "hidden", "items", "mode", "children"]);
        const { mode, open } = this.state;
        return (React.createElement(StyledNavBar, Object.assign({}, rest),
            logo &&
                (typeof logo === 'string' ? (React.createElement(StyledLogoLink, { to: "/" },
                    React.createElement("img", { src: logo, width: 57, height: 57 }))) : (logo)),
            children,
            !hidden &&
                (mode === 'full' ? (React.createElement(DesktopHeader, { leads: items.filter(item => item.display === 'lead'), pins: items.filter(item => item.display === 'pin') })) : mode === 'hamburger' ? (React.createElement(MobileHeader, { entries: items.filter(item => !item.display || item.display === 'mobile'), open: open, onToggle: this.toggle })) : (false))));
    }
}
NavBar.inner = {
    get StyledNavBar() { return StyledNavBar; },
    get StyledLogoLink() { return StyledLogoLink; },
    get DesktopHeader() { return DesktopHeader; },
    get MobileHeader() { return MobileHeader; }
};
