// precise-ui/dist/es6/components/Card/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled from '../../utils/styled';
import { StackPanel, StackPanelDirection } from '../StackPanel';
const StyledCard = styled(StackPanel) `
  position: relative;
  width: 100%;
  height: 100%;
  padding: 0;
  margin: 0;
  box-sizing: border-box;
`;
const StyledCardChildWrapper = styled.div `
  box-sizing: border-box;
  overflow: auto;
  transition: flex-grow 0.4s;
  ${props => props.grow
    ? `
    flex-grow: ${props.grow};
    flex-basis: 0;`
    : ''};
`;
export const Card = Object.assign(((_a) => {
    var { orientation = 'vertical', children, layout: propsLayout } = _a, rest = __rest(_a, ["orientation", "children", "layout"]);
    const layout = propsLayout || React.Children.map(children, _ => 0) || [];
    const cardChildComponents = [];
    React.Children.forEach(children, (child, index) => {
        if (!child) {
            return;
        }
        const _a = child.props, { children: grandChildren } = _a, childProps = __rest(_a, ["children"]);
        const sticky = children && child.type && child.type.displayName === 'Sticky';
        cardChildComponents.push(sticky ? (React.createElement(child.type, Object.assign({ key: index }, childProps), grandChildren)) : (React.createElement(StyledCardChildWrapper, { key: index, grow: layout[index] },
            React.createElement(child.type, Object.assign({}, childProps), grandChildren))));
    });
    return (React.createElement(StyledCard, Object.assign({ direction: orientation === 'vertical' ? StackPanelDirection.topToBottom : StackPanelDirection.leftToRight }, rest), cardChildComponents));
}), { inner: {
        get StyledCardChildWrapper() { return StyledCardChildWrapper; },
        get StyledCard() { return StyledCard; }
    } });
Card.displayName = 'Card';
