// precise-ui/dist/es5/utils/colors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function byte(v) {
    return v < 0 ? 0 : v >= 1 ? 255 : ~~(v * 256);
}
function hsvToRgb(h, s, v, a) {
    if (a === void 0) { a = 1; }
    var i = Math.floor(h * 6);
    var f = h * 6 - i;
    var p = v * (1 - s);
    var q = v * (1 - f * s);
    var t = v * (1 - (1 - f) * s);
    var r = 0;
    var g = 0;
    var b = 0;
    switch (i % 6) {
        case 0:
            (r = v), (g = t), (b = p);
            break;
        case 1:
            (r = q), (g = v), (b = p);
            break;
        case 2:
            (r = p), (g = v), (b = t);
            break;
        case 3:
            (r = p), (g = q), (b = v);
            break;
        case 4:
            (r = t), (g = p), (b = v);
            break;
        case 5:
            (r = v), (g = p), (b = q);
            break;
    }
    return {
        r: byte(r),
        g: byte(g),
        b: byte(b),
        a: a,
    };
}
exports.hsvToRgb = hsvToRgb;
function rgbToHsv(r, g, b, a) {
    if (a === void 0) { a = 1; }
    var max = Math.max(r, g, b);
    var min = Math.min(r, g, b);
    var d = max - min;
    var s = max === 0 ? 0 : d / max;
    var v = max / 255;
    var h = 0;
    switch (max) {
        case min:
            h = 0;
            break;
        case r:
            h = (g - b + d * (g < b ? 6 : 0)) / (6 * d);
            break;
        case g:
            h = (b - r + d * 2) / (6 * d);
            break;
        case b:
            h = (r - g + d * 4) / (6 * d);
            break;
    }
    return {
        h: h,
        s: s,
        v: v,
        a: a,
    };
}
exports.rgbToHsv = rgbToHsv;
function parseColor(color) {
    if (color.match(/^rgb\(/)) {
        var parts = color.substring(4, color.length - 1).split(',');
        return {
            r: +parts[0],
            g: +parts[1],
            b: +parts[2],
            a: 1,
        };
    }
    else if (color.match(/^rgba\(/)) {
        var parts = color.substring(5, color.length - 1).split(',');
        return {
            r: +parts[0],
            g: +parts[1],
            b: +parts[2],
            a: +parts[3],
        };
    }
    else if (color[0] === '#') {
        var value = parseInt(color.substr(1), 16);
        var full = color.length === 7;
        var convert = full
            ? function (v, p) { return (v >> (p * 8)) & 255; }
            : function (v, p) { return ((v >> (p * 4)) & 15) * 17; };
        return {
            r: convert(value, 2),
            g: convert(value, 1),
            b: convert(value, 0),
            a: 1,
        };
    }
    return {
        r: 0,
        g: 0,
        b: 0,
        a: 1,
    };
}
exports.parseColor = parseColor;
function serializeColor(c) {
    return "rgba(" + c.r + ", " + c.g + ", " + c.b + ", " + c.a + ")";
}
exports.serializeColor = serializeColor;
function transparentize(color, alpha) {
    var c = parseColor(color);
    c.a = alpha;
    return serializeColor(c);
}
exports.transparentize = transparentize;
function invertColor(color, blackWhite) {
    if (blackWhite === void 0) { blackWhite = false; }
    var _a = parseColor(color), r = _a.r, g = _a.g, b = _a.b;
    if (!blackWhite) {
        var newR = (255 - r).toString(16);
        var newG = (255 - g).toString(16);
        var newB = (255 - b).toString(16);
        return "rgb(" + newR + ", " + newG + ", " + newB + ")";
    }
    return r * 0.299 + g * 0.587 + b * 0.114 > 186 ? 'rgb(0, 0, 0)' : 'rgb(255, 255, 255)';
}
exports.invertColor = invertColor;
