// precise-ui/dist/es5/utils/displayTo.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var styled_1 = require("./styled");
var common_1 = require("../common");
var themes_1 = require("../themes");
function getMediaQueries(breakpoints) {
    return common_1.ScreenSizeList.map(function (x) { return ({ screen: x, breakpoints: getScreenSizeBreakpoints(x, breakpoints) }); }).reduce(function (acc, next) {
        var _a;
        return (next.breakpoints ? __assign({}, acc, (_a = {}, _a[next.screen] = getWidthBreakpointsQuery(next.breakpoints), _a)) : acc);
    }, {});
}
exports.getMediaQueries = getMediaQueries;
// create max-width media query including screen size
function displayUpTo(screen) {
    var screenBreakpoints = getScreenSizeBreakpoints(screen, themes_1.breakpoints);
    if (!screenBreakpoints) {
        throw new Error('Invalid screen size');
    }
    var query = getWidthBreakpointsQuery({ max: screenBreakpoints.max });
    return getMediaQuery(query);
}
exports.displayUpTo = displayUpTo;
// create min-width media query including screen size
function displayFrom(screen) {
    var screenBreakpoints = getScreenSizeBreakpoints(screen, themes_1.breakpoints);
    if (!screenBreakpoints) {
        throw new Error('Invalid screen size');
    }
    var query = getWidthBreakpointsQuery({ min: screenBreakpoints.min });
    return getMediaQuery(query);
}
exports.displayFrom = displayFrom;
function displayTo(screen) {
    var screenBreakpoints = getScreenSizeBreakpoints(screen, themes_1.breakpoints);
    var query = screenBreakpoints ? getWidthBreakpointsQuery(screenBreakpoints) : screen;
    return getMediaQuery(query);
}
exports.displayTo = displayTo;
// create media query based on previous and next breakpoints
function getWidthBreakpointsQuery(_a) {
    var max = _a.max, min = _a.min;
    if (min === undefined && max === undefined) {
        throw new Error('Both breakpoints cannot be `undefined`');
    }
    if (min !== undefined && max != undefined && max < min) {
        throw new Error("The min(" + min + ") breakpoint must be less than max(" + max + ")");
    }
    var queries = [];
    if (min !== undefined) {
        queries.push("(min-width: " + min + "px)");
    }
    if (max !== undefined) {
        queries.push("(max-width: " + (max - 1) + "px)");
    }
    return "" + queries.join(' and ');
}
exports.getWidthBreakpointsQuery = getWidthBreakpointsQuery;
function getScreenSizeBreakpoints(screen, breakpoints) {
    switch (screen) {
        case 'small':
            return { max: breakpoints.medium };
        case 'medium':
            return { min: breakpoints.medium, max: breakpoints.large };
        case 'large':
            return { min: breakpoints.large, max: breakpoints.xLarge };
        case 'xLarge':
            return { min: breakpoints.xLarge, max: breakpoints.max };
        case 'max':
            return { min: breakpoints.max };
        default:
            return undefined;
    }
}
exports.getScreenSizeBreakpoints = getScreenSizeBreakpoints;
function getMediaQuery(query) {
    return function (strings) {
        var interpolations = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            interpolations[_i - 1] = arguments[_i];
        }
        return styled_1.css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n      @media ", " {\n        ", ";\n      }\n    "], ["\n      @media ", " {\n        ", ";\n      }\n    "])), query, styled_1.css.apply(void 0, [strings].concat(interpolations)));
    };
}
var templateObject_1;
