// precise-ui/dist/es5/utils/eventManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var SimpleEventManager = /** @class */ (function () {
    function SimpleEventManager() {
        this.eventListeners = {};
    }
    SimpleEventManager.prototype.on = function (type, callback) {
        var callbacks = this.eventListeners[type] || [];
        this.eventListeners[type] = callbacks.concat([callback]);
    };
    SimpleEventManager.prototype.off = function (type, callback) {
        var callbacks = this.eventListeners[type] || [];
        this.eventListeners[type] = callbacks.filter(function (cb) { return cb !== callback; });
    };
    SimpleEventManager.prototype.emit = function (type, arg) {
        var _this = this;
        var callbacks = this.eventListeners[type] || [];
        var _loop_1 = function (callback) {
            setTimeout(function () { return callback.call(_this, arg); }, 0);
        };
        for (var _i = 0, callbacks_1 = callbacks; _i < callbacks_1.length; _i++) {
            var callback = callbacks_1[_i];
            _loop_1(callback);
        }
    };
    return SimpleEventManager;
}());
exports.SimpleEventManager = SimpleEventManager;
exports.eventManagers = [new SimpleEventManager()];
function setDefaultEventManager(events) {
    exports.eventManagers.splice(0, 1, events);
}
exports.setDefaultEventManager = setDefaultEventManager;
