// precise-ui/dist/es5/utils/styled.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var styledComponents = require("styled-components");
var themes_1 = require("./../themes");
var _a = styledComponents, styled = _a.default, css = _a.css, keyframes = _a.keyframes, createGlobalStyle = _a.createGlobalStyle, withTheme = _a.withTheme, ThemeProvider = _a.ThemeProvider, ThemeConsumer = _a.ThemeConsumer, ThemeContext = _a.ThemeContext, isStyledComponent = _a.isStyledComponent, ServerStyleSheet = _a.ServerStyleSheet, StyleSheetManager = _a.StyleSheetManager;
exports.css = css;
exports.keyframes = keyframes;
exports.createGlobalStyle = createGlobalStyle;
exports.withTheme = withTheme;
exports.ThemeProvider = ThemeProvider;
exports.ThemeConsumer = ThemeConsumer;
exports.ThemeContext = ThemeContext;
exports.isStyledComponent = isStyledComponent;
exports.ServerStyleSheet = ServerStyleSheet;
exports.StyleSheetManager = StyleSheetManager;
/*
  Helper function which insures that theme is always available in the interpolation callback.
*/
exports.themed = function (interpolation) { return function (props) {
    var _a = props, theme = _a.theme, rest = __rest(_a, ["theme"]);
    var hasNestedProps = function (prop) {
        return typeof prop === 'object' && !Array.isArray(prop) && Object.keys(prop).length > 0;
    };
    var mergedTheme = Object.keys(theme).reduce(function (acc, prop) { return (hasNestedProps(theme[prop]) ? (acc[prop] = __assign({}, acc[prop], theme[prop])) : (acc[prop] = theme[prop]), acc); }, __assign({}, themes_1.light));
    return interpolation(__assign({ theme: mergedTheme }, rest));
}; };
exports.default = styled;
