// precise-ui/dist/es6/components/Tooltip/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled from '../../utils/styled';
import { Flyout } from '../Flyout';
import { tuna, white } from '../../colors';
import { remCalc } from '../../utils/remCalc';
const TooltipContainer = styled.div `
  position: relative;
  width: fit-content;
  display: inline-block;
`;
export class Tooltip extends React.Component {
    constructor(props) {
        super(props);
        this.setTargetRef = (el) => {
            this.targetContainer = el;
        };
        this.onMouseOver = () => {
            if (!this.state.controlled) {
                this.setOpen(true);
            }
        };
        this.onMouseOut = () => {
            if (!this.state.controlled) {
                this.setOpen(false);
            }
        };
        this.state = {
            controlled: props.open !== undefined,
            open: props.open || false,
            targetRect: { top: 0, right: 0, bottom: 0, left: 0, width: 0, height: 0 },
            dirtyFlag: false,
        };
    }
    componentDidMount() {
        this.updateMeasurements();
    }
    UNSAFE_componentWillReceiveProps(nextProps) {
        if (this.state.controlled && nextProps.open !== undefined) {
            this.setOpen(nextProps.open);
        }
    }
    componentDidUpdate() {
        const { dirtyFlag } = this.state;
        if (dirtyFlag) {
            this.updateMeasurements();
            this.setState({
                dirtyFlag: false,
            });
        }
    }
    updateMeasurements() {
        if (this.targetContainer) {
            const targetRect = this.targetContainer.getBoundingClientRect();
            this.setState({
                targetRect,
            });
        }
    }
    setOpen(open) {
        const { onChange } = this.props;
        if (typeof onChange === 'function') {
            onChange({ open });
        }
        this.setState({
            open,
            dirtyFlag: open === true,
        });
    }
    render() {
        const _a = this.props, { open: _0, theme } = _a, props = __rest(_a, ["open", "theme"]);
        const { open } = this.state;
        const tooltipFlyoutTheme = Object.assign({ flyout: { background: tuna, textColor: white, fontSize: remCalc('14px'), maxWidth: '250px' } }, theme);
        return (React.createElement(TooltipContainer, { ref: this.setTargetRef, onFocus: this.onMouseOver, onBlur: this.onMouseOut, onMouseOver: this.onMouseOver, onMouseOut: this.onMouseOut },
            React.createElement(Flyout, Object.assign({}, props, { open: open, theme: tooltipFlyoutTheme }))));
    }
}
Tooltip.inner = {
    get TooltipContainer() { return TooltipContainer; },
    get Flyout() { return Flyout; }
};
