// precise-ui/dist/es5/components/NavBar/index.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var styled_1 = require("../../utils/styled");
var StackPanel_1 = require("../StackPanel");
var Anchor_1 = require("../Anchor");
var Icon_1 = require("../Icon");
var colors_1 = require("../../colors");
var distance_1 = require("../../distance");
var multiply_1 = require("../../utils/multiply");
var textStyles_1 = require("../../textStyles");
var HeaderStackPanel = styled_1.default(StackPanel_1.StackPanel)(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  align-items: center;\n  justify-content: space-between;\n  flex-shrink: 0;\n"], ["\n  align-items: center;\n  justify-content: space-between;\n  flex-shrink: 0;\n"])));
var StyledNavBar = styled_1.default(HeaderStackPanel)(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  padding-top: ", ";\n  margin-bottom: ", ";\n  flex-wrap: wrap;\n"], ["\n  padding-top: ", ";\n  margin-bottom: ", ";\n  flex-wrap: wrap;\n"])), distance_1.distance.xlarge, distance_1.distance.xxxlarge);
var StyledMenuLink = styled_1.default.a(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  ", "\n\n  position: relative;\n\n  cursor: pointer;\n  color: ", ";\n  text-decoration: none;\n  display: flex;\n  margin-left: ", ";\n  align-items: center;\n  i {\n    vertical-align: text-bottom;\n  }\n  span {\n    margin-left: ", ";\n  }\n  &:hover {\n    color: ", ";\n  }\n"], ["\n  ", "\n\n  position: relative;\n\n  cursor: pointer;\n  color: ", ";\n  text-decoration: none;\n  display: flex;\n  margin-left: ", ";\n  align-items: center;\n  i {\n    vertical-align: text-bottom;\n  }\n  span {\n    margin-left: ", ";\n  }\n  &:hover {\n    color: ", ";\n  }\n"])), textStyles_1.getFontStyle({ size: 'xxLarge' }), colors_1.black, distance_1.distance.medium, distance_1.distance.small, colors_1.cyan);
var StyledMenu = styled_1.default.div(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  width: 100%;\n  display: flex;\n  align-items: center;\n  justify-content: right;\n  max-height: 0;\n  overflow: hidden;\n  &.open {\n    max-height: 500px;\n  }\n"], ["\n  width: 100%;\n  display: flex;\n  align-items: center;\n  justify-content: right;\n  max-height: 0;\n  overflow: hidden;\n  &.open {\n    max-height: 500px;\n  }\n"])));
var StyledMenuItems = styled_1.default.ul(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  padding: 0;\n  margin: 0;\n  list-style-type: none;\n  width: 100%;\n"], ["\n  padding: 0;\n  margin: 0;\n  list-style-type: none;\n  width: 100%;\n"])));
var StyledMenuItem = styled_1.default.li(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  ", "\n\n  margin: ", " 0;\n\n  text-align: right;\n  & a {\n    text-decoration: none;\n    color: ", ";\n  }\n  & a:hover {\n    color: ", ";\n  }\n"], ["\n  ", "\n\n  margin: ", " 0;\n\n  text-align: right;\n  & a {\n    text-decoration: none;\n    color: ", ";\n  }\n  & a:hover {\n    color: ", ";\n  }\n"])), textStyles_1.getFontStyle({ size: 'medium' }), distance_1.distance.small, colors_1.grey2, colors_1.cyan);
var StyledLogoLink = styled_1.default(Anchor_1.Anchor)(templateObject_7 || (templateObject_7 = __makeTemplateObject(["\n  margin-left: ", ";\n  margin-right: ", ";\n  height: 57px;\n"], ["\n  margin-left: ", ";\n  margin-right: ", ";\n  height: 57px;\n"])), distance_1.distance.xlarge, distance_1.distance.medium);
var StyledLinks = styled_1.default.div(templateObject_8 || (templateObject_8 = __makeTemplateObject(["\n  margin-right: auto;\n  margin-left: ", ";\n"], ["\n  margin-right: auto;\n  margin-left: ", ";\n"])), multiply_1.multiply(distance_1.distance.xxxlarge, 2));
var StyledLead = styled_1.default.div(templateObject_9 || (templateObject_9 = __makeTemplateObject(["\n  ", "\n\n  margin-right: ", ";\n"], ["\n  ", "\n\n  margin-right: ", ";\n"])), textStyles_1.getFontStyle({ size: 'xxLarge' }), distance_1.distance.medium);
var StyledPin = styled_1.default.div(templateObject_10 || (templateObject_10 = __makeTemplateObject(["\n  ", "\n  \n  margin-left: ", ";\n"], ["\n  ", "\n  \n  margin-left: ", ";\n"])), textStyles_1.getFontStyle({ size: 'xxLarge' }), distance_1.distance.medium);
var breakPoint = 780;
function isFullMode(width) {
    return width > breakPoint;
}
var DesktopHeader = Object.assign((function (_a) {
    var leads = _a.leads, pins = _a.pins;
    return (React.createElement(React.Fragment, null,
        React.createElement(StyledLinks, null, leads.map(function (item) { return (React.createElement(StyledLead, { key: item.id }, item.content)); })),
        React.createElement(HeaderStackPanel, null, pins.map(function (item) { return (React.createElement(StyledPin, { key: item.id }, item.content)); }))));
}), { inner: {
        get StyledLinks() { return StyledLinks; },
        get StyledLead() { return StyledLead; },
        get HeaderStackPanel() { return HeaderStackPanel; },
        get StyledPin() { return StyledPin; }
    } });
var MobileHeader = Object.assign((function (_a) {
    var entries = _a.entries, onToggle = _a.onToggle, open = _a.open;
    return (React.createElement(React.Fragment, null,
        React.createElement(HeaderStackPanel, null,
            React.createElement(StyledMenuLink, { onClick: onToggle },
                React.createElement(Icon_1.Icon, { name: open ? 'Close' : 'Menu' }))),
        React.createElement(StyledMenu, { onClick: onToggle, className: open ? 'open' : '' },
            React.createElement(StyledMenuItems, null, entries.map(function (item) { return (React.createElement(StyledMenuItem, { key: item.id }, item.content)); })))));
}), { inner: {
        get HeaderStackPanel() { return HeaderStackPanel; },
        get StyledMenuLink() { return StyledMenuLink; },
        get Icon() { return Icon_1.Icon; },
        get StyledMenu() { return StyledMenu; },
        get StyledMenuItems() { return StyledMenuItems; },
        get StyledMenuItem() { return StyledMenuItem; }
    } });
/**
 * A simple navigation bar that is responsive by nature.
 */
var NavBar = /** @class */ (function (_super) {
    __extends(NavBar, _super);
    function NavBar(props) {
        var _this = _super.call(this, props) || this;
        _this.toggle = function () {
            var onToggle = _this.props.onToggle;
            var _a = _this.state, open = _a.open, controlledOpen = _a.controlledOpen;
            if (!controlledOpen) {
                _this.setState({
                    open: !open,
                });
            }
            if (typeof onToggle === 'function') {
                onToggle({
                    shouldOpen: !open,
                });
            }
        };
        _this.state = {
            controlledMode: props.mode !== undefined,
            controlledOpen: props.open !== undefined,
            open: !!props.open,
            mode: props.mode,
        };
        return _this;
    }
    NavBar.prototype.UNSAFE_componentWillReceiveProps = function (nextProps) {
        if (this.state.controlledMode) {
            this.setState({
                mode: nextProps.mode,
            });
        }
        if (this.state.controlledOpen) {
            this.setState({
                open: !!nextProps.open,
            });
        }
    };
    NavBar.prototype.render = function () {
        var _a = this.props, logo = _a.logo, hidden = _a.hidden, items = _a.items, _0 = _a.mode, children = _a.children, rest = __rest(_a, ["logo", "hidden", "items", "mode", "children"]);
        var _b = this.state, mode = _b.mode, open = _b.open;
        return (React.createElement(StyledNavBar, __assign({}, rest),
            logo &&
                (typeof logo === 'string' ? (React.createElement(StyledLogoLink, { to: "/" },
                    React.createElement("img", { src: logo, width: 57, height: 57 }))) : (logo)),
            children,
            !hidden &&
                (mode === 'full' ? (React.createElement(DesktopHeader, { leads: items.filter(function (item) { return item.display === 'lead'; }), pins: items.filter(function (item) { return item.display === 'pin'; }) })) : mode === 'hamburger' ? (React.createElement(MobileHeader, { entries: items.filter(function (item) { return !item.display || item.display === 'mobile'; }), open: open, onToggle: this.toggle })) : (false))));
    };
    NavBar.inner = {
        get StyledNavBar() { return StyledNavBar; },
        get StyledLogoLink() { return StyledLogoLink; },
        get DesktopHeader() { return DesktopHeader; },
        get MobileHeader() { return MobileHeader; }
    };
    return NavBar;
}(React.Component));
exports.NavBar = NavBar;
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8, templateObject_9, templateObject_10;
