// precise-ui/dist/es6/components/WindowPopup/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled, { themed } from '../../utils/styled';
import { remCalc } from '../../utils/remCalc';
import { Blocker } from '../Blocker';
import { CloseButton } from '../CloseButton';
import { distance } from '../../distance';
import { getFontStyle } from '../../textStyles';
const StyledMobileDropdownWrapper = styled.div `
  padding: ${distance.medium};
  box-sizing: border-box;
  height: 100%;
  overflow-y: auto;
  background: ${themed(props => props.theme.ui1)};
`;
const StyledMobileWrapper = styled.div `
  box-sizing: border-box;
  margin: 0;
`;
const StyledLabel = styled.div `
  ${getFontStyle({ size: 'small' })}

  color: ${themed(({ theme }) => theme.ui0)};
  padding-bottom: ${distance.medium};
  min-height: ${remCalc('14px')};
`;
export class WindowPopup extends React.Component {
    render() {
        const _a = this.props, { children, label } = _a, props = __rest(_a, ["children", "label"]);
        return (React.createElement(Blocker, Object.assign({}, props),
            React.createElement(StyledMobileDropdownWrapper, { theme: props.theme, tabIndex: 0 },
                React.createElement(StyledLabel, null, label),
                React.createElement(CloseButton, { onClick: props.onClose }),
                React.createElement(StyledMobileWrapper, null, children))));
    }
}
WindowPopup.inner = {
    get Blocker() { return Blocker; },
    get StyledMobileDropdownWrapper() { return StyledMobileDropdownWrapper; },
    get StyledLabel() { return StyledLabel; },
    get CloseButton() { return CloseButton; },
    get StyledMobileWrapper() { return StyledMobileWrapper; }
};
