// @steveeeie/react-page-transition/dist/index.cjs.js
'use strict';Object.defineProperty(exports,"__esModule",{value:!0});function _interopDefault(a){return a&&"object"==typeof a&&"default"in a?a["default"]:a}var React=require("react"),React__default=_interopDefault(React),reactTransitionGroup=require("react-transition-group"),styled=require("styled-components"),styled__default=_interopDefault(styled);const animations={moveToLeft:{keyframes:styled.keyframes`
      from { }
      to { transform: translateX(-100%) rotateZ(0.01deg); }
    `,duration:600,timing:"ease",fill:"both"},moveFromLeft:{keyframes:styled.keyframes`
      from {  transform: translateX(-100%) rotateZ(0.01deg); }
    `,duration:600,timing:"ease",fill:"both"},moveToRight:{keyframes:styled.keyframes`
      from { }
      to { transform: translateX(100%) rotateZ(0.01deg); }
    `,duration:600,timing:"ease",fill:"both"},moveFromRight:{keyframes:styled.keyframes`
      from { transform: translateX(100%) rotateZ(0.01deg); }
    `,duration:600,timing:"ease",fill:"both"},moveToTop:{keyframes:styled.keyframes`
      from { }
      to { transform: translateY(-100%) rotateZ(0.01deg); }
    `,duration:600,timing:"ease",fill:"both"},moveFromTop:{keyframes:styled.keyframes`
      from { transform: translateY(-100%) rotateZ(0.01deg); }
    `,duration:600,timing:"ease",fill:"both"},moveToBottom:{keyframes:styled.keyframes`
      from { }
      to { transform: translateY(100%) rotateZ(0.01deg); }
    `,duration:600,timing:"ease",fill:"both"},moveFromBottom:{keyframes:styled.keyframes`
      from { transform: translateY(100%) rotateZ(0.01deg); }
    `,duration:600,timing:"ease",fill:"both"},fade:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0.3; }
    `,duration:600,timing:"ease",fill:"both"},moveToLeftFade:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0.3; transform: translateX(-100%) rotateZ(0.01deg); }
    `,duration:700,timing:"ease",fill:"both"},moveFromLeftFade:{keyframes:styled.keyframes`
      from { opacity: 0.3;  transform: translateX(-100%) rotateZ(0.01deg); }
    `,duration:700,timing:"ease",fill:"both"},moveToRightFade:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0.3;  transform: translateX(100%) rotateZ(0.01deg); }
    `,duration:700,timing:"ease",fill:"both"},moveFromRightFade:{keyframes:styled.keyframes`
      from { opacity: 0.3; transform: translateX(100%) rotateZ(0.01deg); }
    `,duration:700,timing:"ease",fill:"both"},moveToTopFade:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0.3; transform: translateY(-100%) rotateZ(0.01deg); }
    `,duration:600,timing:"ease",fill:"both"},moveFromTopFade:{keyframes:styled.keyframes`
      from { opacity: 0.3; transform: translateY(-100%) rotateZ(0.01deg); }
    `,duration:700,timing:"ease",fill:"both"},moveToBottomFade:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0.3;  transform: translateY(100%) rotateZ(0.01deg); }
    `,duration:700,timing:"ease",fill:"both"},moveFromBottomFade:{keyframes:styled.keyframes`
      from { opacity: 0.3; transform: translateY(100%) rotateZ(0.01deg); }
    `,duration:700,timing:"ease",fill:"both"},scaleDown:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0; transform: scale(0.8); }
    `,duration:700,timing:"ease",fill:"both"},scaleUp:{keyframes:styled.keyframes`
	    from { opacity: 0;  transform: scale(0.8); }
    `,duration:700,timing:"ease",fill:"both"},scaleUpDown:{keyframes:styled.keyframes`
      from { opacity: 0; transform: scale(1.2); }
    `,duration:500,timing:"ease",fill:"both"},scaleDownUp:{keyframes:styled.keyframes`
	    from { }
	    to { opacity: 0; transform: scale(1.2); }
    `,duration:500,timing:"ease",fill:"both"},scaleDownCenter:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0; transform: scale(0.7); }
    `,duration:400,timing:"ease",fill:"both"},scaleUpCenter:{keyframes:styled.keyframes`
      from { opacity: 0;  transform: scale(0.7); }
    `,duration:400,timing:"ease",fill:"both"},rotateRightSideFirst:{keyframes:styled.keyframes`
      0% { }
      40% { transform: rotateY(15deg); opacity: 0.8; animation-timing-function: ease-out; }
      100% { transform: scale(0.8) translateZ(-200px); opacity:0; }
    `,duration:800,timing:"ease-in",fill:"both",origin:"0% 50%"},rotateLeftSideFirst:{keyframes:styled.keyframes`
      0% { }
      40% { transform: rotateY(-15deg); opacity: 0.8; animation-timing-function: ease-out; }
      100% { transform: scale(0.8) translateZ(-200px); opacity:0; }
    `,duration:800,timing:"ease-in",fill:"both",origin:"0% 50%"},rotateTopSideFirst:{keyframes:styled.keyframes`
      0% { }
      40% { transform: rotateX(15deg); opacity: 0.8; animation-timing-function: ease-out; }
      100% { transform: scale(0.8) translateZ(-200px); opacity:0; }
    `,duration:800,timing:"ease-in",fill:"both",origin:"0% 50%"},rotateBottomSideFirst:{keyframes:styled.keyframes`
      0% { }
      40% { transform: rotateX(-15deg); opacity: 0.8; animation-timing-function: ease-out; }
      100% {transform: scale(0.8) translateZ(-200px); opacity:0; }
    `,duration:800,timing:"ease-in",fill:"both",origin:"0% 50%"},flipOutRight:{keyframes:styled.keyframes`
      from { }
      to { transform: translateZ(-1000px) rotateY(90deg); opacity: 0.2; }
    `,duration:500,timing:"ease-in",fill:"both",origin:"50% 50%"},flipInLeft:{keyframes:styled.keyframes`
      from { transform: translateZ(-1000px) rotateY(-90deg); opacity: 0.2; }
    `,duration:500,timing:"ease-out",fill:"both",origin:"50% 50%"},flipOutLeft:{keyframes:styled.keyframes`
      from { }
      to { transform: translateZ(-1000px) rotateY(-90deg); opacity: 0.2; }
    `,duration:500,timing:"ease-in",fill:"both",origin:"50% 50%"},flipInRight:{keyframes:styled.keyframes`
      from { transform: translateZ(-1000px) rotateY(90deg); opacity: 0.2; }
    `,duration:500,timing:"ease-out",fill:"both",origin:"50% 50%"},flipOutTop:{keyframes:styled.keyframes`
      from { }
      to { transform: translateZ(-1000px) rotateX(90deg); opacity: 0.2; }
    `,duration:500,timing:"ease-in",fill:"both",origin:"50% 50%"},flipInBottom:{keyframes:styled.keyframes`
      from { transform: translateZ(-1000px) rotateX(-90deg); opacity: 0.2; }
    `,duration:500,timing:"ease-out",fill:"both",origin:"50% 50%"},flipOutBottom:{keyframes:styled.keyframes`
      from { }
      to { transform: translateZ(-1000px) rotateX(-90deg); opacity: 0.2; }
    `,duration:500,timing:"ease-in",fill:"both",origin:"50% 50%"},flipInTop:{keyframes:styled.keyframes`
      from { transform: translateZ(-1000px) rotateX(90deg); opacity: 0.2; }
    `,duration:500,timing:"ease-out",fill:"both",origin:"50% 50%"},rotateFall:{keyframes:styled.keyframes`
      0% { transform: rotateZ(0deg); }
      20% { transform: rotateZ(10deg); animation-timing-function: ease-out; }
      40% { transform: rotateZ(17deg); }
      60% { transform: rotateZ(16deg); }
      100% { transform: translateY(100%) rotateZ(17deg); }
    `,duration:1e3,timing:"ease-in",fill:"both",origin:"0% 0%"},rotateOutNewspaper:{keyframes:styled.keyframes`
      from { }
      to { transform: translateZ(-3000px) rotateZ(360deg); opacity: 0; }
    `,duration:500,timing:"ease-in",fill:"both",origin:"50% 50%"},rotateInNewspaper:{keyframes:styled.keyframes`
      from { transform: translateZ(-3000px) rotateZ(-360deg); opacity: 0; }
    `,duration:500,timing:"ease-out",fill:"both",origin:"50% 50%"},rotatePushLeft:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0; transform: rotateY(90deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"0% 50%"},rotatePushRight:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0; transform: rotateY(-90deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"100% 50%"},rotatePushTop:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0; transform: rotateX(-90deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"50% 0%"},rotatePushBottom:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0; transform: rotateX(90deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"50% 100%"},rotatePullRight:{keyframes:styled.keyframes`
      from { opacity: 0; transform: rotateY(-90deg); }
    `,duration:500,timing:"ease",fill:"both",origin:"100% 50%"},rotatePullLeft:{keyframes:styled.keyframes`
      from { opacity: 0; transform: rotateY(90deg); }
    `,duration:500,timing:"ease",fill:"both",origin:"0% 50%"},rotatePullTop:{keyframes:styled.keyframes`
      from { opacity: 0; transform: rotateX(-90deg); }
    `,duration:500,timing:"ease",fill:"both",origin:"50% 0%"},rotatePullBottom:{keyframes:styled.keyframes`
      from { opacity: 0; transform: rotateX(90deg); }
    `,duration:500,timing:"ease",fill:"both",origin:"50% 100%"},rotateFoldRight:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0; transform: translateX(100%) rotateY(90deg); }
    `,duration:700,timing:"ease",fill:"both",origin:"0% 50%"},rotateFoldLeft:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0; transform: translateX(-100%) rotateY(-90deg); }
    `,duration:700,timing:"ease",fill:"both",origin:"100% 50%"},rotateFoldTop:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0; transform: translateY(-100%) rotateX(90deg); }
    `,duration:700,timing:"ease",fill:"both",origin:"50% 100%"},rotateFoldBottom:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0; transform: translateY(100%) rotateX(-90deg); }
    `,duration:700,timing:"ease",fill:"both",origin:"50% 0%"},rotateUnfoldLeft:{keyframes:styled.keyframes`
      from { opacity: 0; transform: translateX(-100%) rotateY(-90deg); }
    `,duration:700,timing:"ease",fill:"both",origin:"100% 50%"},rotateUnfoldRight:{keyframes:styled.keyframes`
      from { opacity: 0;  transform: translateX(100%) rotateY(90deg); }
    `,duration:700,timing:"ease",fill:"both",origin:"0% 50%"},rotateUnfoldTop:{keyframes:styled.keyframes`
      from { opacity: 0; transform: translateY(-100%) rotateX(90deg); }
    `,duration:700,timing:"ease",fill:"both",origin:"50% 100%"},rotateUnfoldBottom:{keyframes:styled.keyframes`
      from { opacity: 0; transform: translateY(100%) rotateX(-90deg); }
    `,duration:700,timing:"ease",fill:"both",origin:"50% 0%"},rotateRoomLeftOut:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0.3; transform: translateX(-100%) rotateY(90deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"100% 50%"},rotateRoomLeftIn:{keyframes:styled.keyframes`
      from { opacity: 0.3; transform: translateX(100%) rotateY(-90deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"0% 50%"},rotateRoomRightOut:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0.3; transform: translateX(100%) rotateY(-90deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"0% 50%"},rotateRoomRightIn:{keyframes:styled.keyframes`
      from { opacity: 0.3; transform: translateX(-100%) rotateY(90deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"100% 50%"},rotateRoomTopOut:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0.3; transform: translateY(-100%) rotateX(-90deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"50% 100%"},rotateRoomTopIn:{keyframes:styled.keyframes`
      from { opacity: 0.3; transform: translateY(100%) rotateX(90deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"50% 0%"},rotateRoomBottomOut:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0.3; transform: translateY(100%) rotateX(90deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"50% 0%"},rotateRoomBottomIn:{keyframes:styled.keyframes`
      from { opacity: 0.3; transform: translateY(-100%) rotateX(-90deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"50% 100%"},rotateCubeLeftOut:{keyframes:styled.keyframes`
      0% { }
      50% { animation-timing-function: ease-out; transform: translateX(-50%) translateZ(-200px) rotateY(-45deg); }
      100% { opacity: 0.3; transform: translateX(-100%) rotateY(-90deg); }
    `,duration:600,timing:"ease-in",fill:"both",origin:"100% 50%"},rotateCubeLeftIn:{keyframes:styled.keyframes`
      0% { opacity: 0.3; transform: translateX(100%) rotateY(90deg); }
      50% { animation-timing-function: ease-out; transform: translateX(50%) translateZ(-200px) rotateY(45deg); }
    `,duration:600,timing:"ease-in",fill:"both",origin:"0% 50%"},rotateCubeRightOut:{keyframes:styled.keyframes`
      0% { }
      50% { animation-timing-function: ease-out; transform: translateX(50%) translateZ(-200px) rotateY(45deg); }
      100% { opacity: 0.3; transform: translateX(100%) rotateY(90deg); }
    `,duration:600,timing:"ease-in",fill:"both",origin:"0% 50%"},rotateCubeRightIn:{keyframes:styled.keyframes`
      0% { opacity: 0.3; transform: translateX(-100%) rotateY(-90deg); }
      50% { animation-timing-function: ease-out; transform: translateX(-50%) translateZ(-200px) rotateY(-45deg); }
    `,duration:600,timing:"ease-in",fill:"both",origin:"100% 50%"},rotateCubeTopOut:{keyframes:styled.keyframes`
      0% {}
      50% { animation-timing-function: ease-out; transform: translateY(-50%) translateZ(-200px) rotateX(45deg); }
      100% { opacity: 0.3; transform: translateY(-100%) rotateX(90deg); }
    `,duration:600,timing:"ease-in",fill:"both",origin:"50% 100%"},rotateCubeTopIn:{keyframes:styled.keyframes`
      0% { opacity: 0.3; transform: translateY(100%) rotateX(-90deg); }
      50% { animation-timing-function: ease-out; transform: translateY(50%) translateZ(-200px) rotateX(-45deg); }
    `,duration:600,timing:"ease-in",fill:"both",origin:"50% 0%"},rotateCubeBottomOut:{keyframes:styled.keyframes`
      0% { }
      50% { animation-timing-function: ease-out; transform: translateY(50%) translateZ(-200px) rotateX(-45deg); }
      100% { opacity: 0.3; -webkit-transform: translateY(100%) rotateX(-90deg); transform: translateY(100%) rotateX(-90deg); }
    `,duration:600,timing:"ease-in",fill:"both",origin:"50% 0%"},rotateCubeBottomIn:{keyframes:styled.keyframes`
      0% { opacity: 0.3; -webkit-transform: translateY(-100%) rotateX(90deg); transform: translateY(-100%) rotateX(90deg); }
      50% { animation-timing-function: ease-out; transform: translateY(-50%) translateZ(-200px) rotateX(45deg); }
    `,duration:600,timing:"ease-in",fill:"both",origin:"50% 100%"},rotateCarouselLeftOut:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0.3; transform: translateX(-150%) scale(0.4) rotateY(-65deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"100% 50%"},rotateCarouselLeftIn:{keyframes:styled.keyframes`
      from { opacity: 0.3; transform: translateX(200%) scale(0.4) rotateY(65deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"0% 50%"},rotateCarouselRightOut:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0.3; transform: translateX(200%) scale(0.4) rotateY(65deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"0% 50%"},rotateCarouselRightIn:{keyframes:styled.keyframes`
      from { opacity: 0.3; transform: translateX(-200%) scale(0.4) rotateY(-65deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"100% 50%"},rotateCarouselTopOut:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0.3; transform: translateY(-200%) scale(0.4) rotateX(65deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"50% 100%"},rotateCarouselTopIn:{keyframes:styled.keyframes`
      from { opacity: 0.3; transform: translateY(200%) scale(0.4) rotateX(-65deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"50% 0%"},rotateCarouselBottomOut:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0.3; transform: translateY(200%) scale(0.4) rotateX(-65deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"50% 0%"},rotateCarouselBottomIn:{keyframes:styled.keyframes`
      from { opacity: 0.3; transform: translateY(-200%) scale(0.4) rotateX(65deg); }
    `,duration:800,timing:"ease",fill:"both",origin:"50% 100%"},rotateSidesOut:{keyframes:styled.keyframes`
      from { }
      to { opacity: 0; transform: translateZ(-500px) rotateY(90deg); }
    `,duration:500,timing:"ease-in",fill:"both",origin:"-50% 50%"},rotateSidesIn:{keyframes:styled.keyframes`
      from { opacity: 0; transform: translateZ(-500px) rotateY(-90deg); }
    `,duration:500,timing:"ease-in",fill:"both",origin:"150% 50%"},rotateSlideOut:{keyframes:styled.keyframes`
      0% { }
      25% { opacity: 0.5; transform: translateZ(-500px); }
      75% { opacity: 0.5; transform: translateZ(-500px) translateX(-200%); }
      100% { opacity: 0.5; transform: translateZ(-500px) translateX(-200%); }
    `,duration:1e3,timing:"ease",fill:"both"},rotateSlideIn:{keyframes:styled.keyframes`
      0%, 25% { opacity: 0.5; transform: translateZ(-500px) translateX(200%); }
      75% { opacity: 0.5; transform: translateZ(-500px); }
      100% { opacity: 1; transform: translateZ(0) translateX(0); }
    `,duration:1e3,timing:"ease",fill:"both"}},presets={moveToLeftFromRight:{exit:{name:"moveToLeft"},enter:{name:"moveFromRight"}},moveToRightFromLeft:{exit:{name:"moveToRight"},enter:{name:"moveFromLeft"}},moveToTopFromBottom:{exit:{name:"moveToTop"},enter:{name:"moveFromBottom"}},moveToBottomFromTop:{exit:{name:"moveToBottom"},enter:{name:"moveFromTop"}},fadeFromRight:{exit:{name:"fade"},enter:{name:"moveFromRight",onTop:!0}},fadeFromLeft:{exit:{name:"fade"},enter:{name:"moveFromLeft",onTop:!0}},fadeFromBottom:{exit:{name:"fade"},enter:{name:"moveFromBottom",onTop:!0}},fadeFromTop:{exit:{name:"fade"},enter:{name:"moveFromTop",onTop:!0}},fadeLeftFadeRight:{exit:{name:"moveToLeftFade"},enter:{name:"moveFromRightFade"}},fadeRightFadeLeft:{exit:{name:"moveToRightFade"},enter:{name:"moveFromLeftFade"}},fadeTopFadeBottom:{exit:{name:"moveToTopFade"},enter:{name:"moveFromBottomFade"}},fadeBottomFadeTop:{exit:{name:"moveToBottomFade"},enter:{name:"moveFromTopFade"}},scaleDownFromRight:{exit:{name:"scaleDown"},enter:{name:"moveFromRight",onTop:!0}},scaleDownFromLeft:{exit:{name:"scaleDown"},enter:{name:"moveFromLeft",onTop:!0}},scaleDownFromBottom:{exit:{name:"scaleDown"},enter:{name:"moveFromBottom",onTop:!0}},scaleDownFromTop:{exit:{name:"scaleDown"},enter:{name:"moveFromTop",onTop:!0}},scaleDownScaleDown:{exit:{name:"scaleDown"},enter:{name:"scaleUpDown",delay:300}},scaleUpScaleUp:{exit:{name:"scaleDownUp"},enter:{name:"scaleUp",delay:300}},moveToLeftScaleUp:{exit:{name:"moveToLeft",onTop:!0},enter:{name:"scaleUp"}},moveToRightScaleUp:{exit:{name:"moveToRight",onTop:!0},enter:{name:"scaleUp"}},moveToTopScaleUp:{exit:{name:"moveToTop",onTop:!0},enter:{name:"scaleUp"}},moveToBottomScaleUp:{exit:{name:"moveToBottom",onTop:!0},enter:{name:"scaleUp"}},scaleDownScaleUp:{exit:{name:"scaleDownCenter"},enter:{name:"scaleUpCenter",delay:400}},glueLeftFromRight:{exit:{name:"rotateRightSideFirst"},enter:{name:"moveFromRight",delay:200,onTop:!0}},glueRightFromLeft:{exit:{name:"rotateLeftSideFirst"},enter:{name:"moveFromLeft",delay:200,onTop:!0}},glueBottomFromTop:{exit:{name:"rotateTopSideFirst"},enter:{name:"moveFromTop",delay:200,onTop:!0}},glueTopFromBottom:{exit:{name:"rotateBottomSideFirst"},enter:{name:"moveFromBottom",delay:200,onTop:!0}},flipRight:{exit:{name:"flipOutRight"},enter:{name:"flipInLeft",delay:500}},flipLeft:{exit:{name:"flipOutLeft"},enter:{name:"flipInRight",delay:500}},flipTop:{exit:{name:"flipOutTop"},enter:{name:"flipInBottom",delay:500}},flipBottom:{exit:{name:"flipOutBottom"},enter:{name:"flipInTop",delay:500}},fall:{exit:{name:"rotateFall",onTop:!0},enter:{name:"scaleUp"}},newspaper:{exit:{name:"rotateOutNewspaper"},enter:{name:"rotateInNewspaper",delay:500}},pushLeftFromRight:{exit:{name:"rotatePushLeft"},enter:{name:"moveFromRight"}},pushRightFromLeft:{exit:{name:"rotatePushRight"},enter:{name:"moveFromLeft"}},pushTopFromBottom:{exit:{name:"rotatePushTop"},enter:{name:"moveFromBottom"}},pushBottomFromTop:{exit:{name:"rotatePushBottom"},enter:{name:"moveFromTop"}},pushLeftPullRight:{exit:{name:"rotatePushLeft"},enter:{name:"rotatePullRight",delay:180}},pushRightPullLeft:{exit:{name:"rotatePushRight"},enter:{name:"rotatePullLeft",delay:180}},pushTopPullBottom:{exit:{name:"rotatePushTop"},enter:{name:"rotatePullBottom",delay:180}},pushBottomPullTop:{exit:{name:"rotatePushBottom"},enter:{name:"rotatePullTop",delay:180}},foldLeftFromRight:{exit:{name:"rotateFoldLeft"},enter:{name:"moveFromRightFade"}},foldRightFromLeft:{exit:{name:"rotateFoldRight"},enter:{name:"moveFromLeftFade"}},foldTopFromBottom:{exit:{name:"rotateFoldTop"},enter:{name:"moveFromBottomFade"}},foldBottomFromTop:{exit:{name:"rotateFoldBottom"},enter:{name:"moveFromTopFade"}},moveToRightUnfoldLeft:{exit:{name:"moveToRightFade"},enter:{name:"rotateUnfoldLeft"}},moveToLeftUnfoldRight:{exit:{name:"moveToLeftFade"},enter:{name:"rotateUnfoldRight"}},moveToBottomUnfoldTop:{exit:{name:"moveToBottomFade"},enter:{name:"rotateUnfoldTop"}},moveToTopUnfoldBottom:{exit:{name:"moveToTopFade"},enter:{name:"rotateUnfoldBottom"}},roomToLeft:{exit:{name:"rotateRoomLeftOut",onTop:!0},enter:{name:"rotateRoomLeftIn"}},roomToRight:{exit:{name:"rotateRoomRightOut",onTop:!0},enter:{name:"rotateRoomRightIn"}},roomToTop:{exit:{name:"rotateRoomTopOut",onTop:!0},enter:{name:"rotateRoomTopIn"}},roomToBottom:{exit:{name:"rotateRoomBottomOut",onTop:!0},enter:{name:"rotateRoomBottomIn"}},cubeToLeft:{exit:{name:"rotateCubeLeftOut",onTop:!0},enter:{name:"rotateCubeLeftIn"}},cubeToRight:{exit:{name:"rotateCubeRightOut",onTop:!0},enter:{name:"rotateCubeRightIn"}},cubeToTop:{exit:{name:"rotateCubeTopOut",onTop:!0},enter:{name:"rotateCubeTopIn"}},cubeToBottom:{exit:{name:"rotateCubeBottomOut",onTop:!0},enter:{name:"rotateCubeBottomIn"}},carouselToLeft:{exit:{name:"rotateCarouselLeftOut",onTop:!0},enter:{name:"rotateCarouselLeftIn"}},carouselToRight:{exit:{name:"rotateCarouselRightOut",onTop:!0},enter:{name:"rotateCarouselRightIn"}},carouselToTop:{exit:{name:"rotateCarouselTopOut",onTop:!0},enter:{name:"rotateCarouselTopIn"}},carouselToBottom:{exit:{name:"rotateCarouselBottomOut",onTop:!0},enter:{name:"rotateCarouselBottomIn"}},slides:{exit:{name:"rotateSidesOut"},enter:{name:"rotateSidesIn",delay:200}},slide:{exit:{name:"rotateSlideOut"},enter:{name:"rotateSlideIn"}}},createAnimationStyles=({keyframes:a,delay:b,duration:c,timing:d,fill:e,origin:f,onTop:g})=>styled.css`
  animation-name: ${a};
  animation-delay: ${b};
  animation-duration: ${c}ms;
  animation-timing-function: ${d};
  animation-fill-mode: ${e};
  transform-origin: ${f||"50% 50%"};

  ${g&&styled.css`
      z-index: 1;
    `}
`,stateMap={entering:({enterAnimation:a})=>styled.css`
      ${createAnimationStyles(a)};
    `,exiting:({exitAnimation:a})=>styled.css`
      ${createAnimationStyles(a)};
    `},PageTransitionGroup=styled__default.div`
  position: relative;
  width: 100%;
  height: 100%;
  perspective: 1200px;
  overflow: hidden;
`,PageTransitionWrapper=styled__default.div`
  backface-visibility: hidden;
  height: 100%;
  left: 0;
  overflow: hidden;
  position: absolute;
  top: 0;
  transform-style: preserve-3d;
  transform: translate3d(0, 0, 0);
  width: 100%;
  will-change: tranform;

  ${({state:a})=>stateMap[a]};
`;function PageTransition({children:a,enterAnimation:b,exitAnimation:c,preset:d,transitionKey:e,...f}){const g=(()=>b?"string"==typeof b?animations[b]:{...animations[b.name],delay:b.delay,onTop:b.onTop}:d?{...animations[presets[d].enter.name],delay:presets[d].enter.delay,onTop:presets[d].enter.onTop}:"rotateSlideIn")(),h=(()=>c?"string"==typeof c?animations[c]:{...animations[c.name],delay:c.delay,onTop:c.onTop}:d?{...animations[presets[d].exit.name],delay:presets[d].exit.delay,onTop:presets[d].exit.onTop}:"rotateSlideIn")(),i=Math.max(g.duration,h.duration);return(/*#__PURE__*/React__default.createElement(PageTransitionGroup,f,/*#__PURE__*/React__default.createElement(reactTransitionGroup.TransitionGroup,{component:null},/*#__PURE__*/React__default.createElement(reactTransitionGroup.Transition,{key:e,timeout:i},b=>/*#__PURE__*/React__default.createElement(PageTransitionWrapper,{enterAnimation:g,exitAnimation:h,state:b},a)))))}var PageTransition$1=/*#__PURE__*/React.memo(PageTransition);exports.PageTransition=PageTransition$1,exports.animations=animations,exports.presets=presets;
