// precise-ui/dist/es6/components/IndicatorKnob/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled, { themed } from '../../utils/styled';
import { distance } from '../../distance';
import { ocean, midnight } from '../../colors';
const transitionDuration = '0.3s';
const transitionDurationBorder = '0.2s';
const transitionEase = 'cubic-bezier(0, 0, 0.25, 1)';
const Indicator = styled('div') `
  width: ${distance.large};
  height: ${distance.large};
  position: absolute;
  border: ${themed(props => props.disabled || !props.focus ? 'none' : `${distance.xsmall} solid ${props.active ? midnight : props.theme.text1}`)};
  border-radius: 50%;
  background: ${themed(props => props.disabled ? props.theme.ui3 : props.active ? props.theme.ui0 : props.theme.text2)};
  transform: translate(-50%, -50%);
  cursor: ${props => (props.disabled ? 'not-allowed' : 'pointer')};
  box-sizing: border-box;
  transition: background ${transitionDuration} ${transitionEase},
    border ${transitionDurationBorder} ${transitionEase}
      ${props => (props.animate ? `, left ${transitionDuration} ${transitionEase}` : '')};

  &:hover {
    border: ${themed(props => props.disabled ? 'none' : `${distance.xsmall} solid ${props.active ? ocean : props.theme.ui5}`)};
  }
`;
function percent(value) {
    return typeof value === 'number' ? `${value * 100}%` : '50%';
}
/**
 * The indicator knob is a button like component that is placed in a container to indicate a position.
 */
export const IndicatorKnob = Object.assign(((_a) => {
    var { x, y, color, style } = _a, props = __rest(_a, ["x", "y", "color", "style"]);
    return (React.createElement(Indicator, Object.assign({ style: Object.assign({}, style, { left: percent(x), top: percent(y), backgroundColor: color }) }, props)));
}), { inner: {
        get Indicator() { return Indicator; }
    } });
IndicatorKnob.displayName = 'IndicatorKnob';
