// precise-ui/dist/es6/quarks/StyledInputRow.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import * as React from 'react';
import styled, { themed } from '../utils/styled';
import { distance } from '../distance';
import { purpleRed } from '../colors';
import { remCalc } from '../utils/remCalc';
const TextFieldBoxWithLabelWrapper = styled.div `
  flex-grow: 1;
  display: flex;
  flex-flow: column-reverse;
  height: 100%;
  position: relative;
  min-width: 0;
  margin: auto;
`;
const TextFieldLabel = styled.label `
  height: 100%;
`;
const TextFieldLabelText = styled('span') `
  display: block;
  font-size: ${remCalc('12px')};
  line-height: ${remCalc('16px')};
  max-width: 66.66%;
  box-sizing: border-box;
  cursor: text;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  transform-origin: left bottom;
  transform: ${props => (props.active || props.filled ? 'none' : 'translate(0, 0.85rem) scale(1.33)')};
  color: ${themed(({ theme, active, filled, error }) => error ? purpleRed : active || filled ? theme.ui0 : theme.text2)};
  transition: all 0.2s;
  position: absolute;
  top: 0;
  left: ${distance.medium};
  right: 0;
  padding-top: ${distance.small};
  background: ${themed(({ multiline, theme }) => (multiline ? theme.ui2 : 'transparent'))};

  > span {
    display: ${props => (props.active || props.filled ? 'none' : '')};
    transition: inherit;
  }
`;
export const StyledInputRow = Object.assign(((_a) => {
    var { children, label, focused = false, error = false, placeholder = '', hasValue = false, multiline = false } = _a, props = __rest(_a, ["children", "label", "focused", "error", "placeholder", "hasValue", "multiline"]);
    return (React.createElement(TextFieldBoxWithLabelWrapper, Object.assign({}, props), label !== undefined ? (React.createElement(TextFieldLabel, null,
        children,
        React.createElement(TextFieldLabelText, { active: focused, filled: !focused && hasValue, error: error, multiline: multiline },
            label,
            placeholder.length > 0 && React.createElement("span", null,
                " \u2014 ",
                placeholder)))) : (children)));
}), { inner: {
        get TextFieldBoxWithLabelWrapper() { return TextFieldBoxWithLabelWrapper; },
        get TextFieldLabel() { return TextFieldLabel; },
        get TextFieldLabelText() { return TextFieldLabelText; }
    } });
